{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook 3: Generate draft EcN model\n",
    "\n",
    "This notebook is based on the work of Norsigian _et al._ (2019). This notebook utilizes the output of notebook 2 (presence/absence matrix and geneID matrix) to generate a draft model of _E. coli_ Nissle from 55 reference models. This is done in the following steps: \n",
    "\n",
    "1. Deletion of missing genes/reaction from reference model iML1515 to generate draft model\n",
    "2. Update gene-protein-reaction rule\n",
    "3. Comparison with 54 reference models based on orthology\n",
    "4. Addition of reactions originating from genes encoding for proteins with >80% homology\n",
    "5. Update gene-protein-reaction rule after each model comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import package needed\n",
    "import cobra\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from glob import glob\n",
    "from cobra.io import load_json_model\n",
    "from cobra import Model, Reaction, Metabolite\n",
    "from cobra.manipulation.delete import delete_model_genes, remove_genes\n",
    "from cobra.manipulation.modify import rename_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Creation of EcN draft model from iML1515"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  1.1 Deletion of missing genes/reaction from reference model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iML1515</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x026d476de048</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1877</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2712</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*BIOMASS_Ec_iML1515_core_75p37M - 1.0*BIOMASS_Ec_iML1515_core_75p37M_reverse_35685</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, periplasm</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iML1515 at 0x26d476de048>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load the base E. coli reconstruction iML1515\n",
    "model = load_json_model('../data/models/iML1515.json')\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load the previously generated homology matrix for MG1655 vs. EcN\n",
    "hom_matrix=pd.read_csv('../tables/MG_ortho_matrix.csv')\n",
    "hom_matrix=hom_matrix.set_index('Unnamed: 0')\n",
    "\n",
    "# Give the ID of the assembly used\n",
    "EcN_ID = \"CP022686.1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CP022686.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2551</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0870</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3368</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2436</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3500</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            CP022686.1\n",
       "Unnamed: 0            \n",
       "b2551              1.0\n",
       "b0870              1.0\n",
       "b3368              1.0\n",
       "b2436              1.0\n",
       "b3500              1.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hom_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Delete missing genes from a copy of the iML1515 model\n",
    "Start with the iML1515 model, identify the missing genes from the matrix, and remove them and their associated reactions from the reference model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get the list of Gene IDs from the homology matrix dataframe for the current strain without a homolog\n",
    "currentStrain=hom_matrix[EcN_ID]\n",
    "nonHomologous=currentStrain[currentStrain==0.0]\n",
    "nonHomologous=nonHomologous.index.tolist()\n",
    "    \n",
    "#s0001 is an artificial gene used in iML1515 for spontaneous reactions and as such has no homologs,\n",
    "#However, it is retained for these spontaneous reactions to function\n",
    "nonHomologous.remove('s0001')\n",
    "    \n",
    "#Define a list of Gene objects from the base reconstruction to be deleted from the current strain\n",
    "toDelete=[]\n",
    "for gene in nonHomologous:\n",
    "    toDelete.append(model.genes.get_by_id(gene))\n",
    "\n",
    "#Establish a model copy and remove the reactions that should not be present according to bbh in notebook 2\n",
    "modelCopy=model.copy()\n",
    "remove_genes(modelCopy, toDelete, remove_reactions=True)\n",
    "modelCopy.id= EcN_ID\n",
    "\n",
    "# Save reduced iML1515 model as EcN model\n",
    "cobra.io.json.save_json_model(modelCopy, str('../data/models/%s_MG.json'%EcN_ID), pretty=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Record gene/reaction addition process "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Record which genes and reactions are present at this stage of the model building process\n",
    "rxn_origin_df = pd.DataFrame(columns=['reaction', 'origin', 'added', 'notebook', 'notes'])\n",
    "rxn_origin_df.set_index('reaction', inplace=True)\n",
    "\n",
    "gene_origin_df = pd.DataFrame(columns=['gene', 'EcN_gene', 'origin', 'added', 'notebook', 'notes'])\n",
    "gene_origin_df.set_index('gene', inplace=True)\n",
    "\n",
    "for gene in modelCopy.genes:\n",
    "    gene_origin_df.loc[gene.id, 'EcN_gene'] = gene.id\n",
    "    gene_origin_df.loc[gene.id, 'origin'] = 'iML1515'\n",
    "    gene_origin_df.loc[gene.id, 'added'] = 'automatic'\n",
    "    gene_origin_df.loc[gene.id, 'notebook'] = '3.1'\n",
    "    gene_origin_df.loc[gene.id, 'notes'] = 'none'\n",
    "        \n",
    "for rxn in modelCopy.reactions:\n",
    "    rxn_origin_df.loc[rxn.id, 'origin'] = 'iML1515'\n",
    "    rxn_origin_df.loc[rxn.id, 'added'] = 'automatic'\n",
    "    rxn_origin_df.loc[rxn.id, 'notebook'] = '3.1'\n",
    "    rxn_origin_df.loc[rxn.id, 'notes'] = 'none'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 Update Model Gene Product Rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CP022686.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2551</th>\n",
       "      <td>CIW80_06560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0870</th>\n",
       "      <td>CIW80_22360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3368</th>\n",
       "      <td>CIW80_11380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2436</th>\n",
       "      <td>CIW80_06055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3500</th>\n",
       "      <td>CIW80_12105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             CP022686.1\n",
       "Unnamed: 0             \n",
       "b2551       CIW80_06560\n",
       "b0870       CIW80_22360\n",
       "b3368       CIW80_11380\n",
       "b2436       CIW80_06055\n",
       "b3500       CIW80_12105"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the geneIDs, set index and remove all genes == None\n",
    "MG_geneIDs_matrix = pd.read_csv('../tables/MG_geneIDs_matrix.csv')\n",
    "MG_geneIDs_matrix.set_index('Unnamed: 0', inplace=True)\n",
    "MG_geneIDs_matrix = MG_geneIDs_matrix[MG_geneIDs_matrix.index != 'None']\n",
    "\n",
    "MG_geneIDs_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utilize the geneIDs matrix to update the GPRs in the EcN model with the proper gene ID\n",
    "mod = '../data/models/%s_MG.json'%EcN_ID\n",
    "model=cobra.io.load_json_model(mod)\n",
    "\n",
    "IDMapping = MG_geneIDs_matrix[EcN_ID].to_dict()\n",
    "IDMappingParsed = {k:v for k,v in IDMapping.items() if v != 'None'}\n",
    "rename_genes(model,IDMappingParsed)\n",
    "\n",
    "# Save the model\n",
    "cobra.io.json.save_json_model(model,mod, pretty=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>origin</th>\n",
       "      <th>added</th>\n",
       "      <th>notebook</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reaction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALATA_D2</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SHCHD2</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPPPGO</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTHOr</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHORD5</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUCCt1pp</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>QUINDH</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LCARSyi</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIOMASS_Ec_iML1515_core_75p37M</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIOMASS_Ec_iML1515_WT_75p37M</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2607 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 origin      added notebook notes\n",
       "reaction                                                         \n",
       "ALATA_D2                        iML1515  automatic      3.1  none\n",
       "SHCHD2                          iML1515  automatic      3.1  none\n",
       "CPPPGO                          iML1515  automatic      3.1  none\n",
       "GTHOr                           iML1515  automatic      3.1  none\n",
       "DHORD5                          iML1515  automatic      3.1  none\n",
       "...                                 ...        ...      ...   ...\n",
       "SUCCt1pp                        iML1515  automatic      3.1  none\n",
       "QUINDH                          iML1515  automatic      3.1  none\n",
       "LCARSyi                         iML1515  automatic      3.1  none\n",
       "BIOMASS_Ec_iML1515_core_75p37M  iML1515  automatic      3.1  none\n",
       "BIOMASS_Ec_iML1515_WT_75p37M    iML1515  automatic      3.1  none\n",
       "\n",
       "[2607 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxn_origin_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.5 Record new gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename the gene names to EcN IDs in the origin table\n",
    "gene_origin_df['EcN_gene'] = gene_origin_df['EcN_gene'].map(IDMapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save as a table\n",
    "rxn_origin_df.to_csv('../tables/rxn_origin.csv')\n",
    "gene_origin_df.to_csv('../tables/gene_origin.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.6 Examine the draft strain specific model contents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CP022686.1 Number of Model Genes: 1390 Number of Model Reactions: 2607\n"
     ]
    }
   ],
   "source": [
    "# Gather the general information on the draft models\n",
    "print (modelCopy.id,'Number of Model Genes:',len(modelCopy.genes),'Number of Model Reactions:',len(model.reactions))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Gene addition to draft EcN model\n",
    "Per strain:\n",
    "1. Identify IDs of all gene homologies <ortho == 1>\n",
    "2. Create a list of all orthologous genes (def ortho_to_list)\n",
    "    \n",
    "Within “strain loop”: Per reaction in ortho list\n",
    "1. Relate gene ID to reaction in strain’s model\n",
    "2. Check presence of the reaction in EcN draft model\n",
    "3. If not present, check homology of all genes related to reaction\n",
    "4. If all genes homologous, add reaction to draft model\n",
    "    \n",
    "Within “strain loop”: After reactions are added:\n",
    "1. Update the GPRs to Nissle gene IDs\n",
    "2. Add the genes/reactions to the overview table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 OrthoIDs_matrix creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CU651637</th>\n",
       "      <th>CP001855</th>\n",
       "      <th>CP002167</th>\n",
       "      <th>CP000468</th>\n",
       "      <th>CP000946</th>\n",
       "      <th>CP000819</th>\n",
       "      <th>CP001665</th>\n",
       "      <th>AM946981</th>\n",
       "      <th>CP001509</th>\n",
       "      <th>CP001396</th>\n",
       "      <th>...</th>\n",
       "      <th>CU928163</th>\n",
       "      <th>CP000243</th>\n",
       "      <th>CP001063</th>\n",
       "      <th>CP000036</th>\n",
       "      <th>CP000034</th>\n",
       "      <th>CP001383</th>\n",
       "      <th>AE014073</th>\n",
       "      <th>AE005674</th>\n",
       "      <th>CP000266</th>\n",
       "      <th>CP000038</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CIW80_00005</th>\n",
       "      <td>LF82_1032</td>\n",
       "      <td>NRG857_06995</td>\n",
       "      <td>UM146_09940</td>\n",
       "      <td>APECO1_564</td>\n",
       "      <td>EcolC_2245</td>\n",
       "      <td>ECB_01368</td>\n",
       "      <td>ECBD_2227</td>\n",
       "      <td>B21_01380</td>\n",
       "      <td>ECD_01368</td>\n",
       "      <td>BWG_1240</td>\n",
       "      <td>...</td>\n",
       "      <td>ECUMN_1659</td>\n",
       "      <td>UTI89_C1635</td>\n",
       "      <td>SbBS512_E1633</td>\n",
       "      <td>SBO_1674</td>\n",
       "      <td>SDY_1774</td>\n",
       "      <td>SFxv_2017</td>\n",
       "      <td>S1472</td>\n",
       "      <td>SF1802</td>\n",
       "      <td>SFV_1794</td>\n",
       "      <td>SSON_1733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_00010</th>\n",
       "      <td>LF82_2792</td>\n",
       "      <td>NRG857_07000</td>\n",
       "      <td>UM146_09935</td>\n",
       "      <td>APECO1_565</td>\n",
       "      <td>EcolC_2244</td>\n",
       "      <td>ECB_01369</td>\n",
       "      <td>ECBD_2226</td>\n",
       "      <td>B21_01381</td>\n",
       "      <td>ECD_01369</td>\n",
       "      <td>BWG_1241</td>\n",
       "      <td>...</td>\n",
       "      <td>ECUMN_1662</td>\n",
       "      <td>UTI89_C1636</td>\n",
       "      <td>SbBS512_E1634</td>\n",
       "      <td>SBO_1673</td>\n",
       "      <td>None</td>\n",
       "      <td>SFxv_2015</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SFV_1792</td>\n",
       "      <td>SSON_1732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_00015</th>\n",
       "      <td>LF82_0064</td>\n",
       "      <td>NRG857_07005</td>\n",
       "      <td>UM146_09930</td>\n",
       "      <td>APECO1_566</td>\n",
       "      <td>EcolC_2243</td>\n",
       "      <td>ECB_01370</td>\n",
       "      <td>ECBD_2225</td>\n",
       "      <td>B21_01382</td>\n",
       "      <td>ECD_01370</td>\n",
       "      <td>BWG_1242</td>\n",
       "      <td>...</td>\n",
       "      <td>ECUMN_1663</td>\n",
       "      <td>UTI89_C1637</td>\n",
       "      <td>SbBS512_E1635</td>\n",
       "      <td>SBO_1672</td>\n",
       "      <td>None</td>\n",
       "      <td>SFxv_2014</td>\n",
       "      <td>S1475</td>\n",
       "      <td>SF1797</td>\n",
       "      <td>SFV_1791</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_00020</th>\n",
       "      <td>LF82_0802</td>\n",
       "      <td>NRG857_07010</td>\n",
       "      <td>UM146_09925</td>\n",
       "      <td>APECO1_567</td>\n",
       "      <td>EcolC_2242</td>\n",
       "      <td>ECB_01371</td>\n",
       "      <td>ECBD_2224</td>\n",
       "      <td>B21_01383</td>\n",
       "      <td>ECD_01371</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>ECUMN_1664</td>\n",
       "      <td>UTI89_C1638</td>\n",
       "      <td>SbBS512_E1636</td>\n",
       "      <td>SBO_1671</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SF1795</td>\n",
       "      <td>SFV_1790</td>\n",
       "      <td>SSON_1725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_00025</th>\n",
       "      <td>LF82_0397</td>\n",
       "      <td>NRG857_07015</td>\n",
       "      <td>UM146_09920</td>\n",
       "      <td>APECO1_568</td>\n",
       "      <td>EcolC_2241</td>\n",
       "      <td>ECB_01374</td>\n",
       "      <td>ECBD_2221</td>\n",
       "      <td>B21_01386</td>\n",
       "      <td>ECD_01374</td>\n",
       "      <td>BWG_1243</td>\n",
       "      <td>...</td>\n",
       "      <td>ECUMN_1665</td>\n",
       "      <td>UTI89_C1639</td>\n",
       "      <td>SbBS512_E1642</td>\n",
       "      <td>SBO_1669</td>\n",
       "      <td>SDY_1764</td>\n",
       "      <td>SFxv_2012</td>\n",
       "      <td>S1478</td>\n",
       "      <td>SF1794</td>\n",
       "      <td>SFV_1789</td>\n",
       "      <td>SSON_1724</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              CU651637      CP001855     CP002167    CP000468    CP000946  \\\n",
       "Unnamed: 0                                                                  \n",
       "CIW80_00005  LF82_1032  NRG857_06995  UM146_09940  APECO1_564  EcolC_2245   \n",
       "CIW80_00010  LF82_2792  NRG857_07000  UM146_09935  APECO1_565  EcolC_2244   \n",
       "CIW80_00015  LF82_0064  NRG857_07005  UM146_09930  APECO1_566  EcolC_2243   \n",
       "CIW80_00020  LF82_0802  NRG857_07010  UM146_09925  APECO1_567  EcolC_2242   \n",
       "CIW80_00025  LF82_0397  NRG857_07015  UM146_09920  APECO1_568  EcolC_2241   \n",
       "\n",
       "              CP000819   CP001665   AM946981   CP001509  CP001396  ...  \\\n",
       "Unnamed: 0                                                         ...   \n",
       "CIW80_00005  ECB_01368  ECBD_2227  B21_01380  ECD_01368  BWG_1240  ...   \n",
       "CIW80_00010  ECB_01369  ECBD_2226  B21_01381  ECD_01369  BWG_1241  ...   \n",
       "CIW80_00015  ECB_01370  ECBD_2225  B21_01382  ECD_01370  BWG_1242  ...   \n",
       "CIW80_00020  ECB_01371  ECBD_2224  B21_01383  ECD_01371      None  ...   \n",
       "CIW80_00025  ECB_01374  ECBD_2221  B21_01386  ECD_01374  BWG_1243  ...   \n",
       "\n",
       "               CU928163     CP000243       CP001063  CP000036  CP000034  \\\n",
       "Unnamed: 0                                                                \n",
       "CIW80_00005  ECUMN_1659  UTI89_C1635  SbBS512_E1633  SBO_1674  SDY_1774   \n",
       "CIW80_00010  ECUMN_1662  UTI89_C1636  SbBS512_E1634  SBO_1673      None   \n",
       "CIW80_00015  ECUMN_1663  UTI89_C1637  SbBS512_E1635  SBO_1672      None   \n",
       "CIW80_00020  ECUMN_1664  UTI89_C1638  SbBS512_E1636  SBO_1671      None   \n",
       "CIW80_00025  ECUMN_1665  UTI89_C1639  SbBS512_E1642  SBO_1669  SDY_1764   \n",
       "\n",
       "              CP001383 AE014073 AE005674  CP000266   CP000038  \n",
       "Unnamed: 0                                                     \n",
       "CIW80_00005  SFxv_2017    S1472   SF1802  SFV_1794  SSON_1733  \n",
       "CIW80_00010  SFxv_2015     None     None  SFV_1792  SSON_1732  \n",
       "CIW80_00015  SFxv_2014    S1475   SF1797  SFV_1791       None  \n",
       "CIW80_00020       None     None   SF1795  SFV_1790  SSON_1725  \n",
       "CIW80_00025  SFxv_2012    S1478   SF1794  SFV_1789  SSON_1724  \n",
       "\n",
       "[5 rows x 55 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the number of genes represented within the 55 models\n",
    "geneIDs_matrix = pd.read_csv('../tables/geneIDs_matrix.csv')\n",
    "geneIDs_matrix.set_index('Unnamed: 0', inplace=True)\n",
    "\n",
    "# Combine with ortho_matrix and remove all genes that do not have homology above 80%\n",
    "ortho_matrix = pd.read_csv('../tables/ortho_matrix.csv')\n",
    "ortho_matrix.set_index('Unnamed: 0', inplace=True)\n",
    "\n",
    "for strain in ortho_matrix.columns:\n",
    "    geneIDs_matrix[strain] = np.where(ortho_matrix[strain] == 0, 'None', geneIDs_matrix[strain])\n",
    "\n",
    "geneIDs_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save as orthoIDs_matrix\n",
    "orthoIDs_matrix = geneIDs_matrix\n",
    "orthoIDs_matrix.to_csv('../tables/orthoIDs_matrix.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Gene addition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# From the ortho_matrix, create a list of genes for the CurrentStrain that have homology with EcN genes \n",
    "def ortho_to_list(CurrentStrain, geneIDs_matrix):\n",
    "\n",
    "    # Load orthology matrix and drop all similarities with existing EcN model\n",
    "    ortho_matrix = pd.read_csv('../tables/ortho_matrix.csv')\n",
    "    ortho_matrix.set_index('Unnamed: 0', inplace=True)\n",
    "\n",
    "    # Rename the current strain in the geneIDs_matrix\n",
    "    current_geneIDs_matrix = geneIDs_matrix[CurrentStrain].rename('geneID')\n",
    "\n",
    "    # Merge ortho_matrix with current geneIDs matrix\n",
    "    ortho_matrix = pd.merge(ortho_matrix, current_geneIDs_matrix, left_index=True, right_index=True)\n",
    "    ortho_matrix.set_index('geneID', inplace=True)\n",
    "    try: ortho_matrix.drop('None', inplace=True)\n",
    "    except: pass\n",
    "\n",
    "    # Create list of all genes that are not yet in model and have homology with current strain\n",
    "    ortho_list = ortho_matrix[ortho_matrix[CurrentStrain] == 1.0].index.tolist()\n",
    "    return(ortho_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Based on the gene IDs in the ortho_list, retrieve reactions from the current_model and add them to the EcN model\n",
    "def gene_addition(modelCopy, current_model, ortho_list):\n",
    "    add_gene = []\n",
    "    \n",
    "    # Record which genes and reactions are added and from which model at this stage of the model building process\n",
    "    g_origin_df = pd.DataFrame(columns=['gene', 'EcN_gene', 'origin', 'added', 'notebook', 'notes'])\n",
    "    g_origin_df.set_index('gene', inplace=True)\n",
    "    \n",
    "    r_origin_df = pd.DataFrame(columns=['reaction', 'origin', 'added', 'notebook', 'notes'])\n",
    "    r_origin_df.set_index('reaction', inplace=True)\n",
    "    \n",
    "    # for all genes that have homology, check if they should be added\n",
    "    for gene in ortho_list:\n",
    "        try:\n",
    "            reactions = current_model.genes.get_by_id(gene).reactions\n",
    "            for reaction in reactions:\n",
    "                \n",
    "                # check if the gene is already present in the model\n",
    "                try:\n",
    "                    modelCopy.reactions.get_by_id(reaction.id)\n",
    "                \n",
    "                # If the reaction is not yet in the model, check the dependencies of genes linked to the reaction\n",
    "                # If the model contains an \"and\" statement, check if all involved genes are present in the ortho_list\n",
    "                # If they are not all present, do not add the reaction\n",
    "                except: \n",
    "                    grr = current_model.reactions.get_by_id(reaction.id).gene_reaction_rule # get gene reaction rule\n",
    "                    or_statement = grr.split(' or ') #split based on \"or\" statement\n",
    "                    for or_genes in or_statement:\n",
    "                        if len(or_genes.split(' and ')) == 1: #check whether there is an \"and\" statement\n",
    "                            add_gene.append(True) #if not there are only \"or\" statements, and having one of them in the ortho_list is sufficient\n",
    "                        else:\n",
    "                            and_present = True\n",
    "                            involved_genes = or_genes.split(' and ') #create a list of all the genes in the \"and\" statement\n",
    "                            for gene_x in involved_genes:\n",
    "                                if gene_x.lstrip('(').rstrip(')') not in ortho_list: #check for each gene if it's present in the ortho_list\n",
    "                                    and_present = False #if one of the genes is not present, do not give True for addition on this part\n",
    "                            if and_present == True:\n",
    "                                add_gene.append(True)\n",
    "                                \n",
    "                    if True in add_gene: #only if one of the \"or\" statements (one gene or combination of \"and\" genes) is completely present, add reaction\n",
    "                        modelCopy.add_reactions([current_model.reactions.get_by_id(reaction.id)]) # if not present, add\n",
    "                        print('added:', reaction.id)\n",
    "                        \n",
    "                        # Add gene & reaction to overview table\n",
    "                        g_origin_df.loc[gene, 'EcN_gene'] = gene\n",
    "                        g_origin_df.loc[gene, 'origin'] = current_model\n",
    "                        g_origin_df.loc[gene, 'added'] = 'automatic'\n",
    "                        g_origin_df.loc[gene, 'notebook'] = '3.1'\n",
    "                        g_origin_df.loc[gene, 'notes'] = 'none'\n",
    "                        \n",
    "                        r_origin_df.loc[reaction.id, 'origin'] = current_model\n",
    "                        r_origin_df.loc[reaction.id, 'added'] = 'automatic'\n",
    "                        r_origin_df.loc[reaction.id, 'notebook'] = '3.1'\n",
    "                        r_origin_df.loc[reaction.id, 'notes'] = 'none'\n",
    "                        \n",
    "                        # Reset the \"add_gene\" list\n",
    "                        add_gene = []\n",
    "                    else:\n",
    "#                         print('Not added:', reaction.id)\n",
    "                        continue\n",
    "        except:\n",
    "            continue\n",
    "    return(modelCopy, g_origin_df, r_origin_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CU651637</th>\n",
       "      <th>CP001855</th>\n",
       "      <th>CP002167</th>\n",
       "      <th>CP000468</th>\n",
       "      <th>CP000946</th>\n",
       "      <th>CP000819</th>\n",
       "      <th>CP001665</th>\n",
       "      <th>AM946981</th>\n",
       "      <th>CP001509</th>\n",
       "      <th>CP001396</th>\n",
       "      <th>...</th>\n",
       "      <th>CU928163</th>\n",
       "      <th>CP000243</th>\n",
       "      <th>CP001063</th>\n",
       "      <th>CP000036</th>\n",
       "      <th>CP000034</th>\n",
       "      <th>CP001383</th>\n",
       "      <th>AE014073</th>\n",
       "      <th>AE005674</th>\n",
       "      <th>CP000266</th>\n",
       "      <th>CP000038</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CIW80_00005</th>\n",
       "      <td>LF82_1032</td>\n",
       "      <td>NRG857_06995</td>\n",
       "      <td>UM146_09940</td>\n",
       "      <td>APECO1_564</td>\n",
       "      <td>EcolC_2245</td>\n",
       "      <td>ECB_01368</td>\n",
       "      <td>ECBD_2227</td>\n",
       "      <td>B21_01380</td>\n",
       "      <td>ECD_01368</td>\n",
       "      <td>BWG_1240</td>\n",
       "      <td>...</td>\n",
       "      <td>ECUMN_1659</td>\n",
       "      <td>UTI89_C1635</td>\n",
       "      <td>SbBS512_E1633</td>\n",
       "      <td>SBO_1674</td>\n",
       "      <td>SDY_1774</td>\n",
       "      <td>SFxv_2017</td>\n",
       "      <td>S1472</td>\n",
       "      <td>SF1802</td>\n",
       "      <td>SFV_1794</td>\n",
       "      <td>SSON_1733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_00010</th>\n",
       "      <td>LF82_2792</td>\n",
       "      <td>NRG857_07000</td>\n",
       "      <td>UM146_09935</td>\n",
       "      <td>APECO1_565</td>\n",
       "      <td>EcolC_2244</td>\n",
       "      <td>ECB_01369</td>\n",
       "      <td>ECBD_2226</td>\n",
       "      <td>B21_01381</td>\n",
       "      <td>ECD_01369</td>\n",
       "      <td>BWG_1241</td>\n",
       "      <td>...</td>\n",
       "      <td>ECUMN_1662</td>\n",
       "      <td>UTI89_C1636</td>\n",
       "      <td>SbBS512_E1634</td>\n",
       "      <td>SBO_1673</td>\n",
       "      <td>None</td>\n",
       "      <td>SFxv_2015</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SFV_1792</td>\n",
       "      <td>SSON_1732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_00015</th>\n",
       "      <td>LF82_0064</td>\n",
       "      <td>NRG857_07005</td>\n",
       "      <td>UM146_09930</td>\n",
       "      <td>APECO1_566</td>\n",
       "      <td>EcolC_2243</td>\n",
       "      <td>ECB_01370</td>\n",
       "      <td>ECBD_2225</td>\n",
       "      <td>B21_01382</td>\n",
       "      <td>ECD_01370</td>\n",
       "      <td>BWG_1242</td>\n",
       "      <td>...</td>\n",
       "      <td>ECUMN_1663</td>\n",
       "      <td>UTI89_C1637</td>\n",
       "      <td>SbBS512_E1635</td>\n",
       "      <td>SBO_1672</td>\n",
       "      <td>None</td>\n",
       "      <td>SFxv_2014</td>\n",
       "      <td>S1475</td>\n",
       "      <td>SF1797</td>\n",
       "      <td>SFV_1791</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_00020</th>\n",
       "      <td>LF82_0802</td>\n",
       "      <td>NRG857_07010</td>\n",
       "      <td>UM146_09925</td>\n",
       "      <td>APECO1_567</td>\n",
       "      <td>EcolC_2242</td>\n",
       "      <td>ECB_01371</td>\n",
       "      <td>ECBD_2224</td>\n",
       "      <td>B21_01383</td>\n",
       "      <td>ECD_01371</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>ECUMN_1664</td>\n",
       "      <td>UTI89_C1638</td>\n",
       "      <td>SbBS512_E1636</td>\n",
       "      <td>SBO_1671</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SF1795</td>\n",
       "      <td>SFV_1790</td>\n",
       "      <td>SSON_1725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_00025</th>\n",
       "      <td>LF82_0397</td>\n",
       "      <td>NRG857_07015</td>\n",
       "      <td>UM146_09920</td>\n",
       "      <td>APECO1_568</td>\n",
       "      <td>EcolC_2241</td>\n",
       "      <td>ECB_01374</td>\n",
       "      <td>ECBD_2221</td>\n",
       "      <td>B21_01386</td>\n",
       "      <td>ECD_01374</td>\n",
       "      <td>BWG_1243</td>\n",
       "      <td>...</td>\n",
       "      <td>ECUMN_1665</td>\n",
       "      <td>UTI89_C1639</td>\n",
       "      <td>SbBS512_E1642</td>\n",
       "      <td>SBO_1669</td>\n",
       "      <td>SDY_1764</td>\n",
       "      <td>SFxv_2012</td>\n",
       "      <td>S1478</td>\n",
       "      <td>SF1794</td>\n",
       "      <td>SFV_1789</td>\n",
       "      <td>SSON_1724</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              CU651637      CP001855     CP002167    CP000468    CP000946  \\\n",
       "Unnamed: 0                                                                  \n",
       "CIW80_00005  LF82_1032  NRG857_06995  UM146_09940  APECO1_564  EcolC_2245   \n",
       "CIW80_00010  LF82_2792  NRG857_07000  UM146_09935  APECO1_565  EcolC_2244   \n",
       "CIW80_00015  LF82_0064  NRG857_07005  UM146_09930  APECO1_566  EcolC_2243   \n",
       "CIW80_00020  LF82_0802  NRG857_07010  UM146_09925  APECO1_567  EcolC_2242   \n",
       "CIW80_00025  LF82_0397  NRG857_07015  UM146_09920  APECO1_568  EcolC_2241   \n",
       "\n",
       "              CP000819   CP001665   AM946981   CP001509  CP001396  ...  \\\n",
       "Unnamed: 0                                                         ...   \n",
       "CIW80_00005  ECB_01368  ECBD_2227  B21_01380  ECD_01368  BWG_1240  ...   \n",
       "CIW80_00010  ECB_01369  ECBD_2226  B21_01381  ECD_01369  BWG_1241  ...   \n",
       "CIW80_00015  ECB_01370  ECBD_2225  B21_01382  ECD_01370  BWG_1242  ...   \n",
       "CIW80_00020  ECB_01371  ECBD_2224  B21_01383  ECD_01371      None  ...   \n",
       "CIW80_00025  ECB_01374  ECBD_2221  B21_01386  ECD_01374  BWG_1243  ...   \n",
       "\n",
       "               CU928163     CP000243       CP001063  CP000036  CP000034  \\\n",
       "Unnamed: 0                                                                \n",
       "CIW80_00005  ECUMN_1659  UTI89_C1635  SbBS512_E1633  SBO_1674  SDY_1774   \n",
       "CIW80_00010  ECUMN_1662  UTI89_C1636  SbBS512_E1634  SBO_1673      None   \n",
       "CIW80_00015  ECUMN_1663  UTI89_C1637  SbBS512_E1635  SBO_1672      None   \n",
       "CIW80_00020  ECUMN_1664  UTI89_C1638  SbBS512_E1636  SBO_1671      None   \n",
       "CIW80_00025  ECUMN_1665  UTI89_C1639  SbBS512_E1642  SBO_1669  SDY_1764   \n",
       "\n",
       "              CP001383 AE014073 AE005674  CP000266   CP000038  \n",
       "Unnamed: 0                                                     \n",
       "CIW80_00005  SFxv_2017    S1472   SF1802  SFV_1794  SSON_1733  \n",
       "CIW80_00010  SFxv_2015     None     None  SFV_1792  SSON_1732  \n",
       "CIW80_00015  SFxv_2014    S1475   SF1797  SFV_1791       None  \n",
       "CIW80_00020       None     None   SF1795  SFV_1790  SSON_1725  \n",
       "CIW80_00025  SFxv_2012    S1478   SF1794  SFV_1789  SSON_1724  \n",
       "\n",
       "[5 rows x 55 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the MG1655 model from which all reactions not present in EcN were removed\n",
    "EcN_model = cobra.io.load_json_model('../data/models/%s_MG.json'%EcN_ID)\n",
    "modelCopy = EcN_model.copy()\n",
    "\n",
    "# Load the table with information on origin of genes & reactions\n",
    "gene_origin_df = pd.read_csv('../tables/gene_origin.csv')\n",
    "gene_origin_df.set_index('gene', inplace=True)\n",
    "\n",
    "rxn_origin_df = pd.read_csv('../tables/rxn_origin.csv')\n",
    "rxn_origin_df.set_index('reaction', inplace=True)\n",
    "\n",
    "# Load strain info to get the model for each strain and remove MG1655, so it will not be used in the comparison\n",
    "strain_info = pd.read_csv('../tables/strain_info.csv', usecols=[1,3,6])\n",
    "strain_info = strain_info[strain_info['NCBI ID'] != 'U00096']\n",
    "\n",
    "#load the orthoID matrix. Set index, rename EcN and remove all genes == None\n",
    "orthoIDs_matrix = pd.read_csv('../tables/orthoIDs_matrix.csv')\n",
    "orthoIDs_matrix = orthoIDs_matrix.set_index('Unnamed: 0') #Set EcN gene IDs as index\n",
    "orthoIDs_matrix = orthoIDs_matrix[orthoIDs_matrix.index != 'None'] # Remove all indexes 00 'None'\n",
    "\n",
    "orthoIDs_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding reactions from strain:  CU651637 , model: iLF82_1304\n",
      "added: DARBt3ipp\n",
      "added: CLB6PH\n",
      "added: CLBH\n",
      "added: SALCNH\n",
      "added: CLBtpp\n",
      "added: SALCtpp\n",
      "added: CLBptspp\n",
      "added: SALCptspp\n",
      "added: CELLBpts_1\n",
      "added: TYRTRS\n",
      "added: GTHPi_copy2\n",
      "added: GTHPi_copy1\n",
      "added: PHETRS\n",
      "added: THRTRS\n",
      "added: TAG1PK\n",
      "added: DC6PH\n",
      "added: CHITPH\n",
      "added: CHITOBpts\n",
      "added: ASPTRS\n",
      "added: ARGTRS\n",
      "added: DARBabcpp\n",
      "added: SALSYN\n",
      "added: FEpp\n",
      "added: FEex\n",
      "added: ADOCBLPS\n",
      "added: METTRS\n",
      "added: SALCHS4FEtonex\n",
      "added: DMALRED\n",
      "added: PACt1\n",
      "added: GULNLtex\n",
      "added: 3PGtex\n",
      "added: TAGtex\n",
      "added: RAFFtex\n",
      "added: 4HTHRtex\n",
      "added: PENGt1\n",
      "added: 3hoxpactex\n",
      "added: HPAtex\n",
      "added: DRIBtex\n",
      "added: TETtex\n",
      "added: DARBtex\n",
      "added: DXYLUDtex\n",
      "added: ICITtex\n",
      "added: 2PGtex\n",
      "added: OAAtex\n",
      "added: GALAMtex\n",
      "added: AEPtex\n",
      "added: AIRStex\n",
      "added: SALCNtex\n",
      "added: FOAMtex\n",
      "added: 6APAt1\n",
      "added: ACONCtex\n",
      "added: SALCHSXtex\n",
      "added: 4HOXPACt2pp\n",
      "added: UPLA4NT\n",
      "added: NPHS\n",
      "added: HACD7i\n",
      "added: HACD5i\n",
      "added: HACD1i\n",
      "added: HACD8i\n",
      "added: HACD6i\n",
      "added: DHACOAH\n",
      "added: HACD3i\n",
      "added: HACD4i\n",
      "added: HACD2i\n",
      "added: TTDCAtexi\n",
      "added: DDCAtexi\n",
      "added: TTDCEAtexi\n",
      "added: HDCAtexi\n",
      "added: OCDCAtexi\n",
      "added: HDCEAtexi\n",
      "added: HDCAt2pp\n",
      "added: OCDCEAtexi\n",
      "added: OCDCEAt2pp\n",
      "added: OCDCAt2pp\n",
      "added: TTDCAt2pp\n",
      "added: DDCAt2pp\n",
      "added: INSt2pp_copy2\n",
      "added: SLCYSS\n",
      "added: SLNTtpp\n",
      "added: SELtpp\n",
      "added: PRASCSi_copy1\n",
      "added: PRASCSi_copy2\n",
      "added: URAt2pp_copy1\n",
      "added: HISTRS\n",
      "added: MECDPDH2\n",
      "added: THZPSN\n",
      "added: PSUDS\n",
      "added: OXPTNDH\n",
      "added: ALATRS\n",
      "added: NMNR\n",
      "added: DARBt2rpp\n",
      "added: 3OXCOAT\n",
      "added: LYSTRS\n",
      "added: PFOR\n",
      "added: FLDR\n",
      "added: OMPHHX\n",
      "added: PPCSCT\n",
      "added: MMM2\n",
      "added: INSt2pp_copy1\n",
      "added: 4HTHRtrpp\n",
      "added: PPRDNDH\n",
      "added: ACGALptspp\n",
      "added: GALAMPTSpp\n",
      "added: GALAM6PISO\n",
      "added: ACGAL6PI\n",
      "added: ACGAL6PISO\n",
      "added: ENLIPIDAt1ex\n",
      "added: FMETTRS\n",
      "added: DSCLCOCH\n",
      "added: TRPTRS\n",
      "added: OAAt2_2pp\n",
      "added: GLYTRS\n",
      "added: VPAMTr_copy1\n",
      "added: VPAMTr_copy2\n",
      "added: XYLUt2pp\n",
      "added: 4ABUTD\n",
      "added: ALDD2x\n",
      "added: ALDD31\n",
      "added: GLYALDDr\n",
      "added: IMACTD\n",
      "added: GLXO1\n",
      "added: SELCYSS\n",
      "added: GLCR1TRB1\n",
      "added: GLCR1TRA1\n",
      "added: GALR1TRB2\n",
      "added: GALR1TRA2\n",
      "added: GLCTR3\n",
      "added: HEPKB2\n",
      "added: HEPKA2\n",
      "added: GALR1TRB1\n",
      "added: GALR1TRA1\n",
      "added: HEPT3\n",
      "added: AMMQT8_2\n",
      "added: URFGTT\n",
      "added: OPHHX\n",
      "added: CTECOAI7_copy1\n",
      "added: CTECOAI7_copy2\n",
      "added: CTECOAI6_copy1\n",
      "added: CTECOAI8_copy1\n",
      "added: CTECOAI8_copy2\n",
      "added: CTECOAI6_copy2\n",
      "added: GHBDHx\n",
      "added: NICRNS\n",
      "added: NMNHYD\n",
      "added: NACODA\n",
      "added: SBDH\n",
      "added: TRPTA\n",
      "added: UNK3\n",
      "added: PDHcr\n",
      "added: ASPt2pp_copy1\n",
      "added: GLYBt2pp_copy1\n",
      "added: GLYBt2pp_copy2\n",
      "added: 3HBCOAHL\n",
      "added: GLYt2pp_copy1\n",
      "added: ALAt2pp_copy1\n",
      "added: ARGDr\n",
      "added: VALTRS\n",
      "added: RNMK\n",
      "added: 4HTHRS\n",
      "added: ILETRS\n",
      "added: ENLIPIDAt2ex\n",
      "added: URIDK2r\n",
      "added: PROTRS\n",
      "added: LEUt4rpp\n",
      "added: DADK_copy2\n",
      "added: DADK_copy1\n",
      "added: ALLTAMH\n",
      "added: CYSTRS\n",
      "added: AGMPTRCtpp\n",
      "added: ENTERES2\n",
      "added: ENTERH\n",
      "added: SALCHS4FEabcpp\n",
      "added: FE3DHBZSabcpp\n",
      "added: SALCHS2FEabcpp\n",
      "added: ADOCBLPP\n",
      "added: LEUTRS\n",
      "added: OMMBLHX\n",
      "added: GLNTRS\n",
      "added: GLCOAS\n",
      "added: RNAMtpp\n",
      "added: SERTRS\n",
      "added: SERTRS2\n",
      "added: ASNTRS\n",
      "added: URAt2pp_copy2\n",
      "added: PROD2\n",
      "Number of Model Genes: 1450 Number of Model Reactions: 2792 \n",
      "\n",
      "Adding reactions from strain:  CP001855 , model: iNRG857_1313\n",
      "Number of Model Genes: 1450 Number of Model Reactions: 2792 \n",
      "\n",
      "Adding reactions from strain:  CP002167 , model: iUMN146_1321\n",
      "added: AH6PI\n",
      "added: PPOR\n",
      "added: ACOAD5\n",
      "added: ACOAD3\n",
      "added: ACOAD7\n",
      "added: ACOAD4\n",
      "added: ACOAD2\n",
      "added: ACOAD6\n",
      "added: SALCHS2FEtonex\n",
      "added: FEENTER2tpp\n",
      "added: SALCHS4abcpp\n",
      "added: ENTERGLCT2\n",
      "added: ENTERGLCT1\n",
      "added: MNabc_1\n",
      "Number of Model Genes: 1459 Number of Model Reactions: 2806 \n",
      "\n",
      "Adding reactions from strain:  CP000468 , model: iAPECO1_1312\n",
      "Number of Model Genes: 1459 Number of Model Reactions: 2806 \n",
      "\n",
      "Adding reactions from strain:  CP000946 , model: iEcolC_1368\n",
      "added: ADNt2pp\n",
      "added: THMDt2pp\n",
      "added: FADRx_copy1\n",
      "Number of Model Genes: 1459 Number of Model Reactions: 2809 \n",
      "\n",
      "Adding reactions from strain:  CP000819 , model: iECB_1328\n",
      "added: HEPK2\n",
      "added: GLCTR2\n",
      "Number of Model Genes: 1460 Number of Model Reactions: 2811 \n",
      "\n",
      "Adding reactions from strain:  CP001665 , model: iECBD_1354\n",
      "Number of Model Genes: 1460 Number of Model Reactions: 2811 \n",
      "\n",
      "Adding reactions from strain:  AM946981 , model: iB21_1397\n",
      "Number of Model Genes: 1460 Number of Model Reactions: 2811 \n",
      "\n",
      "Adding reactions from strain:  CP001509 , model: iECD_1391\n",
      "Number of Model Genes: 1460 Number of Model Reactions: 2811 \n",
      "\n",
      "Adding reactions from strain:  CP001396 , model: iBWG_1329\n",
      "Number of Model Genes: 1460 Number of Model Reactions: 2811 \n",
      "\n",
      "Adding reactions from strain:  CP001637 , model: iEcDH1_1363\n",
      "Number of Model Genes: 1460 Number of Model Reactions: 2811 \n",
      "\n",
      "Adding reactions from strain:  AP012030 , model: iECDH1ME8569_1439\n",
      "added: FEENTERtex_copy1\n",
      "Number of Model Genes: 1460 Number of Model Reactions: 2812 \n",
      "\n",
      "Adding reactions from strain:  CU928162 , model: iECED1_1282\n",
      "added: DRBK\n",
      "added: DRIBtpp\n",
      "Number of Model Genes: 1462 Number of Model Reactions: 2814 \n",
      "\n",
      "Adding reactions from strain:  CP000802 , model: iEcHS_1320\n",
      "Number of Model Genes: 1462 Number of Model Reactions: 2814 \n",
      "\n",
      "Adding reactions from strain:  CU928160 , model: iECIAI1_1343\n",
      "added: 3NTD2pp_1\n",
      "added: GLYOX_2\n",
      "added: MTHFD_1\n",
      "added: QULNS_1\n",
      "added: CYTK2_2\n",
      "Number of Model Genes: 1462 Number of Model Reactions: 2819 \n",
      "\n",
      "Adding reactions from strain:  CP002516 , model: iEKO11_1354\n",
      "added: UAG4Ei_copy2\n",
      "added: UAG4Ei_copy1\n",
      "Number of Model Genes: 1463 Number of Model Reactions: 2821 \n",
      "\n",
      "Adding reactions from strain:  AP009240 , model: iECSE_1348\n",
      "Number of Model Genes: 1463 Number of Model Reactions: 2821 \n",
      "\n",
      "Adding reactions from strain:  AP009378 , model: iECSF_1327\n",
      "added: DKFPS2\n",
      "Number of Model Genes: 1464 Number of Model Reactions: 2822 \n",
      "\n",
      "Adding reactions from strain:  CP000948 , model: iECDH10B_1368\n",
      "Number of Model Genes: 1464 Number of Model Reactions: 2822 \n",
      "\n",
      "Adding reactions from strain:  AP009048 , model: iY75_1357\n",
      "Number of Model Genes: 1464 Number of Model Reactions: 2822 \n",
      "\n",
      "Adding reactions from strain:  CP002185 , model: iECW_1372\n",
      "added: RAFH\n",
      "added: SUCR\n",
      "added: RAFHpp\n",
      "added: FFSD\n",
      "added: SUCtpp\n",
      "added: HHDDI\n",
      "added: OPETDC\n",
      "added: OPTCCL\n",
      "added: CMCMSAD\n",
      "added: DHPDO\n",
      "added: CMHMI\n",
      "added: 2HH24DDH1_copy1\n",
      "added: 2HH24DDH1_copy2\n",
      "added: OHEDH\n",
      "added: 2HH24DDH\n",
      "added: 4H2KPILY\n",
      "added: 3HOXPACt2pp\n",
      "added: HPA3MO\n",
      "added: 3HPAOX\n",
      "added: 4HOXPACMOF\n",
      "added: FADRx_copy2\n",
      "added: PENAM\n",
      "added: 4H3NPACAM\n",
      "added: 4PHACTE\n",
      "added: PHACTE\n",
      "added: 35PHACTE\n",
      "added: 34PHACTE\n",
      "added: 2PHACTE\n",
      "added: 3PHACTE\n",
      "added: 3PHACOAOR\n",
      "added: PHACOAOR\n",
      "added: 3NTYROXDApp\n",
      "added: 4HALDD\n",
      "added: 4H3NALDD\n",
      "added: 34DHALDD\n",
      "added: FEENTERtex_copy2\n",
      "added: G3PCT\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  CP002967 , model: iWFL_1372\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  FN554766 , model: iEC042_1314\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  CU928145 , model: iEC55989_1330\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  AP010958 , model: iECO103_1326\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  AP010960 , model: iECO111_1330\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  CP001164 , model: iECH74115_1262\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  AE005174 , model: iZ_1308\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  BA000007 , model: iECs_1301\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  CP001368 , model: iECSP_1301\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  AP010953 , model: iECO26_1355\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  CP001846 , model: iG2583_1286\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  CP000970 , model: iEcSMS35_1347\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  FM180568 , model: iE2348C_1286\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding reactions from strain:  CP000800 , model: iEcE24377_1341\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  FN649414 , model: iETEC_1333\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  CP002729 , model: iUMNK88_1353\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  CP001969 , model: iECOK1_1307\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2859 \n",
      "\n",
      "Adding reactions from strain:  CU928161 , model: iECS88_1305\n",
      "added: CELLBpts\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2860 \n",
      "\n",
      "Adding reactions from strain:  CP000247 , model: iECP_1309\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2860 \n",
      "\n",
      "Adding reactions from strain:  CP001671 , model: iECABU_c1320\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2860 \n",
      "\n",
      "Adding reactions from strain:  AE014075 , model: ic_1306\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2860 \n",
      "\n",
      "Adding reactions from strain:  CU928164 , model: iECIAI39_1322\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2860 \n",
      "\n",
      "Adding reactions from strain:  CP002797 , model: iECNA114_1301\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2860 \n",
      "\n",
      "Adding reactions from strain:  CU928163 , model: iECUMN_1333\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2860 \n",
      "\n",
      "Adding reactions from strain:  CP000243 , model: iUTI89_1310\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2860 \n",
      "\n",
      "Adding reactions from strain:  CP001063 , model: iSbBS512_1146\n",
      "added: GLCtex\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2861 \n",
      "\n",
      "Adding reactions from strain:  CP000036 , model: iSBO_1134\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2861 \n",
      "\n",
      "Adding reactions from strain:  CP000034 , model: iSDY_1059\n",
      "Number of Model Genes: 1476 Number of Model Reactions: 2861 \n",
      "\n",
      "Adding reactions from strain:  CP001383 , model: iSFxv_1172\n",
      "added: ICHORS\n",
      "added: 12PPDStpp_copy2\n",
      "added: MALt5\n",
      "Number of Model Genes: 1477 Number of Model Reactions: 2864 \n",
      "\n",
      "Adding reactions from strain:  AE014073 , model: iS_1188\n",
      "added: SALCHS2FEexs\n",
      "added: DHPTDCs\n",
      "added: SALCHS4FEexs\n",
      "Number of Model Genes: 1478 Number of Model Reactions: 2867 \n",
      "\n",
      "Adding reactions from strain:  AE005674 , model: iSF_1195\n",
      "Number of Model Genes: 1478 Number of Model Reactions: 2867 \n",
      "\n",
      "Adding reactions from strain:  CP000266 , model: iSFV_1184\n",
      "Number of Model Genes: 1478 Number of Model Reactions: 2867 \n",
      "\n",
      "Adding reactions from strain:  CP000038 , model: iSSON_1240\n",
      "Number of Model Genes: 1478 Number of Model Reactions: 2867 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Loop over the strains and add reactions\n",
    "for strain in strain_info['NCBI ID']:\n",
    "    CurrentStrain = strain\n",
    "\n",
    "    # Load model\n",
    "    CurrentModel = strain_info.loc[strain_info['NCBI ID'] == CurrentStrain, 'Model Name'].values[0]\n",
    "    current_model = cobra.io.load_json_model('../data/models/%s.json'%CurrentModel)\n",
    "    print(\"Adding reactions from strain: \", CurrentStrain,', model:', current_model.id)\n",
    "    \n",
    "    # Get ortho_list without reactions already present in model\n",
    "    ortho_list = ortho_to_list(CurrentStrain, orthoIDs_matrix)\n",
    "    \n",
    "    # Add reactions of each model to the EcN model\n",
    "    modelCopy, g_origin_df, r_origin_df = gene_addition(modelCopy, current_model, ortho_list)\n",
    "    \n",
    "    #Utilize the geneIDs matrix to update the GPRs in each of the strain-specific models with the proper gene ID\n",
    "    IDMapping=orthoIDs_matrix[CurrentStrain].to_dict()\n",
    "    IDMappingParsed = {v:k for k,v in IDMapping.items() if v != 'None'}  # v:k to make the CurrentStrain geneIDs keys and EcN gene IDs values\n",
    "    rename_genes(modelCopy,IDMappingParsed)\n",
    "    \n",
    "    # Change the gene ID to EcN gene ID in overview table\n",
    "    IDMapping_switch = {y:x for x,y in IDMapping.items()}\n",
    "    g_origin_df['EcN_gene'] = g_origin_df['EcN_gene'].map(IDMapping_switch)\n",
    "    gene_origin_df = gene_origin_df.append(g_origin_df)\n",
    "                                           \n",
    "    rxn_origin_df = rxn_origin_df.append(r_origin_df)\n",
    "    \n",
    "    # Model evaluation\n",
    "    print ('Number of Model Genes:',len(modelCopy.genes),'Number of Model Reactions:',len(modelCopy.reactions), '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Inspection of draft model & reaction/gene tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Model Genes: 1478 Number of Model Reactions: 2867\n",
      "The rxn_origin dataframe has the correct length: True\n"
     ]
    }
   ],
   "source": [
    "# Final model evaluation\n",
    "print ('Number of Model Genes:',len(modelCopy.genes),'Number of Model Reactions:',len(modelCopy.reactions))\n",
    "\n",
    "# Check the correct length of the rxn_origin_dataframe\n",
    "print('The rxn_origin dataframe has the correct length:', len(modelCopy.reactions) == len(rxn_origin_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>origin</th>\n",
       "      <th>added</th>\n",
       "      <th>notebook</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reaction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALATA_D2</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SHCHD2</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPPPGO</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTHOr</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHORD5</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           origin      added notebook notes\n",
       "reaction                                   \n",
       "ALATA_D2  iML1515  automatic      3.1  none\n",
       "SHCHD2    iML1515  automatic      3.1  none\n",
       "CPPPGO    iML1515  automatic      3.1  none\n",
       "GTHOr     iML1515  automatic      3.1  none\n",
       "DHORD5    iML1515  automatic      3.1  none"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxn_origin_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EcN_gene</th>\n",
       "      <th>origin</th>\n",
       "      <th>added</th>\n",
       "      <th>notebook</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2551</th>\n",
       "      <td>CIW80_06560</td>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0870</th>\n",
       "      <td>CIW80_22360</td>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3368</th>\n",
       "      <td>CIW80_11380</td>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2436</th>\n",
       "      <td>CIW80_06055</td>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3500</th>\n",
       "      <td>CIW80_12105</td>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          EcN_gene   origin      added notebook notes\n",
       "gene                                                 \n",
       "b2551  CIW80_06560  iML1515  automatic      3.1  none\n",
       "b0870  CIW80_22360  iML1515  automatic      3.1  none\n",
       "b3368  CIW80_11380  iML1515  automatic      3.1  none\n",
       "b2436  CIW80_06055  iML1515  automatic      3.1  none\n",
       "b3500  CIW80_12105  iML1515  automatic      3.1  none"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_origin_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4 Save model & reaction/gene tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save as a table\n",
    "rxn_origin_df.to_csv('../tables/rxn_origin.csv')\n",
    "gene_origin_df.to_csv('../tables/gene_origin.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the model\n",
    "cobra.io.json.save_json_model(modelCopy, str('../data/models/%s_draft.json'%EcN_ID), pretty=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
