{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3.3 iDK1463 comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import package needed\n",
    "import cobra\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from cobra.io import load_json_model\n",
    "from cobra import Model, Reaction, Metabolite\n",
    "from cobra.manipulation.modify import rename_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load EcN model\n",
    "EcN_ID = 'CP022686.1'\n",
    "# EcN_model = cobra.io.load_json_model('../data/models/%s_cur_4.7.json'%EcN_ID)\n",
    "EcN_model = cobra.io.load_json_model('../data/models/%s_draft_3.1.json'%EcN_ID)\n",
    "\n",
    "# Load EcN model from Kim et al. (2021)\n",
    "EcN2_model = cobra.io.load_json_model('../data/models_ecn/iDK1463.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ec_iML1515_core_75p37M = 25.102448355609557</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>0.1307</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e</td>\n",
       "      <td>0.1307</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e</td>\n",
       "      <td>0.0006276</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>0.0178</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>0.4032</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>443.4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>4.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.2178</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>0.01735</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e</td>\n",
       "      <td>0.0001757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>271.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ni2_e</td>\n",
       "      <td>EX_ni2_e</td>\n",
       "      <td>0.008108</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>840.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>6.321</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>0.00856</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4crsol_c</td>\n",
       "      <td>DM_4crsol_c</td>\n",
       "      <td>-0.005598</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5drib_c</td>\n",
       "      <td>-0.005648</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_amob_c</td>\n",
       "      <td>-5.02E-05</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mththf_c</td>\n",
       "      <td>DM_mththf_c</td>\n",
       "      <td>-0.01125</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e</td>\n",
       "      <td>-567.9</td>\n",
       "      <td>2</td>\n",
       "      <td>55.96%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-893.8</td>\n",
       "      <td>1</td>\n",
       "      <td>44.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>meoh_e</td>\n",
       "      <td>EX_meoh_e</td>\n",
       "      <td>-5.02E-05</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>-975.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x1a8a238e708>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EcN_model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ec_iDK1463_core_59p80M = 0.8803305910087365</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>0.004582</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e</td>\n",
       "      <td>0.004582</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e</td>\n",
       "      <td>2.201E-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>0.0006242</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>0.007265</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e</td>\n",
       "      <td>0.006874</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>0.1718</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.007637</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>0.0006083</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e</td>\n",
       "      <td>6.162E-06</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>9.507</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ni2_e</td>\n",
       "      <td>EX_ni2_e</td>\n",
       "      <td>0.0002843</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>18.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>0.8487</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>0.2217</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>0.0003002</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4crsol_c</td>\n",
       "      <td>DM_4crsol_c</td>\n",
       "      <td>-0.0001963</td>\n",
       "      <td>7</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5drib_c</td>\n",
       "      <td>-0.0001981</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_amob_c</td>\n",
       "      <td>-1.761E-06</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mththf_c</td>\n",
       "      <td>DM_mththf_c</td>\n",
       "      <td>-0.0003944</td>\n",
       "      <td>5</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e</td>\n",
       "      <td>-1.743</td>\n",
       "      <td>2</td>\n",
       "      <td>14.60%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-20.38</td>\n",
       "      <td>1</td>\n",
       "      <td>85.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-43.62</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-9.838</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>meoh_e</td>\n",
       "      <td>EX_meoh_e</td>\n",
       "      <td>-1.761E-06</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x1a8a66f20c8>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EcN2_model.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Compare genes\n",
    "- Get the old and new locus tag of the CP007799.1 genome\n",
    "- Compare this with the locus tags from CP022686.1\n",
    "- Create a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import packages needed\n",
    "from glob import glob\n",
    "from Bio import Entrez, SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1 Create common gene ID table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>old_locus</th>\n",
       "      <th>ncbigene</th>\n",
       "      <th>ncbiprotein</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>thrL</td>\n",
       "      <td>WP_001386572.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00010</th>\n",
       "      <td>ECOLIN_00010</td>\n",
       "      <td>thrA</td>\n",
       "      <td>WP_001264710.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00015</th>\n",
       "      <td>ECOLIN_00015</td>\n",
       "      <td>thrB</td>\n",
       "      <td>WP_000241663.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00020</th>\n",
       "      <td>ECOLIN_00020</td>\n",
       "      <td>thrC</td>\n",
       "      <td>WP_001704269.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00030</th>\n",
       "      <td>ECOLIN_00025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000738734.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   old_locus ncbigene     ncbiprotein\n",
       "locus                                                \n",
       "ECOLIN_RS00005           NaN     thrL  WP_001386572.1\n",
       "ECOLIN_RS00010  ECOLIN_00010     thrA  WP_001264710.1\n",
       "ECOLIN_RS00015  ECOLIN_00015     thrB  WP_000241663.1\n",
       "ECOLIN_RS00020  ECOLIN_00020     thrC  WP_001704269.1\n",
       "ECOLIN_RS00030  ECOLIN_00025      NaN  WP_000738734.1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get both locus tags from the CP007799.1 genbank file\n",
    "\n",
    "in_file = '../data/models_ecn/CP007799.1.gb'\n",
    "\n",
    "gene_annot_df = pd.DataFrame(columns=('locus', 'old_locus', 'ncbigene', 'ncbiprotein'))\n",
    "gene_annot_df.set_index('locus', inplace=True)\n",
    "\n",
    "handle = open(in_file)\n",
    "\n",
    "records = SeqIO.parse(handle, \"genbank\")\n",
    "for record in records:\n",
    "    for f in record.features:\n",
    "        if f.type=='CDS':\n",
    "\n",
    "            # Get locus tag\n",
    "            if 'locus_tag' in f.qualifiers.keys():\n",
    "                locus = f.qualifiers['locus_tag'][0]\n",
    "            \n",
    "            # get old_locus id\n",
    "            if 'old_locus_tag' in f.qualifiers.keys():\n",
    "                old_locus_id = f.qualifiers['old_locus_tag'][0]\n",
    "                gene_annot_df.loc[locus, 'old_locus'] = old_locus_id\n",
    "                \n",
    "            # get gene id\n",
    "            if 'gene' in f.qualifiers.keys():\n",
    "                gene_id = f.qualifiers['gene'][0]\n",
    "                gene_annot_df.loc[locus, 'ncbigene'] = gene_id\n",
    "            \n",
    "            # get protein_id\n",
    "            if 'protein_id' in f.qualifiers.keys():\n",
    "                prot_id = f.qualifiers['protein_id'][0]\n",
    "                gene_annot_df.loc[locus, 'ncbiprotein'] = prot_id\n",
    "\n",
    "gene_annot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "missing genes of CP0226861. = 22\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>old_locus</th>\n",
       "      <th>ncbigene</th>\n",
       "      <th>ncbiprotein</th>\n",
       "      <th>CP022686.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00010</th>\n",
       "      <td>ECOLIN_00010</td>\n",
       "      <td>thrA</td>\n",
       "      <td>WP_001264710.1</td>\n",
       "      <td>CIW80_17685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00015</th>\n",
       "      <td>ECOLIN_00015</td>\n",
       "      <td>thrB</td>\n",
       "      <td>WP_000241663.1</td>\n",
       "      <td>CIW80_17690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00020</th>\n",
       "      <td>ECOLIN_00020</td>\n",
       "      <td>thrC</td>\n",
       "      <td>WP_001704269.1</td>\n",
       "      <td>CIW80_17695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00030</th>\n",
       "      <td>ECOLIN_00025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WP_000738734.1</td>\n",
       "      <td>CIW80_17700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00035</th>\n",
       "      <td>ECOLIN_00030</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>WP_000906193.1</td>\n",
       "      <td>CIW80_17705</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   old_locus ncbigene     ncbiprotein   CP022686.1\n",
       "locus                                                             \n",
       "ECOLIN_RS00010  ECOLIN_00010     thrA  WP_001264710.1  CIW80_17685\n",
       "ECOLIN_RS00015  ECOLIN_00015     thrB  WP_000241663.1  CIW80_17690\n",
       "ECOLIN_RS00020  ECOLIN_00020     thrC  WP_001704269.1  CIW80_17695\n",
       "ECOLIN_RS00030  ECOLIN_00025      NaN  WP_000738734.1  CIW80_17700\n",
       "ECOLIN_RS00035  ECOLIN_00030     yaaA  WP_000906193.1  CIW80_17705"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a conversion dataframe\n",
    "conv_df = pd.read_csv('../tables/new_old_EcN_geneIDs_matrix.csv')\n",
    "conv_df.rename(columns={'Unnamed: 0': 'old_locus'}, inplace=True)\n",
    "conv_df.set_index('old_locus', inplace=True)\n",
    "\n",
    "# Merge both dataframes\n",
    "convRS_df = pd.merge(gene_annot_df, conv_df, left_on=\"old_locus\", right_index = True)\n",
    "\n",
    "# Create a dictionary\n",
    "EcN_dict = convRS_df['CP022686.1'].to_dict()\n",
    "\n",
    "# Number of genes not present in CP022868.1\n",
    "print(\"missing genes of CP0226861. =\", len(convRS_df[convRS_df['CP022686.1'] == 'None']))\n",
    "\n",
    "# Inspect dataframe\n",
    "convRS_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Differences between models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a dataframe of all genes of both models\n",
    "EcN_gen_df = pd.DataFrame(columns=['EcN_gen'])\n",
    "EcN2_gen_df = pd.DataFrame(columns=['EcN2_gen'])\n",
    "\n",
    "loc= 0\n",
    "for gen in EcN_model.genes:\n",
    "    EcN_gen_df.loc[loc,'EcN_gen'] = gen.id\n",
    "    loc+=1\n",
    "    \n",
    "loc= 0\n",
    "for gen in EcN2_model.genes:\n",
    "    EcN2_gen_df.loc[loc,'EcN2_gen'] = gen.id\n",
    "    loc+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes specific to EcN:  60 Number of genes specific to EcN2:  37\n"
     ]
    }
   ],
   "source": [
    "# Convert the EcN2_gen to CP022686.1 IDs\n",
    "EcN2_gen_df = pd.merge(EcN2_gen_df, convRS_df, left_on=\"EcN2_gen\", right_index = True)\n",
    "\n",
    "# Create two sets\n",
    "EcN_gen = set(EcN_gen_df['EcN_gen'].to_list())\n",
    "EcN2_gen = set(EcN2_gen_df['CP022686.1'].to_list())\n",
    "\n",
    "# Compare\n",
    "EcN_g_only =set(EcN_gen) - set(EcN2_gen)\n",
    "EcN2_g_only = set(EcN2_gen) - set(EcN_gen)\n",
    "\n",
    "print('Number of genes specific to EcN: ', len(EcN_g_only),'Number of genes specific to EcN2: ', len(EcN2_g_only)) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Compare reactions\n",
    "- Compare the reactions\n",
    "- Add missing reactions to the updated EcN_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load EcN model > use the non-curated model, as copies were not removed from their model\n",
    "EcN_ID = 'CP022686.1'\n",
    "EcN_model = cobra.io.load_json_model('../data/models/%s_draft_3.1.json'%EcN_ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 Compare reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "102 219\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'23DOGULNt4pp',\n",
       " '3SLAR',\n",
       " '5DGLCNR',\n",
       " '6D6SFK',\n",
       " '6D6SPA',\n",
       " 'AATA',\n",
       " 'AIRStpp',\n",
       " 'AKGDH2',\n",
       " 'ALAASPabcpp',\n",
       " 'ALAASPtex',\n",
       " 'ALAASPtpp',\n",
       " 'ALAGLNabcpp',\n",
       " 'ALAGLNtex',\n",
       " 'ALAGLNtpp',\n",
       " 'ALAGLUtex',\n",
       " 'ALAGLUtpp',\n",
       " 'ALAGLYabcpp',\n",
       " 'ALAGLYtex',\n",
       " 'ALAGLYtpp',\n",
       " 'ALAHIStpp',\n",
       " 'ALALEUtpp',\n",
       " 'ALATHRtpp',\n",
       " 'ALAt2pp_copy2',\n",
       " 'AMPEP1',\n",
       " 'AMPEP10',\n",
       " 'AMPEP11',\n",
       " 'AMPEP13',\n",
       " 'AMPEP14',\n",
       " 'AMPEP22',\n",
       " 'AMPEP23',\n",
       " 'AMPEP24',\n",
       " 'AMPEP26',\n",
       " 'AMPEP27',\n",
       " 'AMPEP29',\n",
       " 'AMPEP32',\n",
       " 'AMPEP5',\n",
       " 'AMPEP8',\n",
       " 'AOXSr',\n",
       " 'ASNabcpp',\n",
       " 'BDH',\n",
       " 'BIOMASS_Ec_iDK1463_core_59p80M',\n",
       " 'CELLB6PH',\n",
       " 'CYNTt2pp',\n",
       " 'CYSIabcpp',\n",
       " 'D5KGPA',\n",
       " 'DM_lipidA_core_e_p',\n",
       " 'DXYLUDtpp',\n",
       " 'ETHAtpp',\n",
       " 'EX_13ppd_e',\n",
       " 'EX_23dhbzs3_e',\n",
       " 'EX_2hb_e',\n",
       " 'EX_2obut_e',\n",
       " 'EX_2pg_e',\n",
       " 'EX_3hoxpac_e',\n",
       " 'EX_3ntym_e',\n",
       " 'EX_3pg_e',\n",
       " 'EX_4abz_e',\n",
       " 'EX_4hoxpac_e',\n",
       " 'EX_4hphac_e',\n",
       " 'EX_4hthr_e',\n",
       " 'EX_5aptn_e',\n",
       " 'EX_6apa_e',\n",
       " 'EX_AEP_e',\n",
       " 'EX_C02356_e',\n",
       " 'EX_CE0074_e',\n",
       " 'EX_L_alagly_e',\n",
       " 'EX_abt__D_e',\n",
       " 'EX_abt__L_e',\n",
       " 'EX_acglu_e',\n",
       " 'EX_acolipa_e',\n",
       " 'EX_acon_C_e',\n",
       " 'EX_ad_e',\n",
       " 'EX_airs_e',\n",
       " 'EX_ala_L_Thr__L_e',\n",
       " 'EX_ala_L_asp__L_e',\n",
       " 'EX_ala_L_gln__L_e',\n",
       " 'EX_ala_L_glu__L_e',\n",
       " 'EX_ala_L_his__L_e',\n",
       " 'EX_ala_L_leu__L_e',\n",
       " 'EX_arab__D_e',\n",
       " 'EX_asn__D_e',\n",
       " 'EX_asp__D_e',\n",
       " 'EX_bhb_e',\n",
       " 'EX_btd_RR_e',\n",
       " 'EX_bz_e',\n",
       " 'EX_chitob_e',\n",
       " 'EX_citr__L_e',\n",
       " 'EX_colipa20Oag_e',\n",
       " 'EX_colipaOA_e',\n",
       " 'EX_colipa_e',\n",
       " 'EX_colipap_e',\n",
       " 'EX_cspg_c_e',\n",
       " 'EX_cysi__L_e',\n",
       " 'EX_dextrin_e',\n",
       " 'EX_diact_e',\n",
       " 'EX_drib_e',\n",
       " 'EX_dxyl_e',\n",
       " 'EX_fe3dhbzs3_e',\n",
       " 'EX_fe_e',\n",
       " 'EX_feroxB_e',\n",
       " 'EX_feroxBfe_e',\n",
       " 'EX_feroxE_e',\n",
       " 'EX_feroxEfe_e',\n",
       " 'EX_feroxG_e',\n",
       " 'EX_feroxGfe_e',\n",
       " 'EX_frmd_e',\n",
       " 'EX_galam_e',\n",
       " 'EX_glu__D_e',\n",
       " 'EX_glx_e',\n",
       " 'EX_gly_asn__L_e',\n",
       " 'EX_gly_asp__L_e',\n",
       " 'EX_gly_gln__L_e',\n",
       " 'EX_gly_glu__L_e',\n",
       " 'EX_gly_met__L_e',\n",
       " 'EX_gly_pro__L_e',\n",
       " 'EX_glycogen_e',\n",
       " 'EX_guln__L_e',\n",
       " 'EX_ham_e',\n",
       " 'EX_hista_e',\n",
       " 'EX_icit_e',\n",
       " 'EX_kdo2lipid4_e',\n",
       " 'EX_kokdolipidA_e',\n",
       " 'EX_lipa_cold_e',\n",
       " 'EX_lipa_e',\n",
       " 'EX_lys__D_e',\n",
       " 'EX_madg_e',\n",
       " 'EX_mbdg_e',\n",
       " 'EX_met_L_ala__L_e',\n",
       " 'EX_mma_e',\n",
       " 'EX_o16a4colipa_e',\n",
       " 'EX_o6a4colipa_e',\n",
       " 'EX_oaa_e',\n",
       " 'EX_oxa_e',\n",
       " 'EX_pac_e',\n",
       " 'EX_pala_e',\n",
       " 'EX_pectin_e',\n",
       " 'EX_peng_e',\n",
       " 'EX_pep_e',\n",
       " 'EX_raffin_e',\n",
       " 'EX_rbt_e',\n",
       " 'EX_remnant1_e',\n",
       " 'EX_rnam_e',\n",
       " 'EX_salchs2_e',\n",
       " 'EX_salchs2fe_e',\n",
       " 'EX_salchs4_e',\n",
       " 'EX_salchs4fe_e',\n",
       " 'EX_salchsx_e',\n",
       " 'EX_salcn_e',\n",
       " 'EX_srb__L_e',\n",
       " 'EX_tag__D_e',\n",
       " 'EX_tcb_e',\n",
       " 'EX_tet_e',\n",
       " 'EX_tton_e',\n",
       " 'EX_udcpo4_e',\n",
       " 'EX_udcpo5_e',\n",
       " 'EX_val__D_e',\n",
       " 'EX_xylt_e',\n",
       " 'FE3DHBZR',\n",
       " 'GLYASNabcpp',\n",
       " 'GLYASNtex',\n",
       " 'GLYASNtpp',\n",
       " 'GLYASPabcpp',\n",
       " 'GLYASPtex',\n",
       " 'GLYASPtpp',\n",
       " 'GLYGLNtpp',\n",
       " 'GLYGLUtpp',\n",
       " 'GLYLGLNabcpp',\n",
       " 'GLYLGLNtex',\n",
       " 'GLYLGLUabcpp',\n",
       " 'GLYLGLUtex',\n",
       " 'GLYLGLUtpp',\n",
       " 'GLYLMETabcpp',\n",
       " 'GLYLMETtex',\n",
       " 'GLYMETtpp',\n",
       " 'GLYPROabcpp',\n",
       " 'GLYPROtex',\n",
       " 'GLYPROtpp',\n",
       " 'HCINNMt2rpp',\n",
       " 'HPPPNt2rpp',\n",
       " 'INOSTtpp',\n",
       " 'INSTt2',\n",
       " 'INSt2ex',\n",
       " 'LALALGLUabcpp',\n",
       " 'LALALHISabcpp',\n",
       " 'LALALHIStex',\n",
       " 'LALALLEUabcpp',\n",
       " 'LALALLEUtex',\n",
       " 'LALALTHRabcpp',\n",
       " 'LALALTHRtex',\n",
       " 'LCARR',\n",
       " 'L_LACD4',\n",
       " 'MALDDH',\n",
       " 'MAN1PT',\n",
       " 'MANGLYCptspp',\n",
       " 'MDRPD',\n",
       " 'METALAabcpp',\n",
       " 'METALAtex',\n",
       " 'METALAtpp',\n",
       " 'MMETt2pp',\n",
       " 'MTRI',\n",
       " 'MTRK',\n",
       " 'OMPHHX3',\n",
       " 'OPAH',\n",
       " 'PSCLYSt2pp',\n",
       " 'PYDXOR',\n",
       " 'Q23DO',\n",
       " 'SBTD_D2',\n",
       " 'SCYSSL',\n",
       " 'SRB1PR',\n",
       " 'SRBptspp',\n",
       " 'SRBtex',\n",
       " 'SUCCt3pp',\n",
       " 'SUCOAS2',\n",
       " 'TARTRtpp',\n",
       " 'TCBt2ex',\n",
       " 'URCN',\n",
       " 'VALtpp',\n",
       " 'XYLTD_D',\n",
       " 'r0085'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compare the reactions\n",
    "EcN_rxn = []\n",
    "EcN2_rxn = []\n",
    "\n",
    "for rxn in EcN_model.reactions:\n",
    "    EcN_rxn.append(rxn.id)\n",
    "for rxn in EcN2_model.reactions:\n",
    "    EcN2_rxn.append(rxn.id)  \n",
    "\n",
    "EcN_only =set(EcN_rxn) - set(EcN2_rxn)\n",
    "EcN2_only = set(EcN2_rxn) - set(EcN_rxn)\n",
    "\n",
    "print(len(EcN_only), len(EcN2_only))\n",
    "\n",
    "EcN2_only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare the list of reactions to be added\n",
    "rxn_to_add = list(EcN2_only)\n",
    "rxn_to_add.remove('ALAt2pp_copy2') # Original already present\n",
    "rxn_to_add.remove('MALDDH') #To be added in notebook 4.3\n",
    "rxn_to_add.remove('BIOMASS_Ec_iDK1463_core_59p80M') # Added at the end of notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>origin</th>\n",
       "      <th>added</th>\n",
       "      <th>notebook</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reaction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CYTDK2</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XPPT</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HXPRT</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDPK5</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SHK3Dr</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           origin      added  notebook notes\n",
       "reaction                                    \n",
       "CYTDK2    iML1515  automatic       3.1  none\n",
       "XPPT      iML1515  automatic       3.1  none\n",
       "HXPRT     iML1515  automatic       3.1  none\n",
       "NDPK5     iML1515  automatic       3.1  none\n",
       "SHK3Dr    iML1515  automatic       3.1  none"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the table with information on origin of reactions\n",
    "rxn_origin_df = pd.read_csv('../tables/rxn_origin.csv')\n",
    "rxn_origin_df.set_index('reaction', inplace=True)\n",
    "rxn_origin_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Add reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add reaction\n",
    "for reaction in rxn_to_add:\n",
    "    EcN_model.add_reactions([EcN2_model.reactions.get_by_id(reaction)])\n",
    "#     print('added reaction:', reaction)\n",
    "    \n",
    "    # Add reaction to overview table\n",
    "    rxn_origin_df.loc[reaction, 'origin'] = 'iDK1463'\n",
    "    rxn_origin_df.loc[reaction, 'added'] = 'automatic'\n",
    "    rxn_origin_df.loc[reaction, 'notebook'] = '3.3'\n",
    "    rxn_origin_df.loc[reaction, 'notes'] = 'none'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Update Gene ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ECOLIN_RS01800 pepD\n",
      "ECOLIN_RS03795 dtpD\n",
      "ECOLIN_RS23900 dtpC\n",
      "ECOLIN_RS09120 dtpA\n",
      "ECOLIN_RS19975 dtpB\n",
      "ECOLIN_RS22340\n",
      "ECOLIN_RS01525\n",
      "ECOLIN_RS23455\n",
      "ECOLIN_RS13875 ptsH\n",
      "ECOLIN_RS23205\n",
      "ECOLIN_RS23210\n",
      "ECOLIN_RS13880 ptsI\n",
      "ECOLIN_RS23195\n",
      "ECOLIN_RS23200\n",
      "ECOLIN_RS01820 phoE\n",
      "ECOLIN_RS08090 ompN\n",
      "ECOLIN_RS04920 ompF\n",
      "ECOLIN_RS12935 ompC\n",
      "ECOLIN_RS07520 oppC\n",
      "ECOLIN_RS20270 dppB\n",
      "ECOLIN_RS20265 dppC\n",
      "ECOLIN_RS07525 oppD\n",
      "ECOLIN_RS20255 dppF\n",
      "ECOLIN_RS20260 dppD\n",
      "ECOLIN_RS07510 oppA\n",
      "ECOLIN_RS07515 oppB\n",
      "ECOLIN_RS20280 dppA\n",
      "ECOLIN_RS07530 oppF\n",
      "ECOLIN_RS02490 tsx\n",
      "ECOLIN_RS04135 bioF\n",
      "ECOLIN_RS22350\n",
      "ECOLIN_RS20830\n",
      "ECOLIN_RS24935 yjiR\n",
      "ECOLIN_RS08295 ydcR\n",
      "ECOLIN_RS17895 yhaK\n",
      "ECOLIN_RS19680 yhhW\n",
      "ECOLIN_RS23190 sorE\n",
      "ECOLIN_RS27005\n",
      "ECOLIN_RS22345 yihU\n",
      "ECOLIN_RS26995\n",
      "ECOLIN_RS25965 pepQ\n",
      "ECOLIN_RS22185 pepQ\n",
      "ECOLIN_RS09895 ydjJ\n",
      "ECOLIN_RS14410 yphC\n",
      "ECOLIN_RS03815 pxpA\n",
      "ECOLIN_RS03810 pxpC\n",
      "ECOLIN_RS03805 pxpB\n",
      "ECOLIN_RS08745\n",
      "ECOLIN_RS12075 cpsB\n",
      "ECOLIN_RS12000 cpsB\n",
      "ECOLIN_RS01735 yafV\n",
      "ECOLIN_RS23440 sucC\n",
      "ECOLIN_RS23445 sucD\n",
      "ECOLIN_RS08485 yddG\n",
      "ECOLIN_RS14955 ygaH\n",
      "ECOLIN_RS14950 ygaZ\n",
      "ECOLIN_RS20465 yiaN\n",
      "ECOLIN_RS20460 yiaM\n",
      "ECOLIN_RS20470 yiaO\n",
      "ECOLIN_RS13925 ucpA\n",
      "ECOLIN_RS08150 pdxI\n",
      "ECOLIN_RS27000 mtnA\n",
      "ECOLIN_RS23435 lpdA\n",
      "ECOLIN_RS23430 odhB\n",
      "ECOLIN_RS23425\n"
     ]
    }
   ],
   "source": [
    "# Change gene ID\n",
    "for gene in EcN_model.genes:\n",
    "    if 'ECOLIN' in gene.id:\n",
    "        try: # Find out if gene is in annotation overview\n",
    "            gene_annot_df.loc[gene.id, 'ncbigene'] \n",
    "            if pd.isnull(gene_annot_df.loc[gene.id, 'ncbigene']): # Only change name when there is a name in annotation overview\n",
    "                    print(gene.id)\n",
    "                    pass\n",
    "            else:\n",
    "                EcN_model.genes.get_by_id(gene.id).name = gene_annot_df.loc[gene.id, 'ncbigene']\n",
    "                print(gene.id, gene_annot_df.loc[gene.id, 'ncbigene'])\n",
    "        except:\n",
    "            print('Gene not in overview:', gene.id)           "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.5 Rename genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CIW80_15475 Ldh family oxidoreductase\n",
      "CIW80_15215 PTS system mannose/fructose/N-acetylgalactosamine-transporter subunit IIB\n",
      "CIW80_15220 PTS sorbose IIA component\n",
      "CIW80_15205 PTS system mannose/fructose/sorbose family transporter subunit IID\n",
      "CIW80_15210 PTS mannose/fructose/sorbose transporter subunit IIC\n",
      "CIW80_12995 ketose-bisphosphate aldolase\n",
      "CIW80_17025 S-methyl-5-thioribose kinase\n",
      "CIW80_17035 L-fuculose-phosphate aldolase\n",
      "CIW80_15445 sucA, subunit of E1(0) component of 2-oxoglutarate dehydrogenase\n"
     ]
    }
   ],
   "source": [
    "# Add ECOLIN_RS13880 to CIW80_05965 to the dictionary \n",
    "# This gene was not automatically added to the list, but was manually added based on gene name > gene pstI\n",
    "EcN_dict['ECOLIN_RS13880'] = 'CIW80_05965'\n",
    "\n",
    "# rename genes\n",
    "rename_genes(EcN_model, EcN_dict)\n",
    "\n",
    "# Get Biocyc overview\n",
    "EcN_biocyc = pd.read_csv('../tables/EcN_genes_Biocyc.txt', sep='\\t', header=0, names = ['Gene', 'Description', 'Accession1', 'Accession2'])\n",
    "EcN_biocyc.set_index('Accession2', inplace = True)\n",
    "EcN_biocyc\n",
    "\n",
    "# Find out genes that still have an old gene_id\n",
    "for gene in EcN_model.genes:\n",
    "    if 'ECOLIN_RS' in gene.name:\n",
    "        if 'CIW80_RS' in EcN_biocyc.loc[gene.id, 'Gene']:\n",
    "            EcN_model.genes.get_by_id(gene.id).name = EcN_biocyc.loc[gene.id, 'Description']\n",
    "        else:\n",
    "            EcN_model.genes.get_by_id(gene.id).name = EcN_biocyc.loc[gene.id, 'Gene'] + ', ' + EcN_biocyc.loc[gene.id, 'Description']\n",
    "        print(gene.id, gene.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find out genes that still have an old gene_id\n",
    "for gene in EcN_model.genes:\n",
    "    if 'ECOLIN_RS' in gene.id:\n",
    "        print(gene.id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4 Update gene annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>asap</th>\n",
       "      <th>ecogene</th>\n",
       "      <th>ncbigene</th>\n",
       "      <th>ncbigi</th>\n",
       "      <th>refseq_locus_tag</th>\n",
       "      <th>refseq_name</th>\n",
       "      <th>refseq_synonym</th>\n",
       "      <th>sbo</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>refseq_old_locus_tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS12160</th>\n",
       "      <td>ABE-0006840</td>\n",
       "      <td>EG11701</td>\n",
       "      <td>946597</td>\n",
       "      <td>90111379</td>\n",
       "      <td>b2066</td>\n",
       "      <td>udk</td>\n",
       "      <td>[ECK2060, JW2051]</td>\n",
       "      <td>SBO:0000243</td>\n",
       "      <td>P0A8F4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS01805</th>\n",
       "      <td>ABE-0000814</td>\n",
       "      <td>EG10414</td>\n",
       "      <td>944817</td>\n",
       "      <td>16128224</td>\n",
       "      <td>b0238</td>\n",
       "      <td>gpt</td>\n",
       "      <td>[gpp, ECK0239, gxu, JW0228]</td>\n",
       "      <td>SBO:0000243</td>\n",
       "      <td>P0A9M5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS00695</th>\n",
       "      <td>ABE-0000440</td>\n",
       "      <td>EG20098</td>\n",
       "      <td>946624</td>\n",
       "      <td>90111088</td>\n",
       "      <td>b0125</td>\n",
       "      <td>hpt</td>\n",
       "      <td>[ECK0124, JW5009]</td>\n",
       "      <td>SBO:0000243</td>\n",
       "      <td>P0A9M2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS14305</th>\n",
       "      <td>ABE-0008291</td>\n",
       "      <td>EG10650</td>\n",
       "      <td>945611</td>\n",
       "      <td>16130443</td>\n",
       "      <td>b2518</td>\n",
       "      <td>ndk</td>\n",
       "      <td>[JW2502, ECK2514]</td>\n",
       "      <td>SBO:0000243</td>\n",
       "      <td>P0A763</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECOLIN_RS02815</th>\n",
       "      <td>ABE-0001645</td>\n",
       "      <td>EG10032</td>\n",
       "      <td>945097</td>\n",
       "      <td>16128458</td>\n",
       "      <td>b0474</td>\n",
       "      <td>adk</td>\n",
       "      <td>[plsA, JW0463, ECK0468, dnaW]</td>\n",
       "      <td>SBO:0000243</td>\n",
       "      <td>P69441</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       asap  ecogene ncbigene    ncbigi refseq_locus_tag  \\\n",
       "gene                                                                       \n",
       "ECOLIN_RS12160  ABE-0006840  EG11701   946597  90111379            b2066   \n",
       "ECOLIN_RS01805  ABE-0000814  EG10414   944817  16128224            b0238   \n",
       "ECOLIN_RS00695  ABE-0000440  EG20098   946624  90111088            b0125   \n",
       "ECOLIN_RS14305  ABE-0008291  EG10650   945611  16130443            b2518   \n",
       "ECOLIN_RS02815  ABE-0001645  EG10032   945097  16128458            b0474   \n",
       "\n",
       "               refseq_name                 refseq_synonym          sbo  \\\n",
       "gene                                                                     \n",
       "ECOLIN_RS12160         udk              [ECK2060, JW2051]  SBO:0000243   \n",
       "ECOLIN_RS01805         gpt    [gpp, ECK0239, gxu, JW0228]  SBO:0000243   \n",
       "ECOLIN_RS00695         hpt              [ECK0124, JW5009]  SBO:0000243   \n",
       "ECOLIN_RS14305         ndk              [JW2502, ECK2514]  SBO:0000243   \n",
       "ECOLIN_RS02815         adk  [plsA, JW0463, ECK0468, dnaW]  SBO:0000243   \n",
       "\n",
       "               uniprot refseq_old_locus_tag  \n",
       "gene                                         \n",
       "ECOLIN_RS12160  P0A8F4                  NaN  \n",
       "ECOLIN_RS01805  P0A9M5                  NaN  \n",
       "ECOLIN_RS00695  P0A9M2                  NaN  \n",
       "ECOLIN_RS14305  P0A763                  NaN  \n",
       "ECOLIN_RS02815  P69441                  NaN  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reload EcN model from Kim et al. (2021) > for some reason renaming also affects the EcN2 model\n",
    "EcN2_model = cobra.io.load_json_model('../data/models_ecn/iDK1463.json')\n",
    "\n",
    "# Create dataframe\n",
    "gene_annot_EcN2 = pd.DataFrame(columns=['gene', 'asap', 'ecogene', 'ncbigene', 'ncbigi', 'refseq_locus_tag', 'refseq_name', 'refseq_synonym', 'sbo', 'uniprot'])\n",
    "gene_annot_EcN2.set_index('gene', inplace= True)\n",
    "\n",
    "# Add all annotation info of iDK1463 into dataframe\n",
    "for gene in EcN2_model.genes:\n",
    "    for annot in EcN2_model.genes.get_by_id(gene.id).annotation:\n",
    "        gene_annot_EcN2.loc[gene.id, annot] = EcN2_model.genes.get_by_id(gene.id).annotation[annot]\n",
    "    \n",
    "gene_annot_EcN2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge dataframes\n",
    "\n",
    "gene_annot_EcN2 = pd.merge(gene_annot_EcN2, convRS_df['CP022686.1'],left_index=True, right_index=True, how='left')\n",
    "gene_annot_EcN2.set_index('CP022686.1', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update all missing annotations that are present in the iDK1463 model\n",
    "for gene in EcN_model.genes:\n",
    "        for annot in EcN_model.genes.get_by_id(gene.id).annotation:\n",
    "#             print(EcN_model.genes.get_by_id(gene.id).annotation[annot])\n",
    "            if EcN_model.genes.get_by_id(gene.id).annotation[annot] == '':\n",
    "                EcN_model.genes.get_by_id(gene.id).annotation[annot] = gene_annot_EcN2.loc[gene.id, annot]\n",
    "                print(gene, annot, EcN_model.genes.get_by_id(gene.id).annotation[annot])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No annotations of the iDK1463 model were missing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.5 Update subsystem reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "for rxn in rxn_to_add:\n",
    "    \n",
    "# Based on ID\n",
    "    if 'EX' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Extracellular exchange'\n",
    "    \n",
    "    if 'abcpp' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Transport, Inner Membrane'\n",
    "\n",
    "    if 'tpp' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Transport, Inner Membrane'\n",
    "        \n",
    "    if 't2pp' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Transport, Inner Membrane'         \n",
    "        \n",
    "    if 't2rpp' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Transport, Inner Membrane'        \n",
    "        \n",
    "    if 't3pp' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Transport, Inner Membrane'      \n",
    "        \n",
    "    if 't4pp' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Transport, Inner Membrane'       \n",
    "\n",
    "    if 'tex' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Transport, Outer Membrane'\n",
    "        \n",
    "    if 't2ex' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Transport, Outer Membrane'       \n",
    "         \n",
    "# Based on name\n",
    "    if 'PEP:Pyr' in EcN_model.reactions.get_by_id(rxn).name: # Correct the subsystem of these reactions to Inner Membrane\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Transport, Inner Membrane'\n",
    "        \n",
    "    if 'AMPEP' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions to Inner Membrane\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Amino Acid Metabolism'\n",
    "\n",
    "    if 'ACOAD' in EcN_model.reactions.get_by_id(rxn).id: # Correct the subsystem of these reactions to Inner Membrane\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = 'Membrane Lipid Metabolism'        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Subsystem</th>\n",
       "      <th>Reaction Formula</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CELLBpts_1</th>\n",
       "      <td>Cellobiose transport via PEP:Pyr PTS</td>\n",
       "      <td></td>\n",
       "      <td>cellb_e + pep_c --&gt; 6pgg_c + pyr_p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DMALRED</th>\n",
       "      <td>(S)-Malate:(acceptor) oxidoreductase</td>\n",
       "      <td></td>\n",
       "      <td>fad_c + mal__L_c --&gt; fadh2_c + oaa_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHACOAH</th>\n",
       "      <td>2,3-dehydroadipyl-CoA hydratase</td>\n",
       "      <td></td>\n",
       "      <td>23dhacoa_c + h2o_c --&gt; 3hadpcoa_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMACTD</th>\n",
       "      <td>Imidazole acetaldeyde dehydrogenase</td>\n",
       "      <td></td>\n",
       "      <td>h2o_c + im4act_c + nad_c --&gt; 2.0 h_c + im4ac_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALDD2x</th>\n",
       "      <td>Aldehyde dehydrogenase (acetaldehyde, NAD)</td>\n",
       "      <td></td>\n",
       "      <td>acald_c + h2o_c + nad_c --&gt; ac_c + 2.0 h_c + n...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  Name Subsystem  \\\n",
       "ID                                                                 \n",
       "CELLBpts_1        Cellobiose transport via PEP:Pyr PTS             \n",
       "DMALRED           (S)-Malate:(acceptor) oxidoreductase             \n",
       "DHACOAH                2,3-dehydroadipyl-CoA hydratase             \n",
       "IMACTD             Imidazole acetaldeyde dehydrogenase             \n",
       "ALDD2x      Aldehyde dehydrogenase (acetaldehyde, NAD)             \n",
       "\n",
       "                                             Reaction Formula  \n",
       "ID                                                             \n",
       "CELLBpts_1                 cellb_e + pep_c --> 6pgg_c + pyr_p  \n",
       "DMALRED                  fad_c + mal__L_c --> fadh2_c + oaa_c  \n",
       "DHACOAH                     23dhacoa_c + h2o_c --> 3hadpcoa_c  \n",
       "IMACTD      h2o_c + im4act_c + nad_c --> 2.0 h_c + im4ac_c...  \n",
       "ALDD2x      acald_c + h2o_c + nad_c --> ac_c + 2.0 h_c + n...  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identify reactions still missing subsystem\n",
    "df_reactions = pd.DataFrame(columns=['ID', 'Name', 'Subsystem'])\n",
    "df_reactions.set_index('ID', inplace=True)\n",
    "\n",
    "for rxn in EcN_model.reactions:\n",
    "    bigg_id = rxn.id\n",
    "    df_reactions.loc[bigg_id, 'Name'] = rxn.name\n",
    "    df_reactions.loc[bigg_id, 'Reaction Formula'] = rxn.reaction\n",
    "    df_reactions.loc[bigg_id, 'Subsystem'] = rxn.subsystem\n",
    "\n",
    "sub_missing_df = df_reactions[df_reactions['Subsystem'] == '']\n",
    "sub_missing_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "L_LACD4 not in panecoli\n",
      "AOXSr not in panecoli\n",
      "D5KGPA not in panecoli\n",
      "AATA not in panecoli\n",
      "Q23DO not in panecoli\n",
      "SRB1PR not in panecoli\n",
      "MTRK not in panecoli\n",
      "MDRPD not in panecoli\n",
      "SBTD_D2 not in panecoli\n",
      "OPAH not in panecoli\n",
      "CELLB6PH not in panecoli\n",
      "MAN1PT not in panecoli\n",
      "r0085 not in panecoli\n",
      "XYLTD_D not in panecoli\n",
      "SUCOAS2 not in panecoli\n",
      "BDH not in panecoli\n",
      "PYDXOR not in panecoli\n",
      "MTRI not in panecoli\n",
      "AKGDH2 not in panecoli\n"
     ]
    }
   ],
   "source": [
    "# Add all subsystems present in panecoli model\n",
    "panecoli = cobra.io.load_json_model('../data/models/panecoli_universal.json')\n",
    "\n",
    "for rxn in sub_missing_df.index:\n",
    "    try:\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = panecoli.reactions.get_by_id(rxn).subsystem\n",
    "    except:\n",
    "        print(rxn, 'not in panecoli')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manually update names based on BIGG database\n",
    "EcN_model.reactions.get_by_id('D5KGPA').subsystem = 'Inositol Metabolism'\n",
    "EcN_model.reactions.get_by_id('MDRPD').subsystem = 'Arginine and Proline Metabolism'\n",
    "EcN_model.reactions.get_by_id('OPAH').subsystem = 'Glutathione Metabolism'\n",
    "EcN_model.reactions.get_by_id('MTRK').subsystem = 'Methionine Salvage'\n",
    "EcN_model.reactions.get_by_id('Q23DO').subsystem = 'Alternate Carbon Metabolism'\n",
    "EcN_model.reactions.get_by_id('AATA').subsystem = 'Threonine and Lysine Metabolism'\n",
    "EcN_model.reactions.get_by_id('MAN1PT').subsystem = 'Alternate Carbon Metabolism'\n",
    "EcN_model.reactions.get_by_id('L_LACD4').subsystem = 'Pyruvate Metabolism'\n",
    "EcN_model.reactions.get_by_id('MTRI').subsystem = 'Methionine Salvage'\n",
    "EcN_model.reactions.get_by_id('SRB1PR').subsystem = 'Alternate Carbon Metabolism'\n",
    "EcN_model.reactions.get_by_id('BDH').subsystem = 'Alternate Carbon Metabolism'\n",
    "EcN_model.reactions.get_by_id('SUCOAS2').subsystem = 'Citric Acid Cycle'\n",
    "EcN_model.reactions.get_by_id('AOXSr').subsystem = 'Cofactor and Prosthetic Group Biosynthesis'\n",
    "EcN_model.reactions.get_by_id('AKGDH2').subsystem = 'Citric Acid Cycle'\n",
    "EcN_model.reactions.get_by_id('PYDXOR').subsystem = 'Cofactor and Prosthetic Group Biosynthesis'\n",
    "EcN_model.reactions.get_by_id('CELLB6PH').subsystem = 'Alternate Carbon Metabolism'\n",
    "EcN_model.reactions.get_by_id('SBTD_D2').subsystem = 'Alternate Carbon Metabolism'\n",
    "EcN_model.reactions.get_by_id('XYLTD_D').subsystem = 'Pentose Phosphate Pathway'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.6 Update biomass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update Biomass\n",
    "EcN2_model.reactions.BIOMASS_Ec_iDK1463_core_59p80M\n",
    "\n",
    "# Add and change reaction\n",
    "EcN_model.add_reactions([EcN2_model.reactions.BIOMASS_Ec_iDK1463_core_59p80M])\n",
    "EcN_model.reactions.BIOMASS_Ec_iDK1463_core_59p80M.subsystem = 'Biomass and maintenance functions'\n",
    "EcN_model.reactions.BIOMASS_Ec_iDK1463_core_59p80M.name = 'E. coli biomass objective function (iHM1533) - core - with 59.80 GAM estimate'\n",
    "EcN_model.reactions.BIOMASS_Ec_iDK1463_core_59p80M.id = 'BIOMASS_EcN_iHM1533_core_59p80M'\n",
    "\n",
    "# Make model objective\n",
    "EcN_model.objective = EcN_model.reactions.BIOMASS_EcN_iHM1533_core_59p80M"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.7 Model adjustments of iDK1463"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Delete several reactions\n",
    "EcN_model.remove_reactions(['ACACtex', 'ABUTtex', 'BUTt2rpp',  'ETHAtex', 'URIC', 'XANtex'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Save model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_EcN_iHM1534_core_59p80M = 28.77687793819021</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e</td>\n",
       "      <td>0.1498</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e</td>\n",
       "      <td>0.1498</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e</td>\n",
       "      <td>0.0007194</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe2_e</td>\n",
       "      <td>EX_fe2_e</td>\n",
       "      <td>0.4622</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>364.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e</td>\n",
       "      <td>5.617</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e</td>\n",
       "      <td>0.2496</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>0.01988</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e</td>\n",
       "      <td>0.0002014</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>310.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ni2_e</td>\n",
       "      <td>EX_ni2_e</td>\n",
       "      <td>0.009295</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>813.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e</td>\n",
       "      <td>7.247</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e</td>\n",
       "      <td>0.009813</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4crsol_c</td>\n",
       "      <td>DM_4crsol_c</td>\n",
       "      <td>-0.006417</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5drib_c</td>\n",
       "      <td>-0.006475</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_amob_c</td>\n",
       "      <td>-5.755E-05</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mththf_c</td>\n",
       "      <td>DM_mththf_c</td>\n",
       "      <td>-0.01289</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e</td>\n",
       "      <td>-501.8</td>\n",
       "      <td>2</td>\n",
       "      <td>53.41%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-875.3</td>\n",
       "      <td>1</td>\n",
       "      <td>46.58%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>meoh_e</td>\n",
       "      <td>EX_meoh_e</td>\n",
       "      <td>-5.755E-05</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>-972.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x1a8a7bf8848>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EcN_model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save as a table\n",
    "rxn_origin_df.to_csv('../tables/rxn_origin.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the model\n",
    "cobra.io.json.save_json_model(EcN_model, str('../data/models/%s_draft_3.3.json'%EcN_ID), pretty=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
