{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4.6 Secondary metabolite pathway addition\n",
    "\n",
    "EcN produces six different iron-chelating siderophores;\n",
    "- Aerobactin _(iuc/aer)_\n",
    "- Enterobactin _(ent)_\n",
    "- Salmochelin _(iro)_\n",
    "- Yersiniabactin _(ybt)_\n",
    "- Hemin-dependent iron acquisition system _(chu)_\n",
    "- Citrate-dependent iron acquisition system _(cit)_\n",
    "\n",
    "The latter two are already present in the model. The first four were added here. Colibactin _(clb/pks)_ was additionally added to the model.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import package needed\n",
    "import cobra\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "from cobra.io import load_json_model\n",
    "from glob import glob\n",
    "from cobra import Model, Reaction, Metabolite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load EcN model\n",
    "EcN_ID = 'CP022686.1'\n",
    "EcN_model = cobra.io.load_json_model('../data/models/%s_cur_4.5.json'%EcN_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the tables with information on origin of genes & reactions\n",
    "gene_origin_df = pd.read_csv('../tables/gene_origin.csv')\n",
    "gene_origin_df.set_index('gene', inplace=True)\n",
    "\n",
    "rxn_origin_df = pd.read_csv('../tables/rxn_origin.csv')\n",
    "rxn_origin_df.set_index('reaction', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rxn_origin_df = rxn_origin_df[rxn_origin_df['added'] != 'manual']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Add pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First remove the ENTCS reaction, which is a lumped reaction of the enterobactin pathway\n",
    "EcN_model.remove_reactions(['ENTCS'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>Sec_met_path</th>\n",
       "      <th>Name</th>\n",
       "      <th>Reaction Formula</th>\n",
       "      <th>Gene Rules</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Subsystem</th>\n",
       "      <th>Compartment</th>\n",
       "      <th>LB</th>\n",
       "      <th>UB</th>\n",
       "      <th>RHEA</th>\n",
       "      <th>BioCyc</th>\n",
       "      <th>MetaNetX (MNX) Equation</th>\n",
       "      <th>KEGG Reaction</th>\n",
       "      <th>SEED Reaction</th>\n",
       "      <th>EC Number</th>\n",
       "      <th>Reactome Reaction</th>\n",
       "      <th>Reference PMID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ICHORS</th>\n",
       "      <td>Draft model</td>\n",
       "      <td>Enterobactin</td>\n",
       "      <td>Isochorismate synthase</td>\n",
       "      <td>chor_c &lt;=&gt; ichor_c</td>\n",
       "      <td>CIW80_05275 or CIW80_20975</td>\n",
       "      <td>menF or entC</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c</td>\n",
       "      <td>-1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:ISOCHORSYN-RXN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ICHORT</th>\n",
       "      <td>Draft model</td>\n",
       "      <td>Enterobactin</td>\n",
       "      <td>Isochorismatase</td>\n",
       "      <td>h2o_c + ichor_c --&gt; 23ddhb_c + pyr_c</td>\n",
       "      <td>CIW80_20985</td>\n",
       "      <td>entB</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:ISOCHORMAT-RXN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHBD</th>\n",
       "      <td>Draft model</td>\n",
       "      <td>Enterobactin</td>\n",
       "      <td>2,3-dihydro-2,3-dihydroxybenzoate dehydrogenase</td>\n",
       "      <td>23ddhb_c + nad_c --&gt; 23dhb_c + h_c + nadh_c</td>\n",
       "      <td>CIW80_20990</td>\n",
       "      <td>entA</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:DHBDEHYD-RXN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ArCPSentB</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>Enterobactin</td>\n",
       "      <td>Phosphopantetheinyl transferase (enterobactin B)</td>\n",
       "      <td>apoArCPentB_c + coa_c --&gt; holoArCPentB_c + h_c...</td>\n",
       "      <td>CIW80_20985 and CIW80_20915</td>\n",
       "      <td>entB and entD</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:ENTDB-RXN</td>\n",
       "      <td>MNXR115861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.7.8.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCPSentF</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>Enterobactin</td>\n",
       "      <td>Phosphopantetheinyl transferase (enterobactin F)</td>\n",
       "      <td>apoPCPentF_c + coa_c --&gt; holoPCPentF_c + h_c +...</td>\n",
       "      <td>CIW80_20935 and CIW80_20915</td>\n",
       "      <td>entF and entD</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:RXN-15889</td>\n",
       "      <td>MNXR119687</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.7.8.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Status  Sec_met_path  \\\n",
       "ID                                      \n",
       "ICHORS      Draft model  Enterobactin   \n",
       "ICHORT      Draft model  Enterobactin   \n",
       "DHBD        Draft model  Enterobactin   \n",
       "ArCPSentB  New Reaction  Enterobactin   \n",
       "PCPSentF   New Reaction  Enterobactin   \n",
       "\n",
       "                                                       Name  \\\n",
       "ID                                                            \n",
       "ICHORS                               Isochorismate synthase   \n",
       "ICHORT                                      Isochorismatase   \n",
       "DHBD        2,3-dihydro-2,3-dihydroxybenzoate dehydrogenase   \n",
       "ArCPSentB  Phosphopantetheinyl transferase (enterobactin B)   \n",
       "PCPSentF   Phosphopantetheinyl transferase (enterobactin F)   \n",
       "\n",
       "                                            Reaction Formula  \\\n",
       "ID                                                             \n",
       "ICHORS                                    chor_c <=> ichor_c   \n",
       "ICHORT                  h2o_c + ichor_c --> 23ddhb_c + pyr_c   \n",
       "DHBD             23ddhb_c + nad_c --> 23dhb_c + h_c + nadh_c   \n",
       "ArCPSentB  apoArCPentB_c + coa_c --> holoArCPentB_c + h_c...   \n",
       "PCPSentF   apoPCPentF_c + coa_c --> holoPCPentF_c + h_c +...   \n",
       "\n",
       "                            Gene Rules      Gene Name  \\\n",
       "ID                                                      \n",
       "ICHORS      CIW80_05275 or CIW80_20975   menF or entC   \n",
       "ICHORT                     CIW80_20985           entB   \n",
       "DHBD                       CIW80_20990           entA   \n",
       "ArCPSentB  CIW80_20985 and CIW80_20915  entB and entD   \n",
       "PCPSentF   CIW80_20935 and CIW80_20915  entF and entD   \n",
       "\n",
       "                                   Subsystem Compartment      LB      UB RHEA  \\\n",
       "ID                                                                              \n",
       "ICHORS     Secondary metabolite biosynthesis           c -1000.0  1000.0  NaN   \n",
       "ICHORT     Secondary metabolite biosynthesis           c     0.0  1000.0  NaN   \n",
       "DHBD       Secondary metabolite biosynthesis           c     0.0  1000.0  NaN   \n",
       "ArCPSentB  Secondary metabolite biosynthesis           c     0.0  1000.0  NaN   \n",
       "PCPSentF   Secondary metabolite biosynthesis           c     0.0  1000.0  NaN   \n",
       "\n",
       "                        BioCyc MetaNetX (MNX) Equation KEGG Reaction  \\\n",
       "ID                                                                     \n",
       "ICHORS     META:ISOCHORSYN-RXN                     NaN           NaN   \n",
       "ICHORT     META:ISOCHORMAT-RXN                     NaN           NaN   \n",
       "DHBD         META:DHBDEHYD-RXN                     NaN           NaN   \n",
       "ArCPSentB       META:ENTDB-RXN              MNXR115861           NaN   \n",
       "PCPSentF        META:RXN-15889              MNXR119687           NaN   \n",
       "\n",
       "          SEED Reaction EC Number Reactome Reaction  Reference PMID  \n",
       "ID                                                                   \n",
       "ICHORS              NaN       NaN               NaN             NaN  \n",
       "ICHORT              NaN       NaN               NaN             NaN  \n",
       "DHBD                NaN       NaN               NaN             NaN  \n",
       "ArCPSentB           NaN   2.7.8.7               NaN             NaN  \n",
       "PCPSentF            NaN   2.7.8.7               NaN             NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the table with information on reactions to add\n",
    "reaction_df = pd.read_csv('../tables/sec_met_paths_reactions.csv', sep=';') #adjust delimiter if needed\n",
    "reaction_df['ID'] = reaction_df['ID'].str.strip() #Remove white spaces\n",
    "reaction_df.set_index('ID', inplace=True)\n",
    "reaction_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>Sec_met_path</th>\n",
       "      <th>Name</th>\n",
       "      <th>Reaction Formula</th>\n",
       "      <th>Gene Rules</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Subsystem</th>\n",
       "      <th>Compartment</th>\n",
       "      <th>LB</th>\n",
       "      <th>UB</th>\n",
       "      <th>rhea</th>\n",
       "      <th>biocyc</th>\n",
       "      <th>metanetx.reaction</th>\n",
       "      <th>kegg.reaction</th>\n",
       "      <th>seed.reaction</th>\n",
       "      <th>ec-code</th>\n",
       "      <th>reactome</th>\n",
       "      <th>pubmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ICHORS</th>\n",
       "      <td>Draft model</td>\n",
       "      <td>Enterobactin</td>\n",
       "      <td>Isochorismate synthase</td>\n",
       "      <td>chor_c &lt;=&gt; ichor_c</td>\n",
       "      <td>CIW80_05275 or CIW80_20975</td>\n",
       "      <td>menF or entC</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c</td>\n",
       "      <td>-1000.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:ISOCHORSYN-RXN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ICHORT</th>\n",
       "      <td>Draft model</td>\n",
       "      <td>Enterobactin</td>\n",
       "      <td>Isochorismatase</td>\n",
       "      <td>h2o_c + ichor_c --&gt; 23ddhb_c + pyr_c</td>\n",
       "      <td>CIW80_20985</td>\n",
       "      <td>entB</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:ISOCHORMAT-RXN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHBD</th>\n",
       "      <td>Draft model</td>\n",
       "      <td>Enterobactin</td>\n",
       "      <td>2,3-dihydro-2,3-dihydroxybenzoate dehydrogenase</td>\n",
       "      <td>23ddhb_c + nad_c --&gt; 23dhb_c + h_c + nadh_c</td>\n",
       "      <td>CIW80_20990</td>\n",
       "      <td>entA</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:DHBDEHYD-RXN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ArCPSentB</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>Enterobactin</td>\n",
       "      <td>Phosphopantetheinyl transferase (enterobactin B)</td>\n",
       "      <td>apoArCPentB_c + coa_c --&gt; holoArCPentB_c + h_c...</td>\n",
       "      <td>CIW80_20985 and CIW80_20915</td>\n",
       "      <td>entB and entD</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:ENTDB-RXN</td>\n",
       "      <td>MNXR115861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.7.8.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCPSentF</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>Enterobactin</td>\n",
       "      <td>Phosphopantetheinyl transferase (enterobactin F)</td>\n",
       "      <td>apoPCPentF_c + coa_c --&gt; holoPCPentF_c + h_c +...</td>\n",
       "      <td>CIW80_20935 and CIW80_20915</td>\n",
       "      <td>entF and entD</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:RXN-15889</td>\n",
       "      <td>MNXR119687</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.7.8.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLBTinter1</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>Colibactin</td>\n",
       "      <td>Colinactin intermediate reaction 1</td>\n",
       "      <td>preclbInter_p --&gt; preclb16a_p + 4.0 h2o_p + 2....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>p</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RXN-21153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLBTinter2</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>Colibactin</td>\n",
       "      <td>Colinactin intermediate reaction 2</td>\n",
       "      <td>preclb16a_p --&gt; preclb16b_p + h_p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>p</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RXN-21102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLBTinter3</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>Colibactin</td>\n",
       "      <td>Colinactin intermediate reaction 3</td>\n",
       "      <td>preclb16b_p + h2o_p + h_p --&gt; clbtn_p + nh4_p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>p</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RXN-21101</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLBTNtex</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>Colibactin</td>\n",
       "      <td>Colibactin secretion (to extracellular)</td>\n",
       "      <td>clbtn_p + h_p --&gt; clbtn_e + h_c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Secondary metabolite biosynthesis</td>\n",
       "      <td>c and p nad e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_clbtn_e</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>Colibactin</td>\n",
       "      <td>Colibactin exchange</td>\n",
       "      <td>clbtn_e --&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Extracellular exchange</td>\n",
       "      <td>e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>177 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Status  Sec_met_path  \\\n",
       "ID                                       \n",
       "ICHORS       Draft model  Enterobactin   \n",
       "ICHORT       Draft model  Enterobactin   \n",
       "DHBD         Draft model  Enterobactin   \n",
       "ArCPSentB   New Reaction  Enterobactin   \n",
       "PCPSentF    New Reaction  Enterobactin   \n",
       "...                  ...           ...   \n",
       "CLBTinter1  New Reaction    Colibactin   \n",
       "CLBTinter2  New Reaction    Colibactin   \n",
       "CLBTinter3  New Reaction    Colibactin   \n",
       "CLBTNtex    New Reaction    Colibactin   \n",
       "EX_clbtn_e  New Reaction    Colibactin   \n",
       "\n",
       "                                                        Name  \\\n",
       "ID                                                             \n",
       "ICHORS                                Isochorismate synthase   \n",
       "ICHORT                                       Isochorismatase   \n",
       "DHBD         2,3-dihydro-2,3-dihydroxybenzoate dehydrogenase   \n",
       "ArCPSentB   Phosphopantetheinyl transferase (enterobactin B)   \n",
       "PCPSentF    Phosphopantetheinyl transferase (enterobactin F)   \n",
       "...                                                      ...   \n",
       "CLBTinter1                Colinactin intermediate reaction 1   \n",
       "CLBTinter2                Colinactin intermediate reaction 2   \n",
       "CLBTinter3                Colinactin intermediate reaction 3   \n",
       "CLBTNtex             Colibactin secretion (to extracellular)   \n",
       "EX_clbtn_e                               Colibactin exchange   \n",
       "\n",
       "                                             Reaction Formula  \\\n",
       "ID                                                              \n",
       "ICHORS                                     chor_c <=> ichor_c   \n",
       "ICHORT                   h2o_c + ichor_c --> 23ddhb_c + pyr_c   \n",
       "DHBD              23ddhb_c + nad_c --> 23dhb_c + h_c + nadh_c   \n",
       "ArCPSentB   apoArCPentB_c + coa_c --> holoArCPentB_c + h_c...   \n",
       "PCPSentF    apoPCPentF_c + coa_c --> holoPCPentF_c + h_c +...   \n",
       "...                                                       ...   \n",
       "CLBTinter1  preclbInter_p --> preclb16a_p + 4.0 h2o_p + 2....   \n",
       "CLBTinter2                  preclb16a_p --> preclb16b_p + h_p   \n",
       "CLBTinter3      preclb16b_p + h2o_p + h_p --> clbtn_p + nh4_p   \n",
       "CLBTNtex                     clbtn_p + h_p --> clbtn_e + h_c    \n",
       "EX_clbtn_e                                        clbtn_e -->   \n",
       "\n",
       "                             Gene Rules      Gene Name  \\\n",
       "ID                                                       \n",
       "ICHORS       CIW80_05275 or CIW80_20975   menF or entC   \n",
       "ICHORT                      CIW80_20985           entB   \n",
       "DHBD                        CIW80_20990           entA   \n",
       "ArCPSentB   CIW80_20985 and CIW80_20915  entB and entD   \n",
       "PCPSentF    CIW80_20935 and CIW80_20915  entF and entD   \n",
       "...                                 ...            ...   \n",
       "CLBTinter1                          NaN            NaN   \n",
       "CLBTinter2                          NaN            NaN   \n",
       "CLBTinter3                          NaN            NaN   \n",
       "CLBTNtex                            NaN            NaN   \n",
       "EX_clbtn_e                          NaN            NaN   \n",
       "\n",
       "                                    Subsystem    Compartment      LB      UB  \\\n",
       "ID                                                                             \n",
       "ICHORS      Secondary metabolite biosynthesis              c -1000.0  1000.0   \n",
       "ICHORT      Secondary metabolite biosynthesis              c     0.0  1000.0   \n",
       "DHBD        Secondary metabolite biosynthesis              c     0.0  1000.0   \n",
       "ArCPSentB   Secondary metabolite biosynthesis              c     0.0  1000.0   \n",
       "PCPSentF    Secondary metabolite biosynthesis              c     0.0  1000.0   \n",
       "...                                       ...            ...     ...     ...   \n",
       "CLBTinter1  Secondary metabolite biosynthesis              p     0.0  1000.0   \n",
       "CLBTinter2  Secondary metabolite biosynthesis              p     0.0  1000.0   \n",
       "CLBTinter3  Secondary metabolite biosynthesis              p     0.0  1000.0   \n",
       "CLBTNtex    Secondary metabolite biosynthesis  c and p nad e     0.0  1000.0   \n",
       "EX_clbtn_e             Extracellular exchange              e     0.0  1000.0   \n",
       "\n",
       "           rhea               biocyc metanetx.reaction kegg.reaction  \\\n",
       "ID                                                                     \n",
       "ICHORS      NaN  META:ISOCHORSYN-RXN               NaN           NaN   \n",
       "ICHORT      NaN  META:ISOCHORMAT-RXN               NaN           NaN   \n",
       "DHBD        NaN    META:DHBDEHYD-RXN               NaN           NaN   \n",
       "ArCPSentB   NaN       META:ENTDB-RXN        MNXR115861           NaN   \n",
       "PCPSentF    NaN       META:RXN-15889        MNXR119687           NaN   \n",
       "...         ...                  ...               ...           ...   \n",
       "CLBTinter1  NaN            RXN-21153               NaN           NaN   \n",
       "CLBTinter2  NaN            RXN-21102               NaN           NaN   \n",
       "CLBTinter3  NaN            RXN-21101               NaN           NaN   \n",
       "CLBTNtex    NaN                  NaN               NaN           NaN   \n",
       "EX_clbtn_e  NaN                  NaN               NaN           NaN   \n",
       "\n",
       "           seed.reaction  ec-code reactome  pubmed  \n",
       "ID                                                  \n",
       "ICHORS               NaN      NaN      NaN     NaN  \n",
       "ICHORT               NaN      NaN      NaN     NaN  \n",
       "DHBD                 NaN      NaN      NaN     NaN  \n",
       "ArCPSentB            NaN  2.7.8.7      NaN     NaN  \n",
       "PCPSentF             NaN  2.7.8.7      NaN     NaN  \n",
       "...                  ...      ...      ...     ...  \n",
       "CLBTinter1           NaN      NaN      NaN     NaN  \n",
       "CLBTinter2           NaN      NaN      NaN     NaN  \n",
       "CLBTinter3           NaN      NaN      NaN     NaN  \n",
       "CLBTNtex             NaN      NaN      NaN     NaN  \n",
       "EX_clbtn_e           NaN      NaN      NaN     NaN  \n",
       "\n",
       "[177 rows x 18 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reaction_df.rename(columns={'RHEA': 'rhea',\n",
    "                            'BioCyc':'biocyc',\n",
    "                            'MetaNetX (MNX) Equation': 'metanetx.reaction',\n",
    "                            'KEGG Reaction': 'kegg.reaction',\n",
    "                            'SEED Reaction': 'seed.reaction',\n",
    "                            'EC Number': 'ec-code',\n",
    "                            'Reactome Reaction': 'reactome',\n",
    "                            'Reference PMID': 'pubmed'},\n",
    "                    inplace = True)\n",
    "\n",
    "reaction_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>> UPDATING ICHORS\n",
      ">>>> UPDATING ICHORT\n",
      ">>>> UPDATING DHBD\n",
      ">>>> ADDING ArCPSentB\n",
      "unknown metabolite 'apoArCPentB_c' created\n",
      "unknown metabolite 'holoArCPentB_c' created\n",
      ">>>> ADDING PCPSentF\n",
      "unknown metabolite 'apoPCPentF_c' created\n",
      "unknown metabolite 'holoPCPentF_c' created\n",
      ">>>> ADDING ArCPpdsentB\n",
      ">>>> ADDING PCPpdsentF\n",
      ">>>> UPDATING DHBS\n",
      ">>>> UPDATING SERASr\n",
      ">>>> ADDING ENTERS1\n",
      "unknown metabolite '23dhbaArCPentB_c' created\n",
      ">>>> ADDING ENTERS2\n",
      "unknown metabolite 'serPCPentF_c' created\n",
      ">>>> ADDING ENTERS3\n",
      "unknown metabolite 'enter1PCPentF_c' created\n",
      ">>>> ADDING ENTERS4\n",
      ">>>> UPDATING ENTERES\n",
      ">>>> UPDATING FEENTERtpp\n",
      ">>>> UPDATING FEENTER2tpp\n",
      ">>>> UPDATING FEENTERtex\n",
      ">>>> UPDATING FEENTERexs\n",
      ">>>> UPDATING FEENTERtonex\n",
      ">>>> UPDATING FEENTERabcpp\n",
      ">>>> UPDATING FEENTERR1\n",
      ">>>> UPDATING FEENTERR2\n",
      ">>>> UPDATING FEENTERR4\n",
      ">>>> UPDATING FEENTERR3\n",
      ">>>> UPDATING FE3DHBZSabcpp\n",
      ">>>> UPDATING DHBSZ3FEabcpp\n",
      ">>>> UPDATING FEENTERES\n",
      ">>>> UPDATING FE3DHBZS3R\n",
      ">>>> UPDATING ENTERES2\n",
      ">>>> UPDATING EX_feenter_e\n",
      ">>>> UPDATING EX_enter_e\n",
      ">>>> UPDATING ENTERGLCT1\n",
      ">>>> UPDATING ENTERGLCT2\n",
      ">>>> ADDING ENTERGLCT3\n",
      "unknown metabolite 'entertg_c' created\n",
      ">>>> UPDATING SALCHS4abcpp\n",
      ">>>> ADDING SALCHS4tex\n",
      ">>>> UPDATING SALCHS4FEexs\n",
      ">>>> UPDATING SALCHS4FEtonex\n",
      ">>>> UPDATING SALCHS4FEabcpp\n",
      ">>>> ADDING SALCHS4FER1\n",
      ">>>> ADDING SALCHS4FER2\n",
      ">>>> ADDING SALCHS4FER3\n",
      ">>>> ADDING SALCHS4Hpp\n",
      "unknown metabolite 'salchs2_p' created\n",
      ">>>> ADDING SALCHS4FEHpp\n",
      ">>>> UPDATING SALCHS2FEabcpp\n",
      ">>>> ADDING SALCHS2tex\n",
      ">>>> UPDATING SALCHS2FEexs\n",
      ">>>> UPDATING SALCHS2FEtonex\n",
      ">>>> ADDING SALCHS2FER1\n",
      "unknown metabolite 'salchs2_c' created\n",
      ">>>> ADDING SALCHS2FER2\n",
      ">>>> ADDING SALCHS2FER3\n",
      ">>>> ADDING SALCHS2H\n",
      "unknown metabolite 'salchs1_c' created\n",
      "unknown metabolite 'salchsx_c' created\n",
      ">>>> ADDING SALCHS1H\n",
      ">>>> ADDING SALCHSXtpp\n",
      ">>>> UPDATING SALCHSXtex\n",
      ">>>> UPDATING EX_salchsx_e\n",
      ">>>> ADDING ARBTNS1\n",
      "unknown metabolite 'n6hlys_c' created\n",
      ">>>> ADDING ARBTNS2\n",
      "unknown metabolite 'n6acn6hlys_c' created\n",
      ">>>> ADDING ARBTNS3\n",
      "unknown metabolite 'n2citn6acn6hlys_c' created\n",
      ">>>> ADDING ARBTNS4\n",
      ">>>> UPDATING ARBTNtpp\n",
      ">>>> UPDATING ARBTNtex\n",
      ">>>> UPDATING ARBTNexs\n",
      ">>>> UPDATING ARBTNtonex\n",
      ">>>> UPDATING ARBTNabcpp\n",
      ">>>> UPDATING ARBTNR3\n",
      ">>>> UPDATING ARBTNR2\n",
      ">>>> UPDATING ARBTNR1\n",
      ">>>> UPDATING SALSYN\n",
      ">>>> ADDING ACPSirp1\n",
      "unknown metabolite 'apoACPirp1_c' created\n",
      "unknown metabolite 'holoACPirp1_c' created\n",
      ">>>> ADDING ArCPSirp2\n",
      "unknown metabolite 'apoArCPirp2_c' created\n",
      "unknown metabolite 'holoArCPirp2_c' created\n",
      ">>>> ADDING PCP1Sirp2\n",
      "unknown metabolite 'apoPCP1irp2_c' created\n",
      "unknown metabolite 'holoPCP1irp2_c' created\n",
      ">>>> ADDING PCP2Sirp2\n",
      "unknown metabolite 'apoPCP2irp2_c' created\n",
      "unknown metabolite 'holoPCP2irp2_c' created\n",
      ">>>> ADDING PCP3Sirp1\n",
      "unknown metabolite 'apoPCP3irp1_c' created\n",
      "unknown metabolite 'holoPCP3irp1_c' created\n",
      ">>>> ADDING ACPpdsirp1\n",
      ">>>> ADDING ArCPpdsirp2\n",
      ">>>> ADDING PCP1pdsirp2\n",
      ">>>> ADDING PCP2pdsirp2\n",
      ">>>> ADDING PCP3pdsirp1\n",
      ">>>> ADDING SALASr\n",
      "unknown metabolite 'salamp_c' created\n",
      ">>>> ADDING CYSASr\n",
      "unknown metabolite 'cysamp_c' created\n",
      ">>>> ADDING YBTS1\n",
      "unknown metabolite 'salArCPirp2_c' created\n",
      ">>>> ADDING YBTS2\n",
      "unknown metabolite 'cysPCP1irp2_c' created\n",
      ">>>> ADDING YBTS3\n",
      "unknown metabolite 'salcysPCP1irp2_c' created\n",
      ">>>> ADDING YBTS4\n",
      "unknown metabolite 'hptPCP1irp2_c' created\n",
      ">>>> ADDING YBTS5\n",
      "unknown metabolite 'cysPCP2irp2_c' created\n",
      ">>>> ADDING YBTS6\n",
      "unknown metabolite 'hptcysPCP2irp2_c' created\n",
      ">>>> ADDING YBTS7\n",
      "unknown metabolite 'hpttPCP2irp2_c' created\n",
      ">>>> ADDING YBTS8\n",
      "unknown metabolite 'malACPirp1_c' created\n",
      ">>>> ADDING YBTS9\n",
      "unknown metabolite 'KSirp1_c' created\n",
      "unknown metabolite 'hpttKSirp1_c' created\n",
      ">>>> ADDING YBTS10\n",
      "unknown metabolite 'hpttmalACPirp1_c' created\n",
      ">>>> ADDING YBTS11\n",
      "unknown metabolite 'ybt1ACPirp1_c' created\n",
      ">>>> ADDING YBTS12\n",
      "unknown metabolite 'cysPCP3irp1_c' created\n",
      ">>>> ADDING YBTS13\n",
      "unknown metabolite 'ybt2PCP3irp1_c' created\n",
      ">>>> ADDING YBTS14\n",
      "unknown metabolite 'ybt3PCP3irp1_c' created\n",
      ">>>> ADDING YBTS15\n",
      "unknown metabolite 'ybt4PCP3irp1_c' created\n",
      ">>>> ADDING YBTS16\n",
      "unknown metabolite 'ybt5PCP3irp1_c' created\n",
      ">>>> ADDING YBTS17\n",
      "unknown metabolite 'TEirp1_c' created\n",
      "unknown metabolite 'ybt_TEirp1_c' created\n",
      ">>>> ADDING YBTS18\n",
      "unknown metabolite 'ybt_c' created\n",
      ">>>> ADDING YBTtpp\n",
      "unknown metabolite 'ybt_p' created\n",
      ">>>> ADDING YBTtex\n",
      "unknown metabolite 'ybt_e' created\n",
      ">>>> ADDING FEYBTexs\n",
      "unknown metabolite 'feybt_e' created\n",
      ">>>> ADDING FEYBTtonex\n",
      "unknown metabolite 'feybt_p' created\n",
      ">>>> ADDING FEYBTabcpp\n",
      "unknown metabolite 'feybt_c' created\n",
      ">>>> ADDING FEYBTR\n",
      ">>>> ADDING ZnYBTexs\n",
      "unknown metabolite 'znybt_e' created\n",
      ">>>> ADDING ZnYBTtonex\n",
      "unknown metabolite 'znybt_p' created\n",
      ">>>> ADDING ZnYBTabcpp\n",
      "unknown metabolite 'znybt_c' created\n",
      ">>>> ADDING ZnYBTR\n",
      ">>>> ADDING CuYBTexs\n",
      "unknown metabolite 'cuybt_e' created\n",
      ">>>> ADDING CuYBTtonex\n",
      "unknown metabolite 'cuybt_p' created\n",
      ">>>> ADDING CuYBTabcpp\n",
      "unknown metabolite 'cuybt_c' created\n",
      ">>>> ADDING CuYBTR\n",
      ">>>> ADDING NiYBTexs\n",
      "unknown metabolite 'niybt_e' created\n",
      ">>>> ADDING NiYBTtonex\n",
      "unknown metabolite 'niybt_p' created\n",
      ">>>> ADDING NiYBTabcpp\n",
      "unknown metabolite 'niybt_c' created\n",
      ">>>> ADDING NiYBTR\n",
      ">>>> ADDING EX_ybt_e\n",
      ">>>> ADDING EX_feybt_e\n",
      ">>>> ADDING EX_znybt_e\n",
      ">>>> ADDING EX_cuybt_e\n",
      ">>>> ADDING EX_niybt_e\n",
      ">>>> ADDING ppantclbN\n",
      "unknown metabolite 'apoClbN_c' created\n",
      "unknown metabolite 'holoClbN_c' created\n",
      ">>>> ADDING ppantclbB\n",
      "unknown metabolite 'apoClbB_c' created\n",
      "unknown metabolite 'holoClbB_c' created\n",
      ">>>> ADDING ppantclbC\n",
      "unknown metabolite 'apoClbC_c' created\n",
      "unknown metabolite 'holoClbC_c' created\n",
      ">>>> ADDING ppantclbH\n",
      "unknown metabolite 'apoClbH_c' created\n",
      "unknown metabolite 'holoClbH_c' created\n",
      ">>>> ADDING ppantclbI\n",
      "unknown metabolite 'apoClbI_c' created\n",
      "unknown metabolite 'holoClbI_c' created\n",
      ">>>> ADDING ppantclbJ\n",
      "unknown metabolite 'apoClbJ_c' created\n",
      "unknown metabolite 'holoClbJ_c' created\n",
      ">>>> ADDING ppantclbK\n",
      "unknown metabolite 'apoClbK_c' created\n",
      "unknown metabolite 'holoClbK_c' created\n",
      ">>>> ADDING ppantclbO\n",
      "unknown metabolite 'apoClbO_c' created\n",
      "unknown metabolite 'holoClbO_c' created\n",
      ">>>> ADDING pdsclbN\n",
      ">>>> ADDING pdsclbB\n",
      ">>>> ADDING pdsclbC\n",
      ">>>> ADDING pdsclbH\n",
      ">>>> ADDING pdsclbI\n",
      ">>>> ADDING pdsclbJ\n",
      ">>>> ADDING pdsclbK\n",
      ">>>> ADDING pdsclbO\n",
      ">>>> ADDING ASPASr\n",
      "unknown metabolite 'aspamp_c' created\n",
      ">>>> ADDING CLBTS1\n",
      "unknown metabolite 'aspClbN_c' created\n",
      ">>>> ADDING CLBTS2\n",
      "unknown metabolite 'clb1ClbN_c' created\n",
      ">>>> ADDING ALAASr\n",
      "unknown metabolite 'alaamp_c' created\n",
      ">>>> ADDING CLBTS3\n",
      "unknown metabolite 'alaClbB_c' created\n",
      ">>>> ADDING CLBTS4\n",
      "unknown metabolite 'clb2ClbB_c' created\n",
      ">>>> ADDING CLBTS5\n",
      "unknown metabolite 'clb3ClbB_c' created\n",
      ">>>> ADDING CLBTS6\n",
      "unknown metabolite 'clb4ClbC_c' created\n",
      ">>>> ADDING AMETASr\n",
      "unknown metabolite 'ametamp_c' created\n",
      ">>>> ADDING CLBTS7\n",
      "unknown metabolite 'ametClbH_c' created\n",
      ">>>> ADDING CLBTS8\n",
      "unknown metabolite 'clb5ClbH_c' created\n",
      ">>>> ADDING CLBTS9\n",
      "unknown metabolite 'clb6ClbH_c' created\n",
      ">>>> ADDING CLBTS10\n",
      "unknown metabolite 'clb7ClbI_c' created\n",
      ">>>> ADDING GLYASr\n",
      "unknown metabolite 'glyamp_c' created\n",
      ">>>> ADDING CLBTS11\n",
      "unknown metabolite 'glyClbJ_c' created\n",
      ">>>> ADDING CLBTS12\n",
      "unknown metabolite 'clb8ClbJ_c' created\n",
      ">>>> ADDING CLBTS13\n",
      "unknown metabolite 'cysClbJ_c' created\n",
      ">>>> ADDING CLBTS14\n",
      "unknown metabolite 'clb10ClbJ_c' created\n",
      ">>>> ADDING CLBTS15\n",
      "unknown metabolite 'clb8_c' created\n",
      ">>>> ADDING CLBTS16\n",
      "unknown metabolite 'clb9_c' created\n",
      ">>>> ADDING CLBTS17\n",
      "unknown metabolite 'serClbH_c' created\n",
      ">>>> ADDING CLBTS18\n",
      "unknown metabolite 'clb11ClbH_c' created\n",
      ">>>> ADDING CLBTS19\n",
      "unknown metabolite 'clb12ClbH_c' created\n",
      ">>>> ADDING CLBTS20\n",
      "unknown metabolite 'aminomalClbG_c' created\n",
      "unknown metabolite 'clb13ClbG_c' created\n",
      ">>>> ADDING CLBTS21\n",
      "unknown metabolite 'clb14ClbK_c' created\n",
      ">>>> ADDING CLBTS22\n",
      "unknown metabolite 'clb15ClbK_c' created\n",
      ">>>> ADDING CLBTS23\n",
      "unknown metabolite 'cysClbK_c' created\n",
      ">>>> ADDING CLBTS24\n",
      "unknown metabolite 'clb16ClbK_c' created\n",
      ">>>> ADDING CLBTS25\n",
      "unknown metabolite 'clb17ClbO_c' created\n",
      ">>>> ADDING CLBTS26\n",
      "unknown metabolite 'clb18ClbO_c' created\n",
      ">>>> ADDING CLBTS27\n",
      "unknown metabolite 'preclb1491_c' created\n",
      ">>>> ADDING PRECLBTNmate\n",
      "unknown metabolite 'preclb1491_p' created\n",
      ">>>> ADDING CLBTNpeptidase\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unknown metabolite 'preclbInter_p' created\n",
      "unknown metabolite 'myrasp_D_p' created\n",
      ">>>> ADDING DM_MYRASP\n",
      ">>>> ADDING CLBTinter1\n",
      "unknown metabolite 'preclb16a_p' created\n",
      ">>>> ADDING CLBTinter2\n",
      "unknown metabolite 'preclb16b_p' created\n",
      ">>>> ADDING CLBTinter3\n",
      "unknown metabolite 'clbtn_p' created\n",
      ">>>> ADDING CLBTNtex\n",
      "unknown metabolite 'clbtn_e' created\n",
      ">>>> ADDING EX_clbtn_e\n"
     ]
    }
   ],
   "source": [
    "# add reactions\n",
    "for rxn in reaction_df.index.values:\n",
    "    \n",
    "    # Create reaction and add to model, add reaction name and subsystem\n",
    "    reaction = cobra.Reaction(rxn)\n",
    "    if reaction not in EcN_model.reactions: \n",
    "        print('>>>> ADDING', rxn) \n",
    "        EcN_model.add_reaction(reaction)\n",
    "\n",
    "        # Add name and subsystem\n",
    "        reaction.name = reaction_df.loc[rxn, 'Name']\n",
    "        reaction.subsystem = reaction_df.loc[rxn, 'Subsystem']\n",
    "\n",
    "        # Add boundaries when given\n",
    "        if pd.isnull(reaction_df.loc[rxn, 'LB']):\n",
    "            pass\n",
    "        else:\n",
    "            reaction.lower_bound = reaction_df.loc[rxn, 'LB']\n",
    "        if pd.isnull(reaction_df.loc[rxn, 'UB']):\n",
    "            pass\n",
    "        else:\n",
    "            reaction.upper_bound = reaction_df.loc[rxn, 'UB']\n",
    "\n",
    "        # Add reaction formula\n",
    "        reaction.build_reaction_from_string(reaction_df.loc[rxn, 'Reaction Formula'])\n",
    "\n",
    "        # Add gene reaction rule when given\n",
    "        if str(reaction_df.loc[rxn, 'Gene Rules']).find('CIW80') != -1: # If \"find\" doesn't find a match, it returns -1\n",
    "            reaction.gene_reaction_rule = reaction_df.loc[rxn, 'Gene Rules']\n",
    "\n",
    "        # Add compartment\n",
    "        if 'and' in reaction_df.loc[rxn, 'Compartment']: # take into account the cases where there are two comparments\n",
    "            compartment = reaction_df.loc[rxn, 'Compartment'].split()\n",
    "            reaction.compartments.add(compartment[0]) # add the compartment in front of \"and\"\n",
    "            reaction.compartments.add(compartment[2]) # add the compartment behind \"and\"\n",
    "            if len(compartment) == 5:\n",
    "                reaction.compartments.add(compartment[4]) # add the compartment behind \"and\" \n",
    "        else:\n",
    "            reaction.compartments.add(reaction_df.loc[rxn, 'Compartment'])\n",
    "\n",
    "        # Add database annotations when present\n",
    "        for annot in ['rhea', 'biocyc','metanetx.reaction', 'kegg.reaction', 'seed.reaction', 'ec-code', 'reactome', 'pubmed']:\n",
    "            if pd.isnull(reaction_df.loc[rxn, annot]):\n",
    "                pass\n",
    "            else:\n",
    "                reaction.annotation[annot] = [reaction_df.loc[rxn, annot]]\n",
    "\n",
    "        # Add to reaction overview\n",
    "        rxn_origin_df.loc[rxn, 'origin'] = reaction_df.loc[rxn, 'Status']\n",
    "        rxn_origin_df.loc[rxn, 'added'] = 'manual'\n",
    "        rxn_origin_df.loc[rxn, 'notebook'] = '4.6'\n",
    "        rxn_origin_df.loc[rxn, 'notes'] = 'none'\n",
    "\n",
    "\n",
    "        ### Genes ###\n",
    "\n",
    "        # Update gene name\n",
    "        gene_name = re.split(' and | or ', str(reaction_df.loc[rxn, 'Gene Name']))\n",
    "        gene_rules = re.split(' and | or ', str(reaction_df.loc[rxn, 'Gene Rules']))\n",
    "\n",
    "        if len(gene_name) == len(gene_rules) and gene_rules[0].find('CIW80') != -1: # If \"find\" doesn't find a match, it returns -1\n",
    "            for idx, gene in enumerate(gene_rules):\n",
    "                EcN_model.genes.get_by_id(gene).name = gene_name[idx]\n",
    "\n",
    "                # Register the genes\n",
    "                gene_origin_df.loc[gene, 'EcN_gene'] = gene\n",
    "                gene_origin_df.loc[gene, 'origin'] = reaction_df.loc[rxn, 'Status']\n",
    "                gene_origin_df.loc[gene, 'added'] = 'manual'\n",
    "                gene_origin_df.loc[gene, 'notebook'] = '4.6'\n",
    "                gene_origin_df.loc[gene, 'notes'] = 'none'\n",
    "            \n",
    "    ### If reaction is present, update information\n",
    "    else:\n",
    "        print('>>>> UPDATING', rxn) \n",
    "        EcN_model.reactions.get_by_id(rxn)\n",
    "        # Add name and subsystem\n",
    "        EcN_model.reactions.get_by_id(rxn).name = reaction_df.loc[rxn, 'Name']\n",
    "        EcN_model.reactions.get_by_id(rxn).subsystem = reaction_df.loc[rxn, 'Subsystem']\n",
    "\n",
    "        # Add boundaries when given\n",
    "        if pd.isnull(reaction_df.loc[rxn, 'LB']):\n",
    "            pass\n",
    "        else:\n",
    "            EcN_model.reactions.get_by_id(rxn).lower_bound = reaction_df.loc[rxn, 'LB']\n",
    "        if pd.isnull(reaction_df.loc[rxn, 'UB']):\n",
    "            pass\n",
    "        else:\n",
    "            EcN_model.reactions.get_by_id(rxn).upper_bound = reaction_df.loc[rxn, 'UB']\n",
    "\n",
    "        # Add reaction formula\n",
    "        EcN_model.reactions.get_by_id(rxn).build_reaction_from_string(reaction_df.loc[rxn, 'Reaction Formula'])\n",
    "\n",
    "        # Add gene reaction rule when given\n",
    "        if str(reaction_df.loc[rxn, 'Gene Rules']).find('CIW80') != -1: # If \"find\" doesn't find a match, it returns -1\n",
    "            EcN_model.reactions.get_by_id(rxn).gene_reaction_rule = reaction_df.loc[rxn, 'Gene Rules']\n",
    "\n",
    "        # Add compartment\n",
    "        if 'and' in reaction_df.loc[rxn, 'Compartment']: # take into account the cases where there are two comparments\n",
    "            compartment = reaction_df.loc[rxn, 'Compartment'].split()\n",
    "            EcN_model.reactions.get_by_id(rxn).compartments.add(compartment[0]) # add the compartment in front of \"and\"\n",
    "            EcN_model.reactions.get_by_id(rxn).compartments.add(compartment[2]) # add the compartment behind \"and\"\n",
    "            if len(compartment) == 5:\n",
    "                EcN_model.reactions.get_by_id(rxn).compartments.add(compartment[4]) # add the compartment behind \"and\" \n",
    "        else:\n",
    "            EcN_model.reactions.get_by_id(rxn).compartments.add(reaction_df.loc[rxn, 'Compartment'])\n",
    "\n",
    "        # Add database annotations when present\n",
    "        for annot in ['rhea', 'biocyc','metanetx.reaction', 'kegg.reaction', 'seed.reaction', 'ec-code', 'reactome', 'pubmed']:\n",
    "            if pd.isnull(reaction_df.loc[rxn, annot]):\n",
    "                pass\n",
    "            else:\n",
    "                EcN_model.reactions.get_by_id(rxn).annotation[annot] = [reaction_df.loc[rxn, annot]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EcN_gene</th>\n",
       "      <th>origin</th>\n",
       "      <th>added</th>\n",
       "      <th>notebook</th>\n",
       "      <th>notes</th>\n",
       "      <th>gene.1</th>\n",
       "      <th>gene.1.1</th>\n",
       "      <th>gene.1.1.1</th>\n",
       "      <th>gene.1.1.1.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b2551</th>\n",
       "      <td>CIW80_06560</td>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0870</th>\n",
       "      <td>CIW80_22360</td>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3368</th>\n",
       "      <td>CIW80_11380</td>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2436</th>\n",
       "      <td>CIW80_06055</td>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0008</th>\n",
       "      <td>CIW80_17715</td>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_03455</th>\n",
       "      <td>CIW80_03455</td>\n",
       "      <td>New Reaction</td>\n",
       "      <td>manual</td>\n",
       "      <td>4.6</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_03445</th>\n",
       "      <td>CIW80_03445</td>\n",
       "      <td>New Reaction</td>\n",
       "      <td>manual</td>\n",
       "      <td>4.6</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_03440</th>\n",
       "      <td>CIW80_03440</td>\n",
       "      <td>New Reaction</td>\n",
       "      <td>manual</td>\n",
       "      <td>4.6</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_03410</th>\n",
       "      <td>CIW80_03410</td>\n",
       "      <td>New Reaction</td>\n",
       "      <td>manual</td>\n",
       "      <td>4.6</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIW80_03395</th>\n",
       "      <td>CIW80_03395</td>\n",
       "      <td>New Reaction</td>\n",
       "      <td>manual</td>\n",
       "      <td>4.6</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1573 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                EcN_gene        origin      added notebook notes  gene.1  \\\n",
       "gene                                                                       \n",
       "b2551        CIW80_06560       iML1515  automatic      3.1  none     NaN   \n",
       "b0870        CIW80_22360       iML1515  automatic      3.1  none     NaN   \n",
       "b3368        CIW80_11380       iML1515  automatic      3.1  none     NaN   \n",
       "b2436        CIW80_06055       iML1515  automatic      3.1  none     NaN   \n",
       "b0008        CIW80_17715       iML1515  automatic      3.1  none     NaN   \n",
       "...                  ...           ...        ...      ...   ...     ...   \n",
       "CIW80_03455  CIW80_03455  New Reaction     manual      4.6  none     NaN   \n",
       "CIW80_03445  CIW80_03445  New Reaction     manual      4.6  none     NaN   \n",
       "CIW80_03440  CIW80_03440  New Reaction     manual      4.6  none     NaN   \n",
       "CIW80_03410  CIW80_03410  New Reaction     manual      4.6  none     NaN   \n",
       "CIW80_03395  CIW80_03395  New Reaction     manual      4.6  none     NaN   \n",
       "\n",
       "             gene.1.1  gene.1.1.1  gene.1.1.1.1  \n",
       "gene                                             \n",
       "b2551             NaN         NaN           NaN  \n",
       "b0870             NaN         NaN           NaN  \n",
       "b3368             NaN         NaN           NaN  \n",
       "b2436             NaN         NaN           NaN  \n",
       "b0008             NaN         NaN           NaN  \n",
       "...               ...         ...           ...  \n",
       "CIW80_03455       NaN         NaN           NaN  \n",
       "CIW80_03445       NaN         NaN           NaN  \n",
       "CIW80_03440       NaN         NaN           NaN  \n",
       "CIW80_03410       NaN         NaN           NaN  \n",
       "CIW80_03395       NaN         NaN           NaN  \n",
       "\n",
       "[1573 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check correct addition of genes to gene overview table\n",
    "gene_origin_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Update metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>metabolite_name</th>\n",
       "      <th>formula</th>\n",
       "      <th>compartment</th>\n",
       "      <th>formula_weight</th>\n",
       "      <th>charge</th>\n",
       "      <th>bigg.metabolite</th>\n",
       "      <th>inchikey</th>\n",
       "      <th>metanetx.chemical</th>\n",
       "      <th>sbo</th>\n",
       "      <th>seed.compound</th>\n",
       "      <th>biocyc</th>\n",
       "      <th>chebi</th>\n",
       "      <th>envipath</th>\n",
       "      <th>hmdb</th>\n",
       "      <th>kegg.compound</th>\n",
       "      <th>kegg.drug</th>\n",
       "      <th>reactome.compound</th>\n",
       "      <th>sabiork.compound</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metabolite_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>holoArCPentB_c</th>\n",
       "      <td>New metabolite</td>\n",
       "      <td>Holo-[EntB aryl-carrier protein]</td>\n",
       "      <td>C14H25N3O8PSR</td>\n",
       "      <td>c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MNXM147417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:Holo-EntB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>apoArCPentB_c</th>\n",
       "      <td>New metabolite</td>\n",
       "      <td>Apo-[EntB aryl-carrier protein]</td>\n",
       "      <td>C3H5NO2R</td>\n",
       "      <td>c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MNXM148907</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:Apo-EntB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>holoPCPentF_c</th>\n",
       "      <td>New metabolite</td>\n",
       "      <td>Holo-[EntF L-seryl-carrier protein]</td>\n",
       "      <td>C14H25N3O8PSR</td>\n",
       "      <td>c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MNXM147418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:Holo-EntF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>apoPCPentF_c</th>\n",
       "      <td>New metabolite</td>\n",
       "      <td>Apo-[EntF L-seryl-carrier protein]</td>\n",
       "      <td>C3H5NO2R</td>\n",
       "      <td>c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MNXM148908</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:Apo-EntF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23dhbaArCPentB_c</th>\n",
       "      <td>New metabolite</td>\n",
       "      <td>2,3-dihydroxybenzoyl-[EntB aryl-carrier protein]</td>\n",
       "      <td>C21H29N3O11PSR</td>\n",
       "      <td>c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MNXM147916</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:23DHB-EntB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>preclbInter_p</th>\n",
       "      <td>New metabolite</td>\n",
       "      <td>unstable precolibactin intermediate</td>\n",
       "      <td>C37H52N9O10S2</td>\n",
       "      <td>p</td>\n",
       "      <td>843.30438</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GYHFSHJPDFYZDE-SHROTLKGSA-Q</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:CPD-22758</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>preclb16a_p</th>\n",
       "      <td>New metabolite</td>\n",
       "      <td>precolibactin 16a</td>\n",
       "      <td>C37H42N9O6S2</td>\n",
       "      <td>p</td>\n",
       "      <td>771.26212</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>JWSSANKUQQCBBA-UWGAKSRVSA-O</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:CPD-22721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>preclb16b_p</th>\n",
       "      <td>New metabolite</td>\n",
       "      <td>precolibactin 16b</td>\n",
       "      <td>C37H39N9O6S2</td>\n",
       "      <td>p</td>\n",
       "      <td>769.24647</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NWUPXZAZUAPYEW-ROUUACIJSA-N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:CPD-22720</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clbtn_p</th>\n",
       "      <td>New metabolite</td>\n",
       "      <td>Colibactin</td>\n",
       "      <td>C37H38N8O7S2</td>\n",
       "      <td>p</td>\n",
       "      <td>770.23048</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZWKHDAZPVITMAI-ROUUACIJSA-N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:CPD-22719</td>\n",
       "      <td>CHEBI:156303</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clbtn_e</th>\n",
       "      <td>New metabolite</td>\n",
       "      <td>Colibactin</td>\n",
       "      <td>C37H38N8O7S2</td>\n",
       "      <td>e</td>\n",
       "      <td>770.23048</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZWKHDAZPVITMAI-ROUUACIJSA-N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>META:CPD-22719</td>\n",
       "      <td>CHEBI:156303</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Status  \\\n",
       "metabolite_ID                      \n",
       "holoArCPentB_c    New metabolite   \n",
       "apoArCPentB_c     New metabolite   \n",
       "holoPCPentF_c     New metabolite   \n",
       "apoPCPentF_c      New metabolite   \n",
       "23dhbaArCPentB_c  New metabolite   \n",
       "...                          ...   \n",
       "preclbInter_p     New metabolite   \n",
       "preclb16a_p       New metabolite   \n",
       "preclb16b_p       New metabolite   \n",
       "clbtn_p           New metabolite   \n",
       "clbtn_e           New metabolite   \n",
       "\n",
       "                                                   metabolite_name  \\\n",
       "metabolite_ID                                                        \n",
       "holoArCPentB_c                    Holo-[EntB aryl-carrier protein]   \n",
       "apoArCPentB_c                      Apo-[EntB aryl-carrier protein]   \n",
       "holoPCPentF_c                  Holo-[EntF L-seryl-carrier protein]   \n",
       "apoPCPentF_c                    Apo-[EntF L-seryl-carrier protein]   \n",
       "23dhbaArCPentB_c  2,3-dihydroxybenzoyl-[EntB aryl-carrier protein]   \n",
       "...                                                            ...   \n",
       "preclbInter_p                  unstable precolibactin intermediate   \n",
       "preclb16a_p                                      precolibactin 16a   \n",
       "preclb16b_p                                      precolibactin 16b   \n",
       "clbtn_p                                                 Colibactin   \n",
       "clbtn_e                                                 Colibactin   \n",
       "\n",
       "                         formula compartment  formula_weight  charge  \\\n",
       "metabolite_ID                                                          \n",
       "holoArCPentB_c     C14H25N3O8PSR           c             NaN    -1.0   \n",
       "apoArCPentB_c           C3H5NO2R           c             NaN     0.0   \n",
       "holoPCPentF_c      C14H25N3O8PSR           c             NaN    -1.0   \n",
       "apoPCPentF_c            C3H5NO2R           c             NaN     0.0   \n",
       "23dhbaArCPentB_c  C21H29N3O11PSR           c             NaN    -1.0   \n",
       "...                          ...         ...             ...     ...   \n",
       "preclbInter_p      C37H52N9O10S2           p       843.30438     3.0   \n",
       "preclb16a_p         C37H42N9O6S2           p       771.26212     1.0   \n",
       "preclb16b_p         C37H39N9O6S2           p       769.24647     0.0   \n",
       "clbtn_p             C37H38N8O7S2           p       770.23048     0.0   \n",
       "clbtn_e             C37H38N8O7S2           e       770.23048     0.0   \n",
       "\n",
       "                  bigg.metabolite                     inchikey  \\\n",
       "metabolite_ID                                                    \n",
       "holoArCPentB_c                NaN                          NaN   \n",
       "apoArCPentB_c                 NaN                          NaN   \n",
       "holoPCPentF_c                 NaN                          NaN   \n",
       "apoPCPentF_c                  NaN                          NaN   \n",
       "23dhbaArCPentB_c              NaN                          NaN   \n",
       "...                           ...                          ...   \n",
       "preclbInter_p                 NaN  GYHFSHJPDFYZDE-SHROTLKGSA-Q   \n",
       "preclb16a_p                   NaN  JWSSANKUQQCBBA-UWGAKSRVSA-O   \n",
       "preclb16b_p                   NaN  NWUPXZAZUAPYEW-ROUUACIJSA-N   \n",
       "clbtn_p                       NaN  ZWKHDAZPVITMAI-ROUUACIJSA-N   \n",
       "clbtn_e                       NaN  ZWKHDAZPVITMAI-ROUUACIJSA-N   \n",
       "\n",
       "                 metanetx.chemical  sbo seed.compound           biocyc  \\\n",
       "metabolite_ID                                                            \n",
       "holoArCPentB_c          MNXM147417  NaN           NaN   META:Holo-EntB   \n",
       "apoArCPentB_c           MNXM148907  NaN           NaN    META:Apo-EntB   \n",
       "holoPCPentF_c           MNXM147418  NaN           NaN   META:Holo-EntF   \n",
       "apoPCPentF_c            MNXM148908  NaN           NaN    META:Apo-EntF   \n",
       "23dhbaArCPentB_c        MNXM147916  NaN           NaN  META:23DHB-EntB   \n",
       "...                            ...  ...           ...              ...   \n",
       "preclbInter_p                  NaN  NaN           NaN   META:CPD-22758   \n",
       "preclb16a_p                    NaN  NaN           NaN   META:CPD-22721   \n",
       "preclb16b_p                    NaN  NaN           NaN   META:CPD-22720   \n",
       "clbtn_p                        NaN  NaN           NaN   META:CPD-22719   \n",
       "clbtn_e                        NaN  NaN           NaN   META:CPD-22719   \n",
       "\n",
       "                         chebi  envipath  hmdb kegg.compound  kegg.drug  \\\n",
       "metabolite_ID                                                             \n",
       "holoArCPentB_c             NaN       NaN   NaN           NaN        NaN   \n",
       "apoArCPentB_c              NaN       NaN   NaN           NaN        NaN   \n",
       "holoPCPentF_c              NaN       NaN   NaN           NaN        NaN   \n",
       "apoPCPentF_c               NaN       NaN   NaN           NaN        NaN   \n",
       "23dhbaArCPentB_c           NaN       NaN   NaN           NaN        NaN   \n",
       "...                        ...       ...   ...           ...        ...   \n",
       "preclbInter_p              NaN       NaN   NaN           NaN        NaN   \n",
       "preclb16a_p                NaN       NaN   NaN           NaN        NaN   \n",
       "preclb16b_p                NaN       NaN   NaN           NaN        NaN   \n",
       "clbtn_p           CHEBI:156303       NaN   NaN           NaN        NaN   \n",
       "clbtn_e           CHEBI:156303       NaN   NaN           NaN        NaN   \n",
       "\n",
       "                  reactome.compound  sabiork.compound  \n",
       "metabolite_ID                                          \n",
       "holoArCPentB_c                  NaN               NaN  \n",
       "apoArCPentB_c                   NaN               NaN  \n",
       "holoPCPentF_c                   NaN               NaN  \n",
       "apoPCPentF_c                    NaN               NaN  \n",
       "23dhbaArCPentB_c                NaN               NaN  \n",
       "...                             ...               ...  \n",
       "preclbInter_p                   NaN               NaN  \n",
       "preclb16a_p                     NaN               NaN  \n",
       "preclb16b_p                     NaN               NaN  \n",
       "clbtn_p                         NaN               NaN  \n",
       "clbtn_e                         NaN               NaN  \n",
       "\n",
       "[110 rows x 19 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the table with information on reactions to add\n",
    "mtb_df = pd.read_csv('../tables/sec_met_paths_metabolites.csv')\n",
    "mtb_df['metabolite_ID'] = mtb_df['metabolite_ID'].str.strip() #Remove white spaces\n",
    "mtb_df.set_index('metabolite_ID', inplace=True)\n",
    "\n",
    "# Names in the csv file are already in the correct annotation format\n",
    "mtb_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>> UPDATING holoArCPentB_c\n",
      ">>>> UPDATING apoArCPentB_c\n",
      ">>>> UPDATING holoPCPentF_c\n",
      ">>>> UPDATING apoPCPentF_c\n",
      ">>>> UPDATING 23dhbaArCPentB_c\n",
      ">>>> UPDATING serPCPentF_c\n",
      ">>>> UPDATING enter1PCPentF_c\n",
      ">>>> UPDATING entertg_c\n",
      ">>>> UPDATING n6hlys_c\n",
      ">>>> UPDATING n6acn6hlys_c\n",
      ">>>> UPDATING n2citn6acn6hlys_c\n",
      ">>>> UPDATING apoACPirp1_c\n",
      ">>>> UPDATING apoArCPirp2_c\n",
      ">>>> UPDATING apoPCP1irp2_c\n",
      ">>>> UPDATING apoPCP2irp2_c\n",
      ">>>> UPDATING apoPCP3irp1_c\n",
      ">>>> UPDATING holoACPirp1_c\n",
      ">>>> UPDATING holoArCPirp2_c\n",
      ">>>> UPDATING holoPCP1irp2_c\n",
      ">>>> UPDATING holoPCP2irp2_c\n",
      ">>>> UPDATING holoPCP3irp1_c\n",
      ">>>> UPDATING salamp_c\n",
      ">>>> UPDATING cysamp_c\n",
      ">>>> UPDATING salArCPirp2_c\n",
      ">>>> UPDATING cysPCP1irp2_c\n",
      ">>>> UPDATING salcysPCP1irp2_c\n",
      "No charge\n",
      ">>>> UPDATING hptPCP1irp2_c\n",
      "No charge\n",
      ">>>> UPDATING cysPCP2irp2_c\n",
      "No charge\n",
      ">>>> UPDATING hptcysPCP2irp2_c\n",
      "No charge\n",
      ">>>> UPDATING hpttPCP2irp2_c\n",
      "No charge\n",
      ">>>> UPDATING malACPirp1_c\n",
      "No charge\n",
      ">>>> UPDATING KSirp1_c\n",
      "No charge\n",
      ">>>> UPDATING hpttKSirp1_c\n",
      "No charge\n",
      ">>>> UPDATING hpttmalACPirp1_c\n",
      "No charge\n",
      ">>>> UPDATING ybt1ACPirp1_c\n",
      "No charge\n",
      ">>>> UPDATING cysPCP3irp1_c\n",
      "No charge\n",
      ">>>> UPDATING ybt2PCP3irp1_c\n",
      "No charge\n",
      ">>>> UPDATING ybt3PCP3irp1_c\n",
      "No charge\n",
      ">>>> UPDATING ybt4PCP3irp1_c\n",
      "No charge\n",
      ">>>> UPDATING ybt5PCP3irp1_c\n",
      "No charge\n",
      ">>>> UPDATING TEirp1_c\n",
      "No charge\n",
      ">>>> UPDATING ybt_TEirp1_c\n",
      "No charge\n",
      ">>>> UPDATING ybt_c\n",
      ">>>> UPDATING ybt_p\n",
      ">>>> UPDATING ybt_e\n",
      ">>>> UPDATING feybt_c\n",
      ">>>> UPDATING feybt_p\n",
      ">>>> UPDATING feybt_e\n",
      ">>>> UPDATING znybt_c\n",
      ">>>> UPDATING znybt_p\n",
      ">>>> UPDATING znybt_e\n",
      ">>>> UPDATING cuybt_c\n",
      ">>>> UPDATING cuybt_p\n",
      ">>>> UPDATING cuybt_e\n",
      ">>>> UPDATING niybt_c\n",
      ">>>> UPDATING niybt_p\n",
      ">>>> UPDATING niybt_e\n",
      ">>>> UPDATING apoClbN_c\n",
      ">>>> UPDATING holoClbN_c\n",
      ">>>> UPDATING apoClbB_c\n",
      ">>>> UPDATING holoClbB_c\n",
      ">>>> UPDATING apoClbC_c\n",
      ">>>> UPDATING holoClbC_c\n",
      ">>>> UPDATING apoClbH_c\n",
      ">>>> UPDATING holoClbH_c\n",
      ">>>> UPDATING apoClbI_c\n",
      ">>>> UPDATING holoClbI_c\n",
      ">>>> UPDATING apoClbJ_c\n",
      ">>>> UPDATING holoClbJ_c\n",
      ">>>> UPDATING apoClbK_c\n",
      ">>>> UPDATING holoClbK_c\n",
      ">>>> UPDATING apoClbO_c\n",
      ">>>> UPDATING holoClbO_c\n",
      ">>>> UPDATING aspamp_c\n",
      ">>>> UPDATING aspClbN_c\n",
      "No charge\n",
      ">>>> UPDATING clb1ClbN_c\n",
      "No charge\n",
      ">>>> UPDATING alaamp_c\n",
      ">>>> UPDATING alaClbB_c\n",
      "No charge\n",
      ">>>> UPDATING clb2ClbB_c\n",
      "No charge\n",
      ">>>> UPDATING clb3ClbB_c\n",
      "No charge\n",
      ">>>> UPDATING clb4ClbC_c\n",
      "No charge\n",
      ">>>> UPDATING ametamp_c\n",
      ">>>> UPDATING ametClbH_c\n",
      "No charge\n",
      ">>>> UPDATING clb5ClbH_c\n",
      "No charge\n",
      ">>>> UPDATING clb6ClbH_c\n",
      "No charge\n",
      ">>>> UPDATING clb7ClbI_c\n",
      "No charge\n",
      ">>>> UPDATING glyamp_c\n",
      ">>>> UPDATING glyClbJ_c\n",
      "No charge\n",
      ">>>> UPDATING clb8ClbJ_c\n",
      "No charge\n",
      ">>>> UPDATING cysClbJ_c\n",
      "No charge\n",
      ">>>> UPDATING clb10ClbJ_c\n",
      "No charge\n",
      ">>>> UPDATING serClbH_c\n",
      "No charge\n",
      ">>>> UPDATING clb11ClbH_c\n",
      "No charge\n",
      ">>>> UPDATING clb12ClbH_c\n",
      "No charge\n",
      ">>>> UPDATING aminomalClbG_c\n",
      "No charge\n",
      ">>>> UPDATING clb13ClbG_c\n",
      "No charge\n",
      ">>>> UPDATING clb14ClbK_c\n",
      "No charge\n",
      ">>>> UPDATING clb15ClbK_c\n",
      "No charge\n",
      ">>>> UPDATING cysClbK_c\n",
      "No charge\n",
      ">>>> UPDATING clb16ClbK_c\n",
      "No charge\n",
      ">>>> UPDATING clb17ClbO_c\n",
      "No charge\n",
      ">>>> UPDATING clb18ClbO_c\n",
      "No charge\n",
      ">>>> UPDATING preclb1491_c\n",
      ">>>> UPDATING preclb1491_p\n",
      ">>>> UPDATING myrasp_D_p\n",
      ">>>> UPDATING preclbInter_p\n",
      ">>>> UPDATING preclb16a_p\n",
      ">>>> UPDATING preclb16b_p\n",
      ">>>> UPDATING clbtn_p\n",
      ">>>> UPDATING clbtn_e\n"
     ]
    }
   ],
   "source": [
    "# update metabolites\n",
    "for mtb in mtb_df.index.values:\n",
    "    print('>>>> UPDATING', mtb)\n",
    "    \n",
    "    # Get the metabolite\n",
    "    mtb = EcN_model.metabolites.get_by_id(mtb)\n",
    "    \n",
    "    # Update name, formula, compartment & charge, when this is not yet present\n",
    "    if len(mtb.name) == 0:\n",
    "        mtb.name = mtb_df.loc[mtb.id, 'metabolite_name']\n",
    "#         mtb.formula = mtb_df.loc[mtb.id, 'formula']\n",
    "        mtb.compartment = mtb_df.loc[mtb.id, 'compartment']\n",
    "        \n",
    "    if pd.isnull(mtb_df.loc[mtb.id, 'charge']):\n",
    "        print('No charge')\n",
    "    else:\n",
    "        mtb.charge = int(mtb_df.loc[mtb.id, 'charge'])\n",
    "\n",
    "    \n",
    "\n",
    "    # Add database annotations when present\n",
    "    for annot in ['bigg.metabolite', 'inchikey', 'metanetx.chemical', 'sbo', 'seed.compound', 'biocyc', 'chebi',\n",
    "                  'envipath', 'hmdb', 'kegg.compound', 'kegg.drug', 'reactome.compound', 'sabiork.compound']:\n",
    "        if pd.isnull(mtb_df.loc[mtb.id, annot]):\n",
    "            pass\n",
    "        else:\n",
    "            mtb.annotation[annot] = [mtb_df.loc[mtb.id, annot]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'charge': 3, 'Fe': 1}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EcN_model.reactions.FEYBTR.check_mass_balance()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Check pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EX_arbtn_e 4.513220338983051\n",
      "EX_feenter_e 3.0581974248927075\n",
      "EX_enter_e 3.0581974248927075\n",
      "EX_salchsx_e 5.7647586451079915\n",
      "EX_ybt_e 1.983303556927076\n",
      "EX_feybt_e 1.9833035569270747\n",
      "EX_znybt_e 1.9833035569270752\n",
      "EX_cuybt_e 1.9833035569270752\n",
      "EX_niybt_e 1.9833035569270752\n",
      "CLBTNtex 1.0111711711711728\n"
     ]
    }
   ],
   "source": [
    "with EcN_model:\n",
    "    for rxn in ['EX_arbtn_e', 'EX_feenter_e', 'EX_enter_e', 'EX_salchsx_e', 'EX_ybt_e', 'EX_feybt_e', 'EX_znybt_e', 'EX_cuybt_e', 'EX_niybt_e', 'CLBTNtex']:\n",
    "#     for rxn in reaction_df.index.values:\n",
    "        EcN_model.objective = EcN_model.reactions.get_by_id(rxn)\n",
    "        solution = EcN_model.slim_optimize()\n",
    "        print(rxn, solution)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>origin</th>\n",
       "      <th>added</th>\n",
       "      <th>notebook</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reaction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CYTDK2</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XPPT</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HXPRT</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDPK5</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SHK3Dr</th>\n",
       "      <td>iML1515</td>\n",
       "      <td>automatic</td>\n",
       "      <td>3.1</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLBTinter1</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>manual</td>\n",
       "      <td>4.6</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLBTinter2</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>manual</td>\n",
       "      <td>4.6</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLBTinter3</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>manual</td>\n",
       "      <td>4.6</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLBTNtex</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>manual</td>\n",
       "      <td>4.6</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_clbtn_e</th>\n",
       "      <td>New Reaction</td>\n",
       "      <td>manual</td>\n",
       "      <td>4.6</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3221 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  origin      added notebook notes\n",
       "reaction                                          \n",
       "CYTDK2           iML1515  automatic      3.1  none\n",
       "XPPT             iML1515  automatic      3.1  none\n",
       "HXPRT            iML1515  automatic      3.1  none\n",
       "NDPK5            iML1515  automatic      3.1  none\n",
       "SHK3Dr           iML1515  automatic      3.1  none\n",
       "...                  ...        ...      ...   ...\n",
       "CLBTinter1  New Reaction     manual      4.6  none\n",
       "CLBTinter2  New Reaction     manual      4.6  none\n",
       "CLBTinter3  New Reaction     manual      4.6  none\n",
       "CLBTNtex    New Reaction     manual      4.6  none\n",
       "EX_clbtn_e  New Reaction     manual      4.6  none\n",
       "\n",
       "[3221 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxn_origin_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Methionine salvage pathway\n",
    "In the paper by Sun et al. (2015), they describe two reactions in which EcN converts methylthioribose (MTR) to methylthioribulose-1-phosphate (MTRu-1-P) by two enzymes:\n",
    "- MTR kinase (mtnK, ecn2295)\n",
    "- methylthioribose-1-phosphate (MTR-1-P) isomerase (mtnA, ecn2294)\n",
    "\n",
    "These reactions are present in the BiGG database as \"MTRK\" (5-methylthioribose kinase) and \"MTRI\" (5-methylthioribose-1-phosphate isomerase). Both are added here to the model.\n",
    "\n",
    "Homology:\n",
    "- ecn2295 > Q986M8 > CIW80_17025 (based on homology)\n",
    "- ecn2294 > CIW80_17030 (based on annotation in genome: mtnA, \"S-methyl-5-thioribose-1-phosphate isomerase\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Added from iDK1463\n",
    "\n",
    "# # Load reference model\n",
    "# ref_model = cobra.io.load_json_model('../data/iAF1260.json')\n",
    "\n",
    "# # Add reactions to EcN\n",
    "# EcN_model.add_reactions([ref_model.reactions.MTRK, ref_model.reactions.MTRI])\n",
    "\n",
    "# # Add genes\n",
    "# EcN_model.reactions.MTRK.gene_reaction_rule = 'CIW80_17025'\n",
    "# EcN_model.reactions.MTRI.gene_reaction_rule = 'CIW80_17030'\n",
    "\n",
    "# # Add gene names\n",
    "# EcN_model.genes.CIW80_17025.name = 'mtnK (methylthioribose (MTR) kinase)'\n",
    "# EcN_model.genes.CIW80_17030.name = 'mtnA (methylthioribose-1-phosphate (MTR-1-P) isomerase)'\n",
    "\n",
    "# for rxn in ['MTRK', 'MTRI']:\n",
    "#     # Add to reaction overview\n",
    "#     rxn_origin_df.loc[rxn, 'origin'] = 'iAF1260'\n",
    "#     rxn_origin_df.loc[rxn, 'added'] = 'manual'\n",
    "#     rxn_origin_df.loc[rxn, 'notebook'] = '4.6'\n",
    "#     rxn_origin_df.loc[rxn, 'notes'] = 'Added based on Sun et al. (2005) in notebook 4.6'\n",
    "        \n",
    "# for gene in ['CIW80_17025', 'CIW80_17030']:\n",
    "#     # Register the genes\n",
    "#     gene_origin_df.loc[gene, 'EcN_gene'] = gene\n",
    "#     gene_origin_df.loc[gene, 'origin'] = 'iAF1260'\n",
    "#     gene_origin_df.loc[gene, 'added'] = 'manual'\n",
    "#     gene_origin_df.loc[gene, 'notebook'] = '4.6'\n",
    "#     gene_origin_df.loc[gene, 'notes'] = 'Added based on Sun et al. (2005) in notebook 4.6'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Gene identifier</strong></td><td>CIW80_17025</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>S-methyl-5-thioribose kinase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x026f57b8ef48</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Functional</strong></td><td>True</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 1 reaction(s)</strong></td><td>\n",
       "                    MTRK</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Gene CIW80_17025 at 0x26f57b8ef48>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EcN_model.genes.CIW80_17025"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Others\n",
    "Overview from Sun et al. (2005) > Enzymes present in E. coli Nissle\n",
    "\n",
    "- 2.3.1.12\tDihydrolipoamide acetyltransferase component (E2) of pyruvate dehydrogenase complex,\tGlycolysis, gluconeogenesis and pyruvate metabolism > PDH, present\n",
    "- 2.7.1.51\tFucK; l-fuculokinase,\tFructose and mannose metabolism > FCLK, present\n",
    "- 2.8.3.1\tPropionate CoA-transferase,\tPyruvate and propanoate metabolism > PPCSCT, present\n",
    "- 3.5.3.6\tArginine deiminase,\tArginine and proline metabolism > ARGDr, present\n",
    "- 1.1.1.31\t3-Hydroxyisobutyrate dehydrogenase,\tVal, Leu and Ile degradation <font color=red> > HIBD, not present. Pathway not present in the model </font>\n",
    "- 1.1.3.15\t(S)-2-Hydroxy-acid oxidase,\tGlyoxylate and dicarboxylate metabolism > <font color=red> no reaction on BiGG. </font>\n",
    "- 2.7.1.53\tl-Xylulokinase,\tPentose and glucuronate interconversions > XYLK/XYLK2, present\n",
    "- 2.7.1.58\t2-Dehydro-3-deoxygalactonokinase,\tGalactose metabolism > DDGALK, present\n",
    "- 2.8.3.8\tAcetate CoA-transferase,\tPropanoate and butanoate metabolism > ACACCT, present\n",
    "- 3.2.1.14\tChitinase,\tAminosugars metabolism > <font color=red> no reaction on BiGG. </font>\n",
    "\n",
    "Other reactions were not added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Save model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1536"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(EcN_model.genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the model\n",
    "cobra.io.json.save_json_model(EcN_model, str('../data/models/%s_cur_4.6.json'%EcN_ID))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save reaction and gene origin tables\n",
    "rxn_origin_df.to_csv('../tables/rxn_origin.csv')\n",
    "gene_origin_df.to_csv('../tables/gene_origin.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
