{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4.8 BOFdat\n",
    "Adapt the biomass objective function using BOFdat, based on experimental data of EcN and K12. \n",
    "\n",
    "- https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006971\n",
    "- https://bofdat.readthedocs.io/en/latest/\n",
    "- https://github.com/jclachance/BOFdat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='red'> !!! Run notebook in python 2.7 environment (bofdat.yml) !!! </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from BOFdat import step1\n",
    "from BOFdat import step2\n",
    "from BOFdat import step3\n",
    "from BOFdat.util import update\n",
    "\n",
    "import cobra\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1. Macromolecular composition\n",
    "The general macromolecular composition for Escherichia coli (E.coli K12 MG1655) is used.\n",
    "- https://github.com/jclachance/BOFdat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set parameters based on dry weight composition\n",
    "dna_weight_fraction = 0.031\n",
    "rna_weight_fraction = 0.205\n",
    "protein_weight_fraction = 0.55\n",
    "lipid_weight_fraction = 0.1\n",
    "metabolite_weight_fraction = 0.1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNA\n",
    "The EcN genome is available on NCBI\n",
    "- https://www.ncbi.nlm.nih.gov/assembly/GCF_003546975.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Give the path to each file as function parameters\n",
    "#Genome file in BioPython supported format (.faa, .fna) and GenBank file\n",
    "#also in BioPython supported format (.gb, .gbff)\n",
    "genome = '../data/genomes_gb/CP022686.1.fna'\n",
    "genbank = '../data/genomes_gb/CP022686.1.gb'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transcriptomics\n",
    "Transcriptomics data of EcN from the study by Zhao et al. 20220 was used\n",
    "\n",
    "- https://www.sciencedirect.com/science/article/pii/S2667370322000030?via%3Dihub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Omic datasets as a 2 column csv file, gene and abundance\n",
    "transcriptomic = '../tables/EcN_transcriptomics.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Proteomics\n",
    "Proteomic data of E.coli K12 MG1655 was used, as no data was available for EcN.\n",
    "- https://github.com/jclachance/BOFdat\n",
    "\n",
    "Gene IDs in the proteomic data needs to be converted to EcN IDs.\n",
    "Construction of the converted df only needed first time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #import packages needed\n",
    "# from glob import glob\n",
    "# from Bio import Entrez, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### Get all gene annotations from MG1655\n",
    "\n",
    "# in_file = '../tables/Ecoli_K12_MG1655.gbff'\n",
    "\n",
    "# gene_annot_df = pd.DataFrame(columns=('locus', 'ncbigene', 'ncbiprotein'))\n",
    "# gene_annot_df.set_index('locus', inplace=True)\n",
    "\n",
    "# handle = open(in_file)\n",
    "\n",
    "# records = SeqIO.parse(handle, \"genbank\")\n",
    "# for record in records:\n",
    "#     for f in record.features:\n",
    "#         if f.type=='CDS':\n",
    "\n",
    "#             # Get locus tag\n",
    "#             if 'locus_tag' in f.qualifiers.keys():\n",
    "#                 locus = f.qualifiers['locus_tag'][0]\n",
    "            \n",
    "# #             # get gene id\n",
    "#             if 'gene' in f.qualifiers.keys():\n",
    "#                 gene_id = f.qualifiers['gene'][0]\n",
    "#                 gene_annot_df.loc[locus, 'ncbigene'] = gene_id\n",
    "            \n",
    "#             # get protein_id\n",
    "#             if 'protein_id' in f.qualifiers.keys():\n",
    "#                 prot_id = f.qualifiers['protein_id'][0]\n",
    "#                 gene_annot_df.loc[locus, 'ncbiprotein'] = prot_id\n",
    "\n",
    "# gene_annot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### Get all gene annotations from EcN\n",
    "\n",
    "# in_file = '../data/genomes_gb/CP022686.1.gb'\n",
    "\n",
    "# EcN_annot_df = pd.DataFrame(columns=('locus', 'ncbigene', 'ncbiprotein'))\n",
    "# EcN_annot_df.set_index('locus', inplace=True)\n",
    "\n",
    "# handle = open(in_file)\n",
    "\n",
    "# records = SeqIO.parse(handle, \"genbank\")\n",
    "# for record in records:\n",
    "#     for f in record.features:\n",
    "#         if f.type=='CDS':\n",
    "\n",
    "#             # Get locus tag\n",
    "#             if 'locus_tag' in f.qualifiers.keys():\n",
    "#                 locus = f.qualifiers['locus_tag'][0]\n",
    "            \n",
    "# #             # get gene id\n",
    "#             if 'gene' in f.qualifiers.keys():\n",
    "#                 gene_id = f.qualifiers['gene'][0]\n",
    "#                 EcN_annot_df.loc[locus, 'ncbigene'] = gene_id\n",
    "            \n",
    "#             # get protein_id\n",
    "#             if 'protein_id' in f.qualifiers.keys():\n",
    "#                 prot_id = f.qualifiers['protein_id'][0]\n",
    "#                 EcN_annot_df.loc[locus, 'ncbiprotein'] = prot_id\n",
    "\n",
    "# EcN_annot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Open proteomics data iML1515\n",
    "# proteomic_df = pd.read_csv('../tables/proteomic.csv')\n",
    "# proteomic_df\n",
    "\n",
    "# # Reset index gene_annot_df\n",
    "# gene_annot_df = gene_annot_df.reset_index().set_index('ncbiprotein')\n",
    "\n",
    "# # Merge dataframes\n",
    "# MG1655_df = pd.merge(proteomic_df, gene_annot_df, left_on='gene_id', right_index=True)\n",
    "\n",
    "# # Convert to EcN gene IDs\n",
    "# conv_df = pd.read_csv('../tables/MG_geneIDs_matrix.csv', names=['MG1655', 'EcN'])\n",
    "# gene_conv_df = pd.merge(MG1655_df, conv_df, left_on='locus', right_on='MG1655')\n",
    "\n",
    "# # Create a df with EcN protein ID and abundance\n",
    "# EcN_proteomic_df = pd.merge(gene_conv_df, EcN_annot_df, left_on='EcN', right_index=True)\n",
    "# EcN_proteomic_df.set_index('ncbiprotein', inplace=True)\n",
    "# EcN_proteomic_df.drop(columns=['gene_id', 'locus', 'ncbigene_x', 'MG1655', 'EcN', 'ncbigene_y'], inplace=True)\n",
    "# EcN_proteomic_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Save as .csv file\n",
    "# EcN_proteomic_df.to_csv('../tables/EcN_proteomics.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomic = '../tables/EcN_proteomics.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lipids\n",
    "Lipidomic data of E.coli K12 MG1655 was used, as no data was available for EcN.\n",
    "- https://github.com/jclachance/BOFdat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Lipidomic abundances and conversion to model identifier\n",
    "lipidomic_abundances = '../tables/lipidomic_abundances.csv'\n",
    "lipidomic_conversion = '../tables/lipidomic_conversion.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.1 Maintenance\n",
    "Maintenance data of E.coli K12 MG1655 was used, as limited data was available for EcN.\n",
    "- https://github.com/jclachance/BOFdat\n",
    "\n",
    "Add the growth data from Revelles paper (glucose/gluconate)\n",
    "- https://doi.org/10.1371/journal.pone.0066386"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Growth data on different carbon sources, uptake and secretion rates\n",
    "# maintenance = '../tables/maintenance.csv'\n",
    "maintenance = '../tables/maintenance_ecn.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EcN model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load EcN model\n",
    "model = '../data/models/CP022686.1_cur_4.7.json'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 Macromolecules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{<Metabolite dttp_c at 0x115b68d0>: -0.025264995594222028,\n",
       " <Metabolite dctp_c at 0x115c0240>: -0.027205885733218617,\n",
       " <Metabolite datp_c at 0x115d46a0>: -0.024535580087683178,\n",
       " <Metabolite ppi_c at 0x115d4780>: 0.10089456005683937,\n",
       " <Metabolite dgtp_c at 0x116d4f28>: -0.02388809864171554}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_coefficients = step1.generate_dna_coefficients(genome,model,DNA_WEIGHT_FRACTION=dna_weight_fraction)\n",
    "dna_coefficients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\maxvan\\Anaconda3\\envs\\bofdat\\lib\\site-packages\\BOFdat\\core\\rna.py:60 UserWarning: Some identifiers not found in provided annotation\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{<Metabolite ctp_c at 0x1500b550>: -0.16705693200583052,\n",
       " <Metabolite atp_c at 0x1503f860>: -0.15682670811169883,\n",
       " <Metabolite utp_c at 0x1505f780>: -0.15393794867659485,\n",
       " <Metabolite ppi_c at 0x15069748>: 0.6397639025268087,\n",
       " <Metabolite gtp_c at 0x150d1c50>: -0.16194231373268453}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_coefficients = step1.generate_rna_coefficients(genbank,model,transcriptomic,RNA_WEIGHT_FRACTION=rna_weight_fraction)\n",
    "rna_coefficients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\maxvan\\Anaconda3\\envs\\bofdat\\lib\\site-packages\\BOFdat\\core\\protein.py:64 UserWarning: Redundancy in dataset identifiers\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{<Metabolite his__L_c at 0x19fb1cc0>: -0.0784666222129269,\n",
       " <Metabolite val__L_c at 0x19fc6208>: -0.4236326541330141,\n",
       " <Metabolite pro__L_c at 0x19fc6240>: -0.23785545643823697,\n",
       " <Metabolite asn__L_c at 0x19fda278>: -0.1988011065795899,\n",
       " <Metabolite leu__L_c at 0x19fda4e0>: -0.43210232676094734,\n",
       " <Metabolite met__L_c at 0x1a19a898>: -0.10251662872662275,\n",
       " <Metabolite tyr__L_c at 0x1a1a4e48>: -0.09992310063926794,\n",
       " <Metabolite cys__L_c at 0x1a1a4eb8>: -0.046655642921015865,\n",
       " <Metabolite trp__L_c at 0x1a1ae940>: -0.030522659777398792,\n",
       " <Metabolite lys__L_c at 0x1a1aea90>: -0.25290101506014606,\n",
       " <Metabolite glu__L_c at 0x1a219da0>: -0.2831016692802292,\n",
       " <Metabolite ile__L_c at 0x1a219f28>: -0.2916454170578126,\n",
       " <Metabolite gln__L_c at 0x1a224dd8>: -0.16515881771324858,\n",
       " <Metabolite ala__L_c at 0x1a2384e0>: -0.8051754272737,\n",
       " <Metabolite phe__L_c at 0x1a238630>: -0.12966182233807855,\n",
       " <Metabolite h2o_c at 0x1a242a20>: 5.470463534121725,\n",
       " <Metabolite gly_c at 0x1a242e48>: -0.8259745895550139,\n",
       " <Metabolite thr__L_c at 0x1a5564a8>: -0.3022121994386853,\n",
       " <Metabolite ser__L_c at 0x1a568518>: -0.3181480851031188,\n",
       " <Metabolite arg__L_c at 0x1a5729b0>: -0.15400440615477717,\n",
       " <Metabolite asp__L_c at 0x1a572d30>: -0.2920038869578946}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_coefficients = step1.generate_protein_coefficients(genbank,model,proteomic,PROTEIN_WEIGHT_FRACTION=protein_weight_fraction)\n",
    "protein_coefficients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\maxvan\\Anaconda3\\envs\\bofdat\\lib\\site-packages\\BOFdat\\core\\lipid.py:76 UserWarning: Redundancy in dataset identifiers\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{<Metabolite pe160_p at 0x1d5033c8>: -0.05984233670391253,\n",
       " <Metabolite pg160_p at 0x1d503400>: -0.014826578855136979,\n",
       " <Metabolite pe181_p at 0x1d50df98>: -0.013250952451946145,\n",
       " <Metabolite pg181_p at 0x1d52c550>: -0.009098089940190764,\n",
       " <Metabolite pe161_p at 0x1d56ceb8>: -0.030710743257354293,\n",
       " <Metabolite pg161_p at 0x1d576358>: -0.013732749319021109}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lipid_coefficients = step1.generate_lipid_coefficients(lipidomic_abundances,lipidomic_conversion,model,LIPID_WEIGHT_FRACTION=lipid_weight_fraction)\n",
    "lipid_coefficients"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.3 Maintenance (GAM & NGAM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cobra\\util\\solver.py:403 UserWarning: solver status is 'infeasible'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('m', 73.039820474777997, 'b', 13.948331959461735)\n",
      "('R2=', 0.67931748410829607)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEICAYAAACwDehOAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi41LCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvSM8oowAAGw9JREFUeJzt3XtwVPX9xvFnc4HEhBBQvGFhEm4GtQVkAh0qVYQJHdDhZgIZA9ZOqQwMhqFAihBovSATisVYQJxOi4EAsaWAqS0zBCtyMQoEqMwKHSz5lQhCazIkIZcNe35/0ByJ7GaD7J6zl/frH8jZePbzTeQ8e7634zAMwxAAAJKi7C4AABA8CAUAgIlQAACYCAUAgIlQAACYCAUAgIlQAACYCAUAgIlQAACYYuwu4GYdPXpU8fHxdpdhqaamJnXu3NnuMixFmyNDpLXZzvY2NTVp0KBBPr8v5ELB4XAoLS3N7jIs5XQ6aXMEoM3hz872Op3ODn0f3UcAABOhAAAwEQoAABOhAAAwEQoAABOhAAAwEQoAABOhAAAwEQoAABOhAAAwEQoAABOhAAAwEQoAABOhAAAwEQoAAFPAQuH48ePKycmRJFVWVmratGnKzs7WsmXL5Ha7JUklJSWaNGmSMjMz9f777weqFABABwUkFN566y0tWbJETU1NkqQVK1YoNzdXxcXFMgxDZWVlunTpkoqKirR161b97ne/0+rVq9Xc3ByIcgAAHRSQUOjVq5cKCwvNr0+ePKn09HRJ0siRI3Xw4EGdOHFCgwcPVqdOndSlSxf16tVLn332WSDKAQB0UEAex5mRkaFz586ZXxuGIYfDIUlKSEhQbW2t6urq1KVLF/N7EhISVFdX5/Pcbre7w4+VCxeNjY20OQLQ5vAXCu215BnNUVFf35DU19crKSlJiYmJqq+vb3P8+pBo71yR9ExXKfKeYyvR5kgRaW3mGc3/M3DgQJWXl0uS9u3bp6FDh+q73/2ujhw5oqamJtXW1urMmTPq37+/FeUA+BZ2VFRpxKt7lZL3F414da92VFTZXRICwJI7hUWLFmnp0qVavXq1UlNTlZGRoejoaOXk5Cg7O1uGYWjevHnq3LmzFeUAuEk7Kqr0i+3/UIPrqiSpqqZBv9j+D0nShME97SwNfhawULjvvvtUUlIiSUpJSdGmTZtu+J7MzExlZmYGqgQAflKw+5QZCK0aXFdVsPsUoRBmWLwGwKcvahpu6jhCF6EAwKd7k+Nv6jhCF6EAwKcFGQMUHxvd5lh8bLQWZAywqSIEiiUDzQBCW+u4QcHuU/qipkH3JsdrQcYAxhPCEKEAoEMmDO5JCEQAuo8AACZCAQBgIhQAACZCAQBgIhQAACZCAQBgIhQAACZCAQBgIhQAACZCAQBgIhQAACZCAQBgIhQAACZCAQBgIhQAACZCAQBgIhQAACZCAQBg4nGcACyxo6Lqhmc8D4izuyp8E6EAIOB2VFTpF9v/oQbXVUlSVU2DfrH9H5ozvLvS0mwuroM8hVo4PrOa7iMAAVew+5QZCK0aXFe18Wi1TRXdnNZQq6ppkKGvQ21HRZXdpfkdoQAg4L6oafB4/FJ9i8WVfDveQq1g9ymbKgocQgFAwN2bHO/xeI+E0OjB9hZq3o6HMkIBQMAtyBig+NjoNsfiY6M1Y0g3myq6Od5CzdvxUBYaMQ0gpLUOyN44++iyzZVd42sQeUHGgDYD5dK1UFuQMcCOcgOKUABgiQmDe94wW8fptD8UvM2Mkr4OM2+hFo6zjwgFAGHN111Ae4PI13+fp1C72fcKBZaFgsvlUl5enqqqqhQVFaUXX3xRMTExysvLk8PhUL9+/bRs2TJFRTHMAcA/OnIX4K9B5I68Vyiw7Ar8wQcfqKWlRVu3btXs2bP1m9/8RitWrFBubq6Ki4tlGIbKysqsKgdABOjIVFJvg8WGpBGv7u3wWoRwmbZqWSikpKTo6tWrcrvdqqurU0xMjE6ePKn09HRJ0siRI3Xw4EGrygEQATpyF+BpZlSrm1mkFi7TVi3rPrrttttUVVWlH/3oR6qurtb69ev1ySefyOFwSJISEhJUW1vr8zxut1tOpzPQ5QaVxsZG2hwBaLP/9UiI0UUPC+R6JMSY7zsgTpozvLs2Hq32+L0Nrqt6pfRTnzOlOvJeofA7tiwU/vCHP+gHP/iB5s+fr/Pnz2vGjBlyuVzm6/X19UpKSvJ5nqioKKWFymYpfuJ0OmlzBKDN/rd4fJLHqaSLxz+otLSv+/nT0qTZ46SUvL/I8HCeS/UtPuvsyHvZ+TvuaBhZFgpJSUmKjY2VJHXt2lUtLS0aOHCgysvLNWzYMO3bt0/Dhw+3qhwAHRDqs2ludirpvcnxqvLQ3dORRWodfa9g/5laFgrPPPOMFi9erOzsbLlcLs2bN08PPvigli5dqtWrVys1NVUZGRlWlQPAB0+zaeZtO6bDlV/ppQkP2Vxdx3VkKmmrW12k5uu99n5eqzc+qgzqGUqWhUJCQoLWrFlzw/FNmzZZVQKAm+BpNo0hafNH/ydJev+zS0H7affbCvQitY1Hqzu0JsJOLF4D4JG3WTOtwdDa9x6Mn3Zvxc3cWdwsb7vCBtMMJVaKAfCovX70bw7GhuJ8fDt42xU2mDbWIxQAeHSzm70F06fdYDVjSDePu8UG08Z6dB8B8CpKkruD33tvcnzQz6yx26jULup5b8+g/hkRCgA8Kth9ymMgxMdGSXLcMEPnsft7hMXeP4EWyDELf6D7CIBH3rqDGl1urZj0kHomx8shqWdyvFZMekjvf3YpLPb+iXTcKQDwqL2FXJ4+7c7bdszjeRhrCC3cKQDwyNsjNL0NikbSIyvDGaEAwKMJg3t67Cby1h9+syGC4ET3EUIeM14C52YGRSPpkZXhjFBASAuXp12Fi2CfWQPf6D5CSAuXp10BwYJQQEgLl6ddAcGCUEBIY8YL4F+EAkIaM14A/2KgGSGNGS+AfxEKCHnMeAH8h+4jAICJUAAAmOg+AtAuXyvGI3FFeTi3mVAA4JWvFeORuKI83NtM9xEAr3ytGI/EFeXh3mZCAYBXvlaMR+KK8nBvM6EAwCtfK8YjcUV5uLeZUADgla8V45G4ojzc28xAMwCvfK0Yj8QV5eHeZkIBwA1ap1xW1TQo2uHQVcNQTy8Xv0hcUR7ObSYUALTxzSmXVw1DUvhNvYRnjCkAaMPTlMtW4TT1Ep4RCgDa8DW1MlymXsIzQgFAG76mVobL1Et4ZumYwptvvqm9e/fK5XJp2rRpSk9PV15enhwOh/r166dly5YpKoqcAuy0IGNAmzGF64XT1Et4ZtkVuLy8XBUVFdqyZYuKiop04cIFrVixQrm5uSouLpZhGCorK7OqHABeTBjcUysmPaSe/7sjiHY4JEk9k+O1YtJDDDKHOcvuFPbv36/+/ftr9uzZqqur08KFC1VSUqL09HRJ0siRI3XgwAGNGTPGqpIAeBHOUy7RPstCobq6Wl988YXWr1+vc+fOadasWTIMQ47/fQpJSEhQbW2tz/O43W45nc5AlxtUGhsbaXMEoM3hLxTaa1koJCcnKzU1VZ06dVJqaqo6d+6sCxcumK/X19crKSnJ53mioqKUlpYWyFKDjtPppM0RgDaHPzvb29Ew8jqm8PHHH2vSpEnKycnR2bNnb7mghx9+WB9++KEMw9CXX36phoYGff/731d5ebkkad++fRo6dOgtvw8A4Nvzeqfw2muvqaCgQDU1NVq9erVef/31W3qjxx57TJ988ommTJkiwzCUn5+v++67T0uXLtXq1auVmpqqjIyMW3oPIFR4enLXgDi7qwLaCYXY2Fj16dNHklRYWOiXN1u4cOENxzZt2uSXcwOhwtuTu+YM764I6klBkOrQlFS32x3oOoCI4e3JXRuPVttUEfA1r3cKX375pbZt22aOAWzbts18LSsry5LigHDkbZuIS/UtFlcC3MhrKDzxxBO6dOnSDX8HcGvuTY5XlYdg6JEQ+MmAnsYyWI+A63n9v/Cee+7R5MmTrawFsJwdF0lP20jEx0ZrxpBuAX1fb2MZElth42texxR27txpZR2A5VovklU1DTL09UVyR0VVQN/3+m0kHPp6+4hRqV0C+r7exjLYChvX83qn0NDQoLNnz8r43wM2rpeSkhLQogArtHeRDPQnZ0/bSDidlwP6nt7GMtgKG9fzGgpnz55Vfn7+DaHgcDj09ttvB7wwINAi7SLpbSyDrbBxPa+hcP/993PxR1iLtIukt7EMtsLG9W5q6+zq6mpt2LAhULUAllqQMUDxsdFtjoXzRdLbWAaDzLie1zuFNWvWmH8/ceKENm/erP3797MVBcJG68UwkqZosiU2fPEaComJifrzn/+szZs3q1OnTqqrq1NZWZni4tigBeGDiyTQltfuo1GjRunUqVNatWqViouLdeeddxIIABDmvN4pTJ8+XaWlpaqqqjJ3NgUAhDevdwozZ87Url27lJOTo9LSUn366acqKCjQ6dOnrawPiCg7Kqo04tW9Ssn7i0a8ujfgC+mAb/K52Up6errS09N1+fJl7dy5UwsXLtSOHTusqA2IKHs/r9UbH1WyDQVs1eEpqUlJScrJySEQgADZeLSabShgu5tapwAgcLxtnR2uK6wRnAK/Vy9gg1DcIrpHQowuegiGcF1hjeDEnQLCjl27n96qGUO6RdQKawQnQgFhJ1S3iB6V2oVtKGA7uo8QdkJ591NWWMNu3Ckg7Hjrg6dvHvCNUEDYibTdTwF/ovsIYScSdz8F/IVQQFiibx74dug+AgCYCAUAgInuIwSVtiuRzzMWAFiMUEDQaF2JzC6hgH0IBQQNbyuRf/nuSWYSARYhFBA0vK04rr7iUvUVl6Rrdw+5244pd9sxSVJyfKyWP/kAIQH4CQPNCBrfZsVxTYNLC945HvSb3QGhwvJQ+O9//6sf/vCHOnPmjCorKzVt2jRlZ2dr2bJlcrvdVpeDIOJpJXJHuNxG0G92B4QKS0PB5XIpPz9fcXFxkqQVK1YoNzdXxcXFMgxDZWVlVpaDIDNhcM8bdglNjo/t0H8bCpvdAaHA0lBYuXKlpk6dqjvvvFOSdPLkSaWnp0uSRo4cqYMHD1pZDoLQhME9dSBvlN6bkaoDeaO0/MkHOnT3wGZ3gH9YNtC8fft2de/eXY888og2bNggSTIMQw6HQ5KUkJCg2tpan+dxu91yOp0BrTXYNDY2RmybB8RJc4Z318aj1R6fSiZJ0Q4p+6HEkP8ZRfLvOVKEQnstC4U//elPcjgcOnTokJxOpxYtWqSvvvrKfL2+vl5JSUk+zxMVFaW0tLRAlhp0nE5nRLc5LU2aPe7a8R0VVfrluyfN2UjhNPso0n/PkcDO9nY0jCwLhc2bN5t/z8nJ0fLly1VQUKDy8nINGzZM+/bt0/Dhw60qByGKje6AwLJ1SuqiRYtUWFiorKwsuVwuZWRk2FkOAEQ8WxavFRUVmX/ftGmTHSUAADxg8RoAwMQ2FyGq7W6i7AcEwD8IhRDEbqIAAoXuoxDkbTdRtnoAcKsIhRDkbUsHtnoAcKsIhRDkbUsHtnoAcKsIhRDkaTfR+NhoLcgYYFNFAMIFA80hqHUwmdlHAPyNUAhRbPcAIBDoPgIAmAgFAICJUAAAmAgFAICJUAAAmAgFAICJUAAAmAgFAICJUAAAmAgFAICJUAAAmAgFAICJDfFgC54xDQQnQgGW4xnTQPCi+wiW4xnTQPAiFGA5njENBC9CAZbjGdNA8CIUYDmeMQ0ELwaaYTmeMQ0EL+4UAAAm7hRgOaakAsGLOwVYjimpQPCy7E7B5XJp8eLFqqqqUnNzs2bNmqW+ffsqLy9PDodD/fr107JlyxQVRU6FO6akAsHLslDYtWuXkpOTVVBQoOrqak2cOFH333+/cnNzNWzYMOXn56usrExjxoyxqiTY5N7keFV5CACmpAL2s+xj+dixY/X888+bX0dHR+vkyZNKT0+XJI0cOVIHDx60qhzYiCmpQPCy7E4hISFBklRXV6e5c+cqNzdXK1eulMPhMF+vra31eR632y2n0xnQWoNNY2NjWLV5QJw0Z3h3bTxarUv1LeqREKMZQ7ppQNxlOZ2XJYVfmzuCNoe/UGivpbOPzp8/r9mzZys7O1tPPPGECgoKzNfq6+uVlJTk8xxRUVFKS0sLZJlBx+l0hl2b09Kk2eO8vx6ObfaFNoc/O9vb0TCyrPvoP//5j5599lktWLBAU6ZMkSQNHDhQ5eXlkqR9+/Zp6NChVpUDAPDAslBYv369Ll++rLVr1yonJ0c5OTnKzc1VYWGhsrKy5HK5lJGRYVU5AAAPLOs+WrJkiZYsWXLD8U2bNllVAgDABxYFAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwEQoAABMhAIAwBRjdwFut1vLly/XqVOn1KlTJ7300kvq3bu33WUBQESy/U5hz549am5u1rZt2zR//ny9+uqrdpcEABHL9lA4cuSIHnnkEUnSoEGD9Omnn9pcEQBELtu7j+rq6pSYmGh+HR0drZaWFsXEeC7N4XDI6XRaVV7QoM2RgTaHP7va29TU1KHvsz0UEhMTVV9fb37tdru9BoJ07W4CABAYtncfDRkyRPv27ZMkHTt2TP3797e5IgCIXA7DMAw7C2idfXT69GkZhqFXXnlFffr0sbMkAIhYtocCACB42N59BAAIHoQCAMAUlKHgdruVn5+vrKws5eTkqLKyss3re/fu1eTJk5WVlaWSkhKbqvQvX20uLS3VU089palTpyo/P19ut9umSv3HV5tbLV26VKtWrbK4usDw1eYTJ04oOztb06ZN09y5czs8jTCY+Wrzrl27NHHiRE2ePFnFxcU2VRkYx48fV05Ozg3Hg/oaZgSh3bt3G4sWLTIMwzAqKiqM5557znytubnZGD16tFFTU2M0NTUZkyZNMi5evGhXqX7TXpsbGhqMxx9/3Lhy5YphGIYxb948Y8+ePbbU6U/ttbnVli1bjMzMTKOgoMDq8gKivTa73W7jySefNM6ePWsYhmGUlJQYZ86csaVOf/L1ex4xYoRRXV1tNDU1mf+2w8GGDRuM8ePHG0899VSb48F+DQvKO4X2VjmfOXNGvXr1UteuXdWpUyc9/PDDOnz4sF2l+k17be7UqZO2bt2q+Ph4SVJLS4s6d+5sS53+5Gs1e0VFhY4fP66srCw7yguI9tr8r3/9S8nJydq4caOefvpp1dTUKDU11a5S/cbX73nAgAGqra1Vc3OzDMOQw+Gwo0y/69WrlwoLC284HuzXsKAMBW+rnFtf69Kli/laQkKC6urqLK/R39prc1RUlO644w5JUlFRka5cuaIRI0bYUqc/tdfmixcv6o033lB+fr5d5QVEe22urq5WRUWFsrOz9fvf/14fffSRDh06ZFepftNemyWpX79+mjx5ssaNG6dHH31USUlJdpTpdxkZGR4X4gb7NSwoQ6G9Vc7ffK2+vr7NDzhU+VrZ7Xa7tXLlSh04cECFhYVh8WmqvTb/7W9/U3V1tWbOnKkNGzaotLRU27dvt6tUv2mvzcnJyerdu7f69u2r2NhYPfLII2GxF1h7bf7ss8/097//XWVlZdq7d6+++uor/fWvf7WrVEsE+zUsKEOhvVXOffr0UWVlpWpqatTc3KzDhw9r8ODBdpXqN75Wdufn56upqUlr1641u5FCXXttnj59urZv366ioiLNnDlT48eP16RJk+wq1W/aa/N3vvMd1dfXmwOxhw8fVr9+/Wyp05/aa3OXLl0UFxenzp07Kzo6Wt27d9fly5ftKtUSwX4Ns33vI0/GjBmjAwcOaOrUqeYq53fffVdXrlxRVlaW8vLy9JOf/ESGYWjy5Mm666677C75lrXX5gcffFB//OMfNXToUM2YMUPStYvmmDFjbK761vj6PYcjX21++eWXNX/+fBmGocGDB+vRRx+1u+Rb5qvNWVlZys7OVmxsrHr16qWJEyfaXXJAhMo1jBXNAABTUHYfAQDsQSgAAEyEAgDARCgAAEyEAgDARCggIvz73//W3LlzlZmZqenTp2vmzJn65z//6ZdzNzU16Z133pEkFRYWasuWLbd8zk2bNt3yOYBvg1BA2GtoaNCsWbP04x//WCUlJXr77bc1Z84c/epXv/LL+S9dumSGgr+sW7fOr+cDOop1Cgh77733no4ePaolS5a0Od66+VpeXp5qampUU1OjN998U+vWrdORI0ckSePHj9eTTz6pZ555Rjt37lRFRYV+9rOf6dChQ7p06ZJeeOEF3XPPPXrvvff07LPPyjAMHT9+XC6XSzU1NXr++ec1atQo8z3Ly8u1atUqxcbGKjMzU3Fxcdq8ebP5+po1a7Rt2zb99re/1ZQpU/TCCy9o2bJlqqyslNvtVm5uroYNG2bNDw4RiTsFhL1z586pV69e5tezZs1STk6Oxo4dqwsXLkiShg8frq1bt+ro0aM6d+6cSkpKVFxcrNLSUl28eFHJyck6f/68PvzwQ9199906efKkysrKNHr0aD333HPq27ev5syZI0m66667tHHjRi1evNhjV1JTU5OKi4s1YcIEnT17Vhs2bFBRUZFSUlK0f/9+zZo1S127dtXy5cv1zjvvqFu3btq8ebPWrl3rt7sbwJug3OYC8Ke77767zcZyrV0zmZmZ5m6dKSkpkq5tazx06FA5HA7Fxsbqe9/7ns6cOaMxY8bogw8+UEVFhX7605/qwIEDqqio0Msvv3zDg3AeeOABSdIdd9yhxsbGG+ppfS9Juv3227Vo0SIlJCTo888/16BBg9p87+nTp3XkyBGdOHFC0rVt06urq9WtW7db/bEAHnGngLD3+OOP69ChQzp27Jh5rLKyUhcuXDB3m239s0+fPmbXkcvlUkVFhXr37q3Ro0ertLRUiYmJGjlypPbs2aPm5mb16NFDUVFRbZ6E52sH26ioa//samtr9frrr+u1117TSy+9pM6dO6u1N7f1z9TUVI0bN05FRUV66623NHbsWHXt2tVPPxngRtwpIOwlJCRo3bp1+vWvf61Vq1appaVFMTExevHFF9WzZ8823/vYY4/p448/VlZWllwul8aOHWt+8m9qatLw4cPVtWtXxcTEmJvV3X777XK5XCooKFBcXFyH60pMTNSQIUM0ceJE3XbbbUpKStLFixclXQunn//853rllVe0ZMkSPf3006qrq1N2drYZKkAgMNAMADDxkQMAYCIUAAAmQgEAYCIUAAAmQgEAYCIUAAAmQgEAYCIUAACm/wcSszOYpxrZygAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "maintenance_cost = step1.generate_maintenance_costs(maintenance,model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.4 Update model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "json_model = cobra.io.load_json_model(model)\n",
    "bofdat_step1 = update.make_new_BOF(json_model,False,True,dna_coefficients,rna_coefficients,protein_coefficients,\n",
    "                    lipid_coefficients,maintenance=maintenance_cost)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "bofdat_step1.to_csv('../tables/bofdat_step1.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 2. Coenzymes & ions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 Inorganic ions\n",
    "\n",
    "The file BIOMASS_universal_components.csv needs to be added to the BOFdat folder\n",
    "- https://github.com/jclachance/BOFdat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_model = '../data/models/CP022686.1_cur_4.7.json'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating list of solvable metabolites\n"
     ]
    }
   ],
   "source": [
    "# path_to_model = 'data/iML1515.json'\n",
    "selected_metabolites = step2.find_coenzymes_and_ions(path_to_model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determine coefficients\n",
    "The default weight_fraction of 0.05 (5%) is used "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from BOFdat.util.update import determine_coefficients\n",
    "WEIGHT_FRACTION = 0.05\n",
    "bd_step2 = determine_coefficients(selected_metabolites,json_model,WEIGHT_FRACTION)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Update biomass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\maxvan\\Anaconda3\\envs\\bofdat\\lib\\site-packages\\BOFdat\\util\\update.py:22 UserWarning: File has 3 columns, assuming index on first column\n"
     ]
    }
   ],
   "source": [
    "#Get the input from the previous step\n",
    "path_to_biomass = '../tables/bofdat_step1.csv'\n",
    "bd_step1 = update.convert_to_dictionary(path_to_biomass)\n",
    "bd_step2.update(bd_step1)\n",
    "\n",
    "# Save updated biomass\n",
    "update.save_biomass(bd_step2,'../tables/bofdat_step2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove u char\n",
    "import json, ast\n",
    "bd_step2_u = ast.literal_eval(json.dumps(bd_step2))\n",
    "\n",
    "# Update biomass\n",
    "update.save_biomass(bd_step2_u,'../tables/bofdat_step2.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 3. Finding specie-specific metabolic end goals"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This step was skipped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Update BOF of EcN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load step 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # Load bd_steps\n",
    "df_bd2 = pd.read_csv('../tables/bofdat_step2.csv', usecols=['Metabolites', 'Coefficients'])\n",
    "df_bd2.set_index('Metabolites', inplace=True)\n",
    "\n",
    "bd2 = {}\n",
    "\n",
    "for mtb in df_bd2.index:\n",
    "    bd2[mtb] = df_bd2.loc[mtb, 'Coefficients']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Update model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load model\n",
    "EcN_model = cobra.io.load_json_model('../data/models/CP022686.1_cur_4.7.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EcN_model.reactions.BIOMASS_EcN_iHM1533_core_59p80M.reaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('pg161_p', 'not in BIOMASS')\n",
      "('pe181_p', 'not in BIOMASS')\n",
      "('na1_c', 'not in BIOMASS')\n",
      "('pyr_c', 'not in BIOMASS')\n",
      "('nadh_c', 'not in BIOMASS')\n",
      "('amp_c', 'not in BIOMASS')\n",
      "('pg181_p', 'not in BIOMASS')\n",
      "('glyc3p_c', 'not in BIOMASS')\n",
      "('nadph_c', 'not in BIOMASS')\n",
      "('pg160_p', 'not in BIOMASS')\n"
     ]
    }
   ],
   "source": [
    "for mtb in bd2:\n",
    "    coef = bd2[mtb]\n",
    "    \n",
    "    # Remove metabolite from existing biomass reaction\n",
    "    try:\n",
    "        comp = EcN_model.reactions.BIOMASS_EcN_iHM1533_core_59p80M.get_coefficient(mtb)\n",
    "        EcN_model.reactions.BIOMASS_EcN_iHM1533_core_59p80M.add_metabolites({mtb: -comp})\n",
    "    except:\n",
    "        print(mtb, 'not in BIOMASS')\n",
    "        \n",
    "# Add all metabolites from step 2\n",
    "EcN_model.reactions.BIOMASS_EcN_iHM1533_core_59p80M.add_metabolites(bd2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EcN_model.reactions.BIOMASS_EcN_iHM1533_core_59p80M.reaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove iML1515 biomass reactions from EcN_model\n",
    "EcN_model.remove_reactions(['BIOMASS_Ec_iML1515_core_75p37M', 'BIOMASS_Ec_iML1515_WT_75p37M'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Update reaction list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the rxn_origin dataframe and update biomass reactions\n",
    "rxn_origin_df = pd.read_csv('../tables/rxn_origin.csv', usecols=['reaction', 'origin', 'added', 'notebook', 'notes'])\n",
    "rxn_origin_df.set_index('reaction', inplace=True)\n",
    "\n",
    "rxn_origin_df.loc['BIOMASS_Ec_iML1515_core_75p37M', 'added'] = 'removed'\n",
    "rxn_origin_df.loc['BIOMASS_Ec_iML1515_core_75p37M', 'notes'] = 'Reaction was replaced by BIOMASS_EcN_iHM1533_core_59p80M in 4.8'\n",
    "\n",
    "rxn_origin_df.loc['BIOMASS_Ec_iML1515_WT_75p37M', 'added'] = 'removed'\n",
    "rxn_origin_df.loc['BIOMASS_Ec_iML1515_WT_75p37M', 'notes'] = 'Reaction was replaced by BIOMASS_EcN_iHM1533_core_59p80M in 4.8'\n",
    "\n",
    "rxn_origin_df.loc['BIOMASS_EcN_iHM1533_core_59p80M', 'origin'] = 'New Reaction'\n",
    "rxn_origin_df.loc['BIOMASS_EcN_iHM1533_core_59p80M', 'added'] = 'manual'\n",
    "rxn_origin_df.loc['BIOMASS_EcN_iHM1533_core_59p80M', 'notebook'] = '4.8'\n",
    "rxn_origin_df.loc['BIOMASS_EcN_iHM1533_core_59p80M', 'notes'] = 'Reaction added in 4.8. Replaces original iML1515 biomass objective functions'\n",
    "\n",
    "# Save\n",
    "rxn_origin_df.to_csv('../tables/rxn_origin.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save updated model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update model ID\n",
    "EcN_model.id = 'iHM1533' #SBML only accepts \"a string comprising only alphanumeric characters and underscores starting with either a letter or an underscore\"\n",
    "\n",
    "# Save the model\n",
    "cobra.io.json.save_json_model(EcN_model, '../data/models/CP022686.1_cur_4.8.json')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
